<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Miguel Camacho Sánchez">

<title>tidyGenR benchmarking</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="benchmarking_files/libs/clipboard/clipboard.min.js"></script>
<script src="benchmarking_files/libs/quarto-html/quarto.js"></script>
<script src="benchmarking_files/libs/quarto-html/popper.min.js"></script>
<script src="benchmarking_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="benchmarking_files/libs/quarto-html/anchor.min.js"></script>
<link href="benchmarking_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="benchmarking_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="benchmarking_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="benchmarking_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="benchmarking_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#benchmarking-tidygenr" id="toc-benchmarking-tidygenr" class="nav-link active" data-scroll-target="#benchmarking-tidygenr">Benchmarking <em>tidyGenR</em></a></li>
  <li><a href="#preparation-of-input-data" id="toc-preparation-of-input-data" class="nav-link" data-scroll-target="#preparation-of-input-data">Preparation of input data</a></li>
  <li><a href="#demultiplex-by-locus" id="toc-demultiplex-by-locus" class="nav-link" data-scroll-target="#demultiplex-by-locus">Demultiplex by locus</a></li>
  <li><a href="#truncate-reads" id="toc-truncate-reads" class="nav-link" data-scroll-target="#truncate-reads">Truncate reads</a></li>
  <li><a href="#exploration-of-the-parameter-space" id="toc-exploration-of-the-parameter-space" class="nav-link" data-scroll-target="#exploration-of-the-parameter-space">Exploration of the parameter space</a></li>
  <li><a href="#variant-and-genotype-calling" id="toc-variant-and-genotype-calling" class="nav-link" data-scroll-target="#variant-and-genotype-calling">Variant and genotype calling</a></li>
  <li><a href="#tidy-data" id="toc-tidy-data" class="nav-link" data-scroll-target="#tidy-data">Tidy data</a></li>
  <li><a href="#genotype-with-amplisat" id="toc-genotype-with-amplisat" class="nav-link" data-scroll-target="#genotype-with-amplisat">Genotype with AmpliSAT</a></li>
  <li><a href="#compare-results" id="toc-compare-results" class="nav-link" data-scroll-target="#compare-results">Compare results</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session Info</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">tidyGenR benchmarking</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Miguel Camacho Sánchez </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<p>Attach libraries:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyGenR)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># declare logicals to trigger EVAL in intensive code chunks</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># they are not run if output already exists.</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># to run fresh analysis outputs need to be removed.</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># demultiplex</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>c_log <span class="ot">&lt;-</span> <span class="st">"output/cutadapt.log"</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>eval_demultiplex <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(c_log)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># fastqc</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>multiqc_report <span class="ot">&lt;-</span> <span class="st">"output/multiqc_report.html"</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>eval_fastqc <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(multiqc_report)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co"># truncate</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>tr_path <span class="ot">&lt;-</span> <span class="st">"output/trunc_in-out.rds"</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>eval_truncate <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(tr_path)</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="co"># explore_dada</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>x_out_path <span class="ot">&lt;-</span> <span class="st">"output/explore_dada.rds"</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>eval_x <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(x_out_path)</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co"># variant_call</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>x_variants_path <span class="ot">&lt;-</span>  <span class="st">"output/variants_x.rds"</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>eval_variant_call <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(x_variants_path)</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="co"># genotype</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>x_genotypes_path <span class="ot">&lt;-</span>  <span class="st">"output/genotypes_x.rds"</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>eval_genotype <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(x_genotypes_path)</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="co"># amplisas</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>amplisas_results <span class="ot">&lt;-</span> <span class="st">""</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>eval_amplisas <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(amplisas_results)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="benchmarking-tidygenr" class="level2">
<h2 class="anchored" data-anchor-id="benchmarking-tidygenr">Benchmarking <em>tidyGenR</em></h2>
<p>This repository accompanies the R package <strong>tidyGenR</strong> available at <a href="https://github.com/csmiguel/tidyGenR" class="uri">https://github.com/csmiguel/tidyGenR</a>. It covers (1) all steps for genotype calling with <em>tidyGenR</em>, (2) exploration of the parameters for variant calling, and (3) comparison of <em>tidyGenR</em> against <a href="https://doi.org/10.1111/1755-0998.12453">AmpliSAS</a>. It uses real data from a population genetics study of the wild rodent <em>Rattus baluensis</em> (Camacho-Sanchez et al.&nbsp;<em>in preparation</em>).</p>
</section>
<section id="preparation-of-input-data" class="level2">
<h2 class="anchored" data-anchor-id="preparation-of-input-data">Preparation of input data</h2>
<p>Raw sequences are deposited in NCBI under the BioStudy <a href="https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP293699">SRP293699</a>, in the BioProject <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA680166">PRJNA680166</a>. From the BioStudy, we can download the <a href="data/raw/SraRunTable.csv">SRA metadata</a> to mapping later sample IDs to SRRs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># table with mapped SRR-sampleIDs</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>srr <span class="ot">&lt;-</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read.csv</span>(<span class="st">"data/raw/SraRunTable.csv"</span>) <span class="sc">|&gt;</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Run, <span class="st">"Sample.Name"</span>) <span class="sc">|&gt;</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">run =</span> Run,</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">sample =</span> <span class="st">"Sample.Name"</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(srr, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">run</th>
<th style="text-align: left;">sample</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">SRR13106799</td>
<td style="text-align: left;">BOR1063</td>
</tr>
<tr class="even">
<td style="text-align: left;">SRR13106800</td>
<td style="text-align: left;">BOR1061</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SRR13106802</td>
<td style="text-align: left;">BOR626</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Raw FASTQ reads from the SRRs can be downloaded with <a href="https://github.com/ncbi/sra-tools">SRA-toolkit</a>. Then, they are renamed based on the <code>data/raw/SraRunTable.csv</code> mapping file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download raw reads from NCBI</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># they are not downloaded if already present in 'data/raw'.</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">invisible</span>(</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">apply</span>(srr, <span class="dv">1</span>, <span class="cf">function</span>(x) {</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># name paths</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  srr_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, x[<span class="dv">1</span>])</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  srr1 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">1</span>], <span class="st">"_1.fastq"</span>))</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  srr2 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">1</span>], <span class="st">"_2.fastq"</span>))</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  sam1 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">2</span>], <span class="st">"_1.fastq"</span>))</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  sam2 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">2</span>], <span class="st">"_2.fastq"</span>))</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># fetch SRR</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">any</span>(<span class="fu">dir.exists</span>(srr_path) <span class="sc">||</span> <span class="fu">file.exists</span>(sam1))) {</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"prefetch"</span>, <span class="fu">paste</span>(<span class="st">"-O"</span>, <span class="st">"data/raw/"</span>, x[<span class="dv">1</span>]))</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(sam1)) {</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># reformat to FASTQ</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"fasterq-dump"</span>, <span class="fu">paste</span>(<span class="st">"-O"</span>, <span class="st">"data/raw/"</span>, srr_path))</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># rename SRR with sample codes</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"mv"</span>, <span class="fu">paste</span>(srr1, sam1))</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"mv"</span>, <span class="fu">paste</span>(srr2, sam2))</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>    <span class="co"># rm SRR</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"rm"</span>, <span class="fu">paste0</span>(<span class="st">"-r "</span>, srr_path, <span class="st">"*"</span>))</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>List some of the downloaded files:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(<span class="st">"data/raw"</span>, <span class="at">pattern =</span> <span class="st">"fastq"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data/raw/BOR1061_1.fastq" "data/raw/BOR1061_2.fastq"
[3] "data/raw/BOR1063_1.fastq"</code></pre>
</div>
</div>
<p>Check input raw fastq:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>freads <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/raw"</span>, <span class="at">pattern =</span> <span class="st">"1.fastq"</span>,</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>rreads <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/raw"</span>, <span class="at">pattern =</span> <span class="st">"2.fastq"</span>,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>chr <span class="ot">&lt;-</span> <span class="fu">check_raw_reads</span>(freads, rreads, <span class="at">low_readcount =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>All F and (R files) passed check on number of reads above 10.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Sample names are unique.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>All F files have their corresponding R file.</code></pre>
</div>
</div>
<p>Input FASTQ complies with expected input for <em>tidyGenR</em>. A total of 44 samples are detected:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>chr<span class="sc">$</span>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "BOR1061" "BOR1063" "BOR1069" "BOR1070" "BOR1071" "BOR1074" "BOR1075"
 [8] "BOR1076" "BOR1077" "BOR1079" "BOR1080" "BOR1082" "BOR1086" "BOR1088"
[15] "BOR1090" "BOR1091" "BOR1097" "BOR1098" "BOR1101" "BOR1107" "BOR1111"
[22] "BOR1113" "BOR1115" "BOR1125" "BOR1127" "BOR1129" "BOR1131" "BOR1133"
[29] "BOR1135" "BOR1137" "BOR1139" "BOR1141" "BOR1143" "BOR1147" "BOR1149"
[36] "BOR1151" "BOR1155" "BOR614"  "BOR615"  "BOR616"  "BOR618"  "BOR619" 
[43] "BOR621"  "BOR626" </code></pre>
</div>
</div>
</section>
<section id="demultiplex-by-locus" class="level2">
<h2 class="anchored" data-anchor-id="demultiplex-by-locus">Demultiplex by locus</h2>
<p>Reads are demultiplexed by locus using primer sequences in paired-end mode.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load primer data</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"primers"</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># path to cutadapt</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>cutadapt <span class="ot">&lt;-</span> <span class="fu">system</span>(<span class="st">"which cutadapt"</span>, <span class="at">intern =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># path to folder save locus-demultiplexed FASTQ</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>demult <span class="ot">&lt;-</span> <span class="st">"data/intermediate/demultiplexed"</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co"># print primers</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(primers, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">locus</th>
<th style="text-align: left;">fw</th>
<th style="text-align: left;">rv</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">nfkbia</td>
<td style="text-align: left;">GCCTCCAAACACACAGTCAT</td>
<td style="text-align: left;">TGAGGAGAGCTATGACACGG</td>
</tr>
<tr class="even">
<td style="text-align: left;">chrna9</td>
<td style="text-align: left;">TTATCTGGGAGAGCGTGACC</td>
<td style="text-align: left;">TTGGGAAARGATGAACCGGC</td>
</tr>
<tr class="odd">
<td style="text-align: left;">rogdi</td>
<td style="text-align: left;">AGAARCCGGCTCACTACCC</td>
<td style="text-align: left;">GAGGCACAGCTTGTTGAGG</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># demultiplex</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">demultiplex</span>(</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">interpreter =</span> <span class="st">"/bin/bash"</span>,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">cutadapt =</span> cutadapt,</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">freads =</span> freads,</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">rreads =</span> rreads,</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">primers =</span> primers,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">sh_out =</span> <span class="st">"code/demultiplex.sh"</span>,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">write_demultiplexed =</span> demult,</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">log_out =</span> c_log,</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">mode =</span> <span class="st">"pe"</span>,</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">run =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Glimpse demultiplexed FASTQ:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(demult, <span class="at">pattern =</span> <span class="st">"fastq"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data/intermediate/demultiplexed/BOR1061.abcg8.1.fastq.gz" 
[2] "data/intermediate/demultiplexed/BOR1061.abcg8.2.fastq.gz" 
[3] "data/intermediate/demultiplexed/BOR1061.alkbh7.1.fastq.gz"</code></pre>
</div>
</div>
<p>Remove files with few reads:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">remove_poor_fastq</span>(demult,</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">min_reads =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Make sequence-quality reports using FastQC and MultiQC. Instead of running it over the &gt;2000 demultiplexed files, it will be run over 100 random files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># instead of running fastqc to all files &gt; 2000, I run</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co">#    it on 100 random files.</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>dem_files <span class="ot">&lt;-</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>(demult, <span class="st">"fastq.gz"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># select 100 random files</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>random_dem <span class="ot">&lt;-</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample</span>(dem_files, <span class="dv">100</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co"># copy to temp dir</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>temp_dem <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">tempdir</span>(), <span class="st">"dem"</span>)</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(temp_dem)</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="fu">file.copy</span>(<span class="at">from =</span> random_dem, <span class="at">to =</span> temp_dem)</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="co"># run fastqc</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="fu">system2</span>(<span class="st">"fastqc"</span>,</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>        <span class="fu">paste</span>(<span class="st">"--noextract -o"</span>,</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>              temp_dem,</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>              <span class="fu">paste0</span>(temp_dem, <span class="st">"/*fastq.gz"</span>)))</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a><span class="co"># run multiqc</span></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a><span class="fu">system2</span>(<span class="st">"multiqc"</span>,</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>        <span class="fu">paste</span>(<span class="st">"-o output"</span>, temp_dem))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="truncate-reads" class="level2">
<h2 class="anchored" data-anchor-id="truncate-reads">Truncate reads</h2>
<p>Reads are truncated to a given length for each locus. The truncation lengths depend on the sequence qualities for forward and reverse reads. After per-base sequence-qualities in the <a href="output/multiqc_report.html">output/multiqc_report.html</a>, <strong>270</strong> nt for forward reads and <strong>180</strong> for reverse reads seem reasonable truncation lengths. For some loci, specific truncation lengths were set to maximize the number of reads returned. That is, a <code>data.frame</code> with truncation lengths for forward and reverse reads was built to maximize the amount of information yielded by each locus. For instance, the amplicons for some loci, as <em>nfkbia</em>, are long and it is a good trade-off to keep the low quality ends but to be sure both F and R reads overlap.</p>
<p>A <code>data.frame</code> with locus-specific truncation lengths was built.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>loci <span class="ot">&lt;-</span> </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  tidyGenR<span class="sc">:::</span><span class="fu">check_names_demultiplexed</span>(demult,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">fw_pattern =</span> <span class="st">"1.fastq.gz"</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">rv_pattern =</span> <span class="st">"2.fastq.gz"</span>)<span class="sc">$</span>loci</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>trunc_fr <span class="ot">&lt;-</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(<span class="at">locus =</span> loci,</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">trunc_f =</span> <span class="dv">270</span>,</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">trunc_r =</span> <span class="dv">180</span>)</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co"># introduce manual values</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"rgd735029"</span>), <span class="st">"trunc_f"</span>] <span class="ot">&lt;-</span>  <span class="dv">245</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"rgd735029"</span>), <span class="st">"trunc_r"</span>] <span class="ot">&lt;-</span>  <span class="dv">155</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"fancg"</span>), <span class="st">"trunc_r"</span>] <span class="ot">&lt;-</span> <span class="dv">190</span></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"nfkbia"</span>), <span class="st">"trunc_r"</span>] <span class="ot">&lt;-</span> <span class="dv">240</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"tmem87a"</span>), <span class="st">"trunc_r"</span>] <span class="ot">&lt;-</span> <span class="dv">160</span></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse data.frame with locus-specific truncation lengths</span></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(trunc_fr, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">locus</th>
<th style="text-align: right;">trunc_f</th>
<th style="text-align: right;">trunc_r</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">abcg8</td>
<td style="text-align: right;">270</td>
<td style="text-align: right;">180</td>
</tr>
<tr class="even">
<td style="text-align: left;">alkbh7</td>
<td style="text-align: right;">270</td>
<td style="text-align: right;">180</td>
</tr>
<tr class="odd">
<td style="text-align: left;">apeh17</td>
<td style="text-align: right;">270</td>
<td style="text-align: right;">180</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Truncate reads according to locus-specific truncation lengths:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>tr_dir <span class="ot">&lt;-</span> <span class="st">"data/intermediate/truncated"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># truncate</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>trunc_out <span class="ot">&lt;-</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">trunc_amp</span>(<span class="at">in_dir =</span> demult,</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">fw_pattern =</span> <span class="st">"1.fastq.gz"</span>,</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">rv_pattern =</span> <span class="st">"2.fastq.gz"</span>,</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">trunc_fr =</span> trunc_fr,</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">write_trun =</span> tr_dir,</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">max_ee =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">5</span>),</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">trunc_q =</span> <span class="dv">2</span>)</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="co"># save reads in and out</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(trunc_out, tr_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The output from <code>trunc_amp()</code> is a list of matrices with IN and OUT reads after truncation.</p>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># see trunc_out</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(trunc_out[<span class="fu">seq_len</span>(<span class="dv">3</span>)], head, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$abcg8
        reads.in reads.out
BOR1061     1716      1039
BOR1063     1616       992
BOR1069     1634       961

$alkbh7
        reads.in reads.out
BOR1061      546       398
BOR1063      499       361
BOR1069      546       374

$apeh17
        reads.in reads.out
BOR1061     3719      2773
BOR1063     3761      2793
BOR1069     3756      2838</code></pre>
</div>
</div>
<p>Truncated FASTQ with low number of reads can be removed:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">remove_poor_fastq</span>(tr_dir,</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">min_reads =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>2376 files have been KEPT.
0 files REMOVED:</code></pre>
</div>
</div>
</section>
<section id="exploration-of-the-parameter-space" class="level2">
<h2 class="anchored" data-anchor-id="exploration-of-the-parameter-space">Exploration of the parameter space</h2>
<p>The function <code>explore_dada()</code> can be used to explore the effect of some DADA2 parameters (“OMEGA_A”, “BAND_SIZE”, “pool”) on the sensitivity on the variant calling.</p>
<ul>
<li><em>omega_a</em>: threshold for variants to be significant overabundant <em>log(-log(birth_pval))</em> (see <a href="https://doi.org/10.1186/1471-2105-13-283">Rosen et al.&nbsp;2012</a>).</li>
<li><em>band_size</em>: positive numbers set a band size in Needleman-Wunsch alignments and ends free alignment is performed. A value of zero turns off banding, triggering full Needleman-Wunsch alignments, in which gapless alignment is performed (see <a href="https://github.com/benjjneb/dada2/issues/1982">issue</a>).</li>
<li><em>pool</em>: calling variants pooling samples can increase sensitivity (see <a href="https://benjjneb.github.io/dada2/pseudo.html">dicussion</a>).</li>
</ul>
<p>The returned plots can be used to guide the selection of the best <em>OMEGA_A</em> in <code>variant_call()</code>, or frequency (<em>maf</em>) and abundance thresholds (<em>ad</em>) for filtering variants.</p>
<p>Explore variants:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># declare candidate OMEGA_A to use in variant_call()</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>candidate_omega_a <span class="ot">&lt;-</span> <span class="dv">10</span><span class="sc">^-</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># paths to forward and reverse truncated reads</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>ftrun <span class="ot">&lt;-</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>(tr_dir, <span class="at">pattern =</span> <span class="st">"_F_"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>rtrun <span class="ot">&lt;-</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>(tr_dir, <span class="at">pattern =</span> <span class="st">"_R_"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="co"># candidate omega value to annotate vline in plots.</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>v_line <span class="ot">&lt;-</span> <span class="fu">log</span>(<span class="sc">-</span><span class="fu">log</span>(candidate_omega_a))</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a><span class="co"># run explore_dada() with band_size = 0, non pooling and omega_a = 0.9</span></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="co"># forward</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>F_ind_0 <span class="ot">&lt;-</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">explore_dada</span>(ftrun, <span class="at">band_size =</span> <span class="dv">0</span>, <span class="at">vline =</span> v_line, <span class="at">hline_fr =</span> <span class="fl">0.1</span>)</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a><span class="co"># reverse</span></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>R_ind_0 <span class="ot">&lt;-</span></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">explore_dada</span>(rtrun, <span class="at">band_size =</span> <span class="dv">0</span>, <span class="at">vline =</span> v_line, <span class="at">hline_fr =</span> <span class="fl">0.1</span>)</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a><span class="co"># save results</span></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(<span class="fu">list</span>(<span class="at">F_ind_0 =</span> F_ind_0, <span class="at">R_ind_0 =</span> R_ind_0), x_out_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Exploration of DADA2 clustering for forward (A, C) and reverse (B, D) reads in <a href="#fig-explore_dada">Figure&nbsp;1</a>. The Y-axis represents the frequency of the variant in each locus and sample. The <em>log(-log(birth_pval))</em> transformation in the X-axis is related to the <em>p</em>-value of a variant being significantly overabundant. Larger x-values represent likely true variants. For representation purposes <em>birth_pval</em> of 0 (thus negative infinite), are converted to 10. Points are color-coded according to the variant rank in read abundance for its given locus and sample. For diploid individuals, <span style="color: green;">green</span> are likely true variants and <span style="color: red;">red</span> are likely false variants. <span style="color: grey;">Grey</span> dashed lines are thresholds used for <code>variant_call()</code>:</p>
<div class="cell">

</div>
<div class="cell" data-warnings="false">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>ppool <span class="ot">&lt;-</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  (F_ind_0<span class="sc">$</span>p1 <span class="sc">|</span> R_ind_0<span class="sc">$</span>p1) <span class="sc">/</span> (F_ind_0<span class="sc">$</span>p2 <span class="sc">|</span> R_ind_0<span class="sc">$</span>p2) <span class="sc">+</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="st">"Dada F/R pool = F, omega_a 0.9, band_size = 0"</span>,</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">tag_levels =</span> <span class="st">"A"</span>)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="co"># save plot with combined loci</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"output/explore_dada.pdf"</span>, ppool, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">5</span>)</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="co"># print plot</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>ppool</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-explore_dada" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="benchmarking_files/figure-html/fig-explore_dada-1.png" class="img-fluid figure-img" width="768"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;1: Exploration of DADA2 variants.</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>After the exploration variants in <a href="#fig-explore_dada">Figure&nbsp;1</a>, it seems an <em>OMEGA_A</em> = 0.01, implying a cut-off of 1.53 in the X-axis and a frequency threshold (Y-axis) of 0.1, excludes most <span style="color: red;">artifacts</span> while maximizing <span style="color: green;">true positives</span>.</p>
<p>The results can also be explored <strong>per-locus</strong>. For instance, <a href="#fig-explore_dada">Figure&nbsp;1</a> B can be expanded per locus in <a href="#fig-explore_dada_loci">Figure&nbsp;2</a>.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># list of plots per locus</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>lplots <span class="ot">&lt;-</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(<span class="at">loci_f_ind_0_logp =</span> F_ind_0<span class="sc">$</span>p3,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">loci_r_ind_0_logp =</span> R_ind_0<span class="sc">$</span>p3,</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">loci_f_ind_0_abun =</span> F_ind_0<span class="sc">$</span>p4,</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">loci_r_ind_0_abun =</span> R_ind_0<span class="sc">$</span>p4)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save plots per locus</span></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a><span class="fu">invisible</span>(</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(<span class="fu">seq_along</span>(lplots), <span class="cf">function</span>(x) {</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">paste0</span>(<span class="st">"output/"</span>, <span class="fu">names</span>(lplots)[x], <span class="st">".pdf"</span>),</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>         lplots[[x]], <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">6</span>)</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-warnings="false">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>lplots<span class="sc">$</span>loci_r_ind_0_logp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-explore_dada_loci" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="benchmarking_files/figure-html/fig-explore_dada_loci-1.png" class="img-fluid figure-img" style="width:90.0%"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2: Exploration of DADA2 variants per locus for R reads.</figcaption><p></p>
</figure>
</div>
</div>
</div>
</section>
<section id="variant-and-genotype-calling" class="level2">
<h2 class="anchored" data-anchor-id="variant-and-genotype-calling">Variant and genotype calling</h2>
<p>Variant calling is run using <em>OMEGA_A</em> = 0.01, and under different parameters:</p>
<ul>
<li><em>band_size</em>: 0, 16</li>
<li><em>pool</em>: TRUE, FALSE</li>
</ul>
<div class="cell">

</div>
<p>Samples are genotyped from variant data with defaults <code>ploidy = 2</code> and <code>ADt = 10</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>genotypes_x <span class="ot">&lt;-</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(variants_x, genotype)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">

</div>
</section>
<section id="tidy-data" class="level2">
<h2 class="anchored" data-anchor-id="tidy-data">Tidy data</h2>
<p>A strength from <em>tidyGenR</em> is that variants and genotypes from <code>variant_call()</code> and <code>genotype()</code> are returned as <code>tidy</code> tables: one row per observation and variables in columns. Lets have a look at the data structure.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse tidy variants</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(variants_x<span class="sc">$</span>ind_bs0))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 6%">
<col style="width: 86%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">sample</th>
<th style="text-align: left;">locus</th>
<th style="text-align: left;">variant</th>
<th style="text-align: right;">reads</th>
<th style="text-align: right;">nt</th>
<th style="text-align: left;">md5</th>
<th style="text-align: left;">sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">abcg8</td>
<td style="text-align: left;">01</td>
<td style="text-align: right;">1028</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">c3eab42bea56937040aa79a6bbea2724</td>
<td style="text-align: left;">TTGCCCACCCTGTTCATCCATGGAGCAGAAGCCTGCCTGATGTCTCTCATCATTGGCTTCCTTTACTACGGCCACGCAGATAAGCCGCTCTCCTTCGTGGACATGGCAGCCCTCCTGTTCATGATAGGAGCGCTCATTCCTTTTAATGTCATTCTGGATGTCGTCTCCAAATGTGAGTGTCACCCGCCCTCCTCACCAGACATCGGGACAGTGGGACAGCCTCCCTGGGCACTGCACTGAGGCCAAGCTCTGTGCTTCCGCTGGTACCCACGGCATTACAAGAGATGCGACCTCAGTAACACTCTTCGCTCATTCACCTCCCTCTCCCCCATCTCCAGGTCACTCGGAGCGCTCGCTGCTGTACTATGAACTGGAGGACGGACTGTACAC</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">alkbh7</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">393</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">620057f2e547e688d1220db91a3a1d01</td>
<td style="text-align: left;">GCCTTCCTGGCCCCATCCCCTCTGGGAGGGAGCGGCAAATCACTGAGATGCGTCGGCCCCGGGGGACCCGGTGAGCCCCAAAAAATGATTCTTCATCTCTAAGGATCTCATGGGAGAAGTCATATCGGGCTGAACCCCTGCAAGAAAATAAAGGCCAAGTAGGTGAAGGGAGGGAAGTCTGTCCCTTCATCATTTCTGTACTTTATCTGGGTAGTGGTGATGGTGACTTCCTCTGCTATGAGAGCAGAAACAAGAGCTGTTGGAAACACTTGGCCTCATCTGGGGGTTCTAGGCTAGGTCATACCTTAGGATATAGAGAGAACCAGGTTCCAGCAACAGTTCCAGCCACTGCTCAGGTTCCTGTGTATGAACCAGCTTCATAACACTTGG</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">apeh17</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">2747</td>
<td style="text-align: right;">415</td>
<td style="text-align: left;">394a483d3ae2ce5605f1674b2400986c</td>
<td style="text-align: left;">AAAGCCAGTGGAGCCACGATAGTTCACTGCAAGTGAGACAAGATAGTCGGGCAATGTTCTCAGTCCCCAACCCAATTCCATGTGTCTGTGACATAGGCCCTAGGACCAGCTGCATCACGTGCTGTCAAGGGGAGGTAGCCAAGAGAGATGAAGCTGCTACCCTGAGACCACAGCATCCTTTCCACAGAAGCTTCAAGTTAATCCACGTGATCACCAGGTTATGAAGCCAGAGGCTAAGCCAGGGCAATTTCTAGTAGTAGTTTCTTTTTGTTTTCAAAACAGGGTTTTTCTGTGCAGTCCTGGATGATCTTGGACTCAAACCTTCACCTGCCTCTGCCTCCTAAGTGCTGGATTAAAGGAGTGTGCCCCCATTGCCCAGCCCGTTTCTAAAGAGTAAGTACACCATAGAAAAGCA</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">cd27</td>
<td style="text-align: left;">01</td>
<td style="text-align: right;">1143</td>
<td style="text-align: right;">379</td>
<td style="text-align: left;">a897a500c934797d4b3662415fc92456</td>
<td style="text-align: left;">AGTCTTCCTGGATAGGGATGACGCTGCCCTCCTCCTCCCTGGGGCAGCTGTAAGGACAAAGCTCTTCAGGTACTGCCTGGCTATCTTCATCTGTGCAAAGACAATTAGCCAAGTGTTGGTCAGCAGTGGAGAGAAGAGAGGGGAAGGTGAGGAGAGAGGAGAAGGCCGAGTGGAGGCTGGGGCATGGGGGAGCCAGGGGAGCCTGTGGGAAGGACACTTGAAGAACCAGAGAAGGTGGGTGAAGGTGGGATGGGGGCTTTAGGTGTGGGTGGCAGAGCTGAGAGGGCAGGAGGGAAGGCCTGTGCCTTACTTGGCCCGTGATTTCTTCTTTGACGGAAGAACAAGATCCCACCCAGGACGAAGACAAGAAGCATGCTGG</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">chrna9</td>
<td style="text-align: left;">01</td>
<td style="text-align: right;">21</td>
<td style="text-align: right;">408</td>
<td style="text-align: left;">3d77c726462281336567895361ceebc1</td>
<td style="text-align: left;">TGCAGTGTGACATTCAGCACCGCGTCCGTATCCTCGACTGGACGCAGAGCACTGGAGTAGTCTTCAAAAAGATCGCTGAACAATTTCTGAGCATATTTCCCATTTGCTGTCTCTACGGCTGTTCAAAGAGAAGCACCCGGATGGGCATTTCAAAACAAGACCAACTCTGGGGTCAAAATGGGGGATGGCAACTTTGGATGAGTTCTTTTTTGGGGGCAGGGGACTGGTCTTACCTACCTTGATGATCGAGAGTCAGAAAGAGGTGCGAGCACGCCCATGTACTCACGTCAGTCCTCCTGTCTCTCTCACCAAGGCTCTAAGGAGCCTCTATGGAGCACTGCGCTAGCCCCTCACCTCTGATTCCAGAAGCAGCAAAATACATCCAGCAAAAGGAGATGCAGGACTGGG</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">chrna9</td>
<td style="text-align: left;">02</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">408</td>
<td style="text-align: left;">9b805048a29d6233f1ac41f2a6aa1421</td>
<td style="text-align: left;">TGCAGTGTGACATTCAGCACCGCGTCCGTATCCTCGACTGGACGCAGAGCACTGGAGTAGTCTTCAAAAAGATCGCTGAATAATTTCTGAGCATATTTCCCATTTGCTGTCTCTACGGCTGTTCAAAGAGAAGCACCCGGATGGGCATTTCAAAACAAGACCAACTCTGGGATCAAAATGGGGGATGGCAACTTTGGATGAGTTCTTTTTTGGGGGCAGGGGACTGGTCTTACCTACCTTGATGATCAAGAGTCAGAAAGAGGTGCGAGCACGCCCATGTACTCACGTCAGTCCTCCTGTCTCTCTCACCAAGGCTCTAAGGAGCCTCTATGGAGCACTGCGCTAGCCCCTCACCTCTGATTCCAGAAGCAGCAAAATACATCCAGCAAAAGGAGATGCAGGACTGGG</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse tidy genotypes</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(genotypes_x<span class="sc">$</span>ind_bs0))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 6%">
<col style="width: 84%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">sample</th>
<th style="text-align: left;">locus</th>
<th style="text-align: left;">allele</th>
<th style="text-align: right;">allele_no</th>
<th style="text-align: right;">reads</th>
<th style="text-align: right;">nt</th>
<th style="text-align: left;">md5</th>
<th style="text-align: left;">sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">abcg8</td>
<td style="text-align: left;">01</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">514.0</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">c3eab42bea56937040aa79a6bbea2724</td>
<td style="text-align: left;">TTGCCCACCCTGTTCATCCATGGAGCAGAAGCCTGCCTGATGTCTCTCATCATTGGCTTCCTTTACTACGGCCACGCAGATAAGCCGCTCTCCTTCGTGGACATGGCAGCCCTCCTGTTCATGATAGGAGCGCTCATTCCTTTTAATGTCATTCTGGATGTCGTCTCCAAATGTGAGTGTCACCCGCCCTCCTCACCAGACATCGGGACAGTGGGACAGCCTCCCTGGGCACTGCACTGAGGCCAAGCTCTGTGCTTCCGCTGGTACCCACGGCATTACAAGAGATGCGACCTCAGTAACACTCTTCGCTCATTCACCTCCCTCTCCCCCATCTCCAGGTCACTCGGAGCGCTCGCTGCTGTACTATGAACTGGAGGACGGACTGTACAC</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">abcg8</td>
<td style="text-align: left;">01</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">514.0</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">c3eab42bea56937040aa79a6bbea2724</td>
<td style="text-align: left;">TTGCCCACCCTGTTCATCCATGGAGCAGAAGCCTGCCTGATGTCTCTCATCATTGGCTTCCTTTACTACGGCCACGCAGATAAGCCGCTCTCCTTCGTGGACATGGCAGCCCTCCTGTTCATGATAGGAGCGCTCATTCCTTTTAATGTCATTCTGGATGTCGTCTCCAAATGTGAGTGTCACCCGCCCTCCTCACCAGACATCGGGACAGTGGGACAGCCTCCCTGGGCACTGCACTGAGGCCAAGCTCTGTGCTTCCGCTGGTACCCACGGCATTACAAGAGATGCGACCTCAGTAACACTCTTCGCTCATTCACCTCCCTCTCCCCCATCTCCAGGTCACTCGGAGCGCTCGCTGCTGTACTATGAACTGGAGGACGGACTGTACAC</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">alkbh7</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">196.5</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">620057f2e547e688d1220db91a3a1d01</td>
<td style="text-align: left;">GCCTTCCTGGCCCCATCCCCTCTGGGAGGGAGCGGCAAATCACTGAGATGCGTCGGCCCCGGGGGACCCGGTGAGCCCCAAAAAATGATTCTTCATCTCTAAGGATCTCATGGGAGAAGTCATATCGGGCTGAACCCCTGCAAGAAAATAAAGGCCAAGTAGGTGAAGGGAGGGAAGTCTGTCCCTTCATCATTTCTGTACTTTATCTGGGTAGTGGTGATGGTGACTTCCTCTGCTATGAGAGCAGAAACAAGAGCTGTTGGAAACACTTGGCCTCATCTGGGGGTTCTAGGCTAGGTCATACCTTAGGATATAGAGAGAACCAGGTTCCAGCAACAGTTCCAGCCACTGCTCAGGTTCCTGTGTATGAACCAGCTTCATAACACTTGG</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">alkbh7</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">196.5</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">620057f2e547e688d1220db91a3a1d01</td>
<td style="text-align: left;">GCCTTCCTGGCCCCATCCCCTCTGGGAGGGAGCGGCAAATCACTGAGATGCGTCGGCCCCGGGGGACCCGGTGAGCCCCAAAAAATGATTCTTCATCTCTAAGGATCTCATGGGAGAAGTCATATCGGGCTGAACCCCTGCAAGAAAATAAAGGCCAAGTAGGTGAAGGGAGGGAAGTCTGTCCCTTCATCATTTCTGTACTTTATCTGGGTAGTGGTGATGGTGACTTCCTCTGCTATGAGAGCAGAAACAAGAGCTGTTGGAAACACTTGGCCTCATCTGGGGGTTCTAGGCTAGGTCATACCTTAGGATATAGAGAGAACCAGGTTCCAGCAACAGTTCCAGCCACTGCTCAGGTTCCTGTGTATGAACCAGCTTCATAACACTTGG</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">apeh17</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1373.5</td>
<td style="text-align: right;">415</td>
<td style="text-align: left;">394a483d3ae2ce5605f1674b2400986c</td>
<td style="text-align: left;">AAAGCCAGTGGAGCCACGATAGTTCACTGCAAGTGAGACAAGATAGTCGGGCAATGTTCTCAGTCCCCAACCCAATTCCATGTGTCTGTGACATAGGCCCTAGGACCAGCTGCATCACGTGCTGTCAAGGGGAGGTAGCCAAGAGAGATGAAGCTGCTACCCTGAGACCACAGCATCCTTTCCACAGAAGCTTCAAGTTAATCCACGTGATCACCAGGTTATGAAGCCAGAGGCTAAGCCAGGGCAATTTCTAGTAGTAGTTTCTTTTTGTTTTCAAAACAGGGTTTTTCTGTGCAGTCCTGGATGATCTTGGACTCAAACCTTCACCTGCCTCTGCCTCCTAAGTGCTGGATTAAAGGAGTGTGCCCCCATTGCCCAGCCCGTTTCTAAAGAGTAAGTACACCATAGAAAAGCA</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">apeh17</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1373.5</td>
<td style="text-align: right;">415</td>
<td style="text-align: left;">394a483d3ae2ce5605f1674b2400986c</td>
<td style="text-align: left;">AAAGCCAGTGGAGCCACGATAGTTCACTGCAAGTGAGACAAGATAGTCGGGCAATGTTCTCAGTCCCCAACCCAATTCCATGTGTCTGTGACATAGGCCCTAGGACCAGCTGCATCACGTGCTGTCAAGGGGAGGTAGCCAAGAGAGATGAAGCTGCTACCCTGAGACCACAGCATCCTTTCCACAGAAGCTTCAAGTTAATCCACGTGATCACCAGGTTATGAAGCCAGAGGCTAAGCCAGGGCAATTTCTAGTAGTAGTTTCTTTTTGTTTTCAAAACAGGGTTTTTCTGTGCAGTCCTGGATGATCTTGGACTCAAACCTTCACCTGCCTCTGCCTCCTAAGTGCTGGATTAAAGGAGTGTGCCCCCATTGCCCAGCCCGTTTCTAAAGAGTAAGTACACCATAGAAAAGCA</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="genotype-with-amplisat" class="level2">
<h2 class="anchored" data-anchor-id="genotype-with-amplisat">Genotype with AmpliSAT</h2>
<p>Genotyping was also carried with <a href="https://doi.org/10.1111/1755-0998.12453">AmpliSAT</a>, a software written in PERL with similar characteristics to <em>tidyGenR</em>. To compare their performance we run AmpliSAT in a DOCKER container to genotype the same raw data. The steps are detailed <a href="code/amplisas">here</a>. AmpliSAS returns results in a multisheet EXCEL and in plain text files, one per locus. The function <code>amplisas2tidy()</code> permits to read plain text results from AmpliSAS into tidy variants.</p>
<p>Create input for AmpliSAS:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># amplicon metadata</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"code/amplisas/01_create_amplicon_data.R"</span>)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="co"># append barcodes to reads</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"code/amplisas/02_add_barcodes.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>AmpliSAS is run in a DOCKER container. The steps are detailed <a href="code/amplisat_analysis.md">here</a>.</p>
<p>AmpliSAS results can be read to tidy variants:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>fp <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/intermediate/amplisas/results_amplisas/filtered"</span>,</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">pattern =</span> <span class="st">"txt$"</span>,</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="co"># amplisas results to tidy variants</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>var_amplisas <span class="ot">&lt;-</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">amplisas2tidy</span>(fp)</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a><span class="co"># to genotypes</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>gen_amplisas <span class="ot">&lt;-</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">genotype</span>(var_amplisas)</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(var_amplisas, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 6%">
<col style="width: 86%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">sample</th>
<th style="text-align: left;">locus</th>
<th style="text-align: left;">variant</th>
<th style="text-align: right;">reads</th>
<th style="text-align: right;">nt</th>
<th style="text-align: left;">md5</th>
<th style="text-align: left;">sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">abcg8</td>
<td style="text-align: left;">97168</td>
<td style="text-align: right;">467</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">c3eab42bea56937040aa79a6bbea2724</td>
<td style="text-align: left;">TTGCCCACCCTGTTCATCCATGGAGCAGAAGCCTGCCTGATGTCTCTCATCATTGGCTTCCTTTACTACGGCCACGCAGATAAGCCGCTCTCCTTCGTGGACATGGCAGCCCTCCTGTTCATGATAGGAGCGCTCATTCCTTTTAATGTCATTCTGGATGTCGTCTCCAAATGTGAGTGTCACCCGCCCTCCTCACCAGACATCGGGACAGTGGGACAGCCTCCCTGGGCACTGCACTGAGGCCAAGCTCTGTGCTTCCGCTGGTACCCACGGCATTACAAGAGATGCGACCTCAGTAACACTCTTCGCTCATTCACCTCCCTCTCCCCCATCTCCAGGTCACTCGGAGCGCTCGCTGCTGTACTATGAACTGGAGGACGGACTGTACAC</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">alkbh7</td>
<td style="text-align: left;">62348</td>
<td style="text-align: right;">185</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">620057f2e547e688d1220db91a3a1d01</td>
<td style="text-align: left;">GCCTTCCTGGCCCCATCCCCTCTGGGAGGGAGCGGCAAATCACTGAGATGCGTCGGCCCCGGGGGACCCGGTGAGCCCCAAAAAATGATTCTTCATCTCTAAGGATCTCATGGGAGAAGTCATATCGGGCTGAACCCCTGCAAGAAAATAAAGGCCAAGTAGGTGAAGGGAGGGAAGTCTGTCCCTTCATCATTTCTGTACTTTATCTGGGTAGTGGTGATGGTGACTTCCTCTGCTATGAGAGCAGAAACAAGAGCTGTTGGAAACACTTGGCCTCATCTGGGGGTTCTAGGCTAGGTCATACCTTAGGATATAGAGAGAACCAGGTTCCAGCAACAGTTCCAGCCACTGCTCAGGTTCCTGTGTATGAACCAGCTTCATAACACTTGG</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">apeh17</td>
<td style="text-align: left;">81256</td>
<td style="text-align: right;">1120</td>
<td style="text-align: right;">415</td>
<td style="text-align: left;">394a483d3ae2ce5605f1674b2400986c</td>
<td style="text-align: left;">AAAGCCAGTGGAGCCACGATAGTTCACTGCAAGTGAGACAAGATAGTCGGGCAATGTTCTCAGTCCCCAACCCAATTCCATGTGTCTGTGACATAGGCCCTAGGACCAGCTGCATCACGTGCTGTCAAGGGGAGGTAGCCAAGAGAGATGAAGCTGCTACCCTGAGACCACAGCATCCTTTCCACAGAAGCTTCAAGTTAATCCACGTGATCACCAGGTTATGAAGCCAGAGGCTAAGCCAGGGCAATTTCTAGTAGTAGTTTCTTTTTGTTTTCAAAACAGGGTTTTTCTGTGCAGTCCTGGATGATCTTGGACTCAAACCTTCACCTGCCTCTGCCTCCTAAGTGCTGGATTAAAGGAGTGTGCCCCCATTGCCCAGCCCGTTTCTAAAGAGTAAGTACACCATAGAAAAGCA</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="compare-results" class="level2">
<h2 class="anchored" data-anchor-id="compare-results">Compare results</h2>
<p><code>compare_calls()</code> was used for comparing results of <em>tidyGenR</em> run with different sets of parameters between them and agains AmpliSAT results. The comparison indicates that the set of parameters that returns the most similar genotype calls to AmpliSAT is using a band size = 0.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compare genotypes of the four runs plus amplisat</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   genotypes</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>gen_comb <span class="ot">&lt;-</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(genotypes_x,</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">list</span>(gen_amplisas))</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(gen_comb) <span class="ot">&lt;-</span> <span class="fu">c</span>(stringr<span class="sc">::</span><span class="fu">str_c</span>(<span class="st">"tidyGenR_"</span>, <span class="fu">names</span>(genotypes_x)), <span class="st">"AmpliSAT"</span>)</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>comp_gen <span class="ot">&lt;-</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">suppressMessages</span>(<span class="fu">compare_calls</span>(gen_comb,</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"output/comp_gen_noReads.xlsx"</span>,</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>                <span class="at">creads =</span> <span class="cn">FALSE</span>))</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(comp_gen<span class="sc">$</span>dist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">method1</th>
<th style="text-align: left;">method2</th>
<th style="text-align: right;">dist_euc</th>
<th style="text-align: right;">dist_eucp</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">tidyGenR_ind_bs16</td>
<td style="text-align: left;">tidyGenR_ind_bs0</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">0.0205420</td>
</tr>
<tr class="even">
<td style="text-align: left;">tidyGenR_ind_bs16</td>
<td style="text-align: left;">tidyGenR_pool_bs16</td>
<td style="text-align: right;">42</td>
<td style="text-align: right;">0.0183566</td>
</tr>
<tr class="odd">
<td style="text-align: left;">tidyGenR_ind_bs16</td>
<td style="text-align: left;">tidyGenR_pool_bs0</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">0.0205420</td>
</tr>
<tr class="even">
<td style="text-align: left;">tidyGenR_ind_bs16</td>
<td style="text-align: left;">AmpliSAT</td>
<td style="text-align: right;">74</td>
<td style="text-align: right;">0.0323427</td>
</tr>
<tr class="odd">
<td style="text-align: left;">tidyGenR_ind_bs0</td>
<td style="text-align: left;">tidyGenR_pool_bs16</td>
<td style="text-align: right;">51</td>
<td style="text-align: right;">0.0222902</td>
</tr>
<tr class="even">
<td style="text-align: left;">tidyGenR_ind_bs0</td>
<td style="text-align: left;">tidyGenR_pool_bs0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="odd">
<td style="text-align: left;">tidyGenR_ind_bs0</td>
<td style="text-align: left;">AmpliSAT</td>
<td style="text-align: right;">29</td>
<td style="text-align: right;">0.0126748</td>
</tr>
<tr class="even">
<td style="text-align: left;">tidyGenR_pool_bs16</td>
<td style="text-align: left;">tidyGenR_pool_bs0</td>
<td style="text-align: right;">51</td>
<td style="text-align: right;">0.0222902</td>
</tr>
<tr class="odd">
<td style="text-align: left;">tidyGenR_pool_bs16</td>
<td style="text-align: left;">AmpliSAT</td>
<td style="text-align: right;">78</td>
<td style="text-align: right;">0.0340909</td>
</tr>
<tr class="even">
<td style="text-align: left;">tidyGenR_pool_bs0</td>
<td style="text-align: left;">AmpliSAT</td>
<td style="text-align: right;">29</td>
<td style="text-align: right;">0.0126748</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>A plot produced by <code>compare_calls()</code> depicts the differences between strategies. It is useful for spotting potential biases and problematic loci (<a href="#fig-comp_gen_all">Figure&nbsp;3</a>).</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>comp_gen<span class="sc">$</span>plot2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-comp_gen_all" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="benchmarking_files/figure-html/fig-comp_gen_all-1.png" class="img-fluid figure-img" width="1152"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;3: Comparison between genotype calls.</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>Problematic loci, with low coverage and with conflicting genotypes between different strategies were checked manually by aligning them and compared the with dereplicated reads. Setting band_size = 0 yielded the best results as it recovered alleles which had indels at their ends. Pooling samples did not affect genotype calls, although it has some influence on the number of reads supporting each allele. Thus samples not sharing alleles (i.e.&nbsp;different species) can be genotyped equally efficiently as samples sharing alleles (i.e.&nbsp;same species). Then, the run with <em>band size = 0</em> and <em>non-pooled</em> samples was chosen as the best strategy and compared pairwise with AmpliSAT.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compare best call with amplisas</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   variants</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>comp_gen_sel <span class="ot">&lt;-</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">compare_calls</span>(<span class="fu">list</span>(<span class="at">tidyGenR_ind_bs0 =</span> genotypes_x<span class="sc">$</span>ind_bs0,</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">AmpliSAT =</span> gen_amplisas),</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"output/comp_gen_indbs0_amplisat.xlsx"</span>,</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>                  <span class="at">creads =</span> <span class="cn">TRUE</span>)</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(comp_gen_sel, <span class="st">"output/comp_gen_sel.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>pcomp1 <span class="ot">&lt;-</span> comp_gen_sel<span class="sc">$</span>plot2</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(pcomp1,</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">file =</span> <span class="st">"output/comp_amplisat_indo.pdf"</span>,</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">width =</span> <span class="dv">10</span>,</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">height =</span> <span class="fl">5.4</span>)</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>pcomp1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-compgen-amplisas" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="benchmarking_files/figure-html/fig-compgen-amplisas-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;4: Comparison of genotype call tidyGenR vs AmpliSAT</figcaption><p></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>pcomp2 <span class="ot">&lt;-</span> comp_gen_sel<span class="sc">$</span>plot3</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(pcomp2,</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">file =</span> <span class="st">"output/comp_amplisat_ind0_boxplot.pdf"</span>,</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">width =</span> <span class="dv">5</span>,</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">height =</span> <span class="dv">3</span>)</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>pcomp2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-compgen-amplisas_nreads" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="benchmarking_files/figure-html/fig-compgen-amplisas_nreads-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;5: Comparison of number of reads supporting alleles in tidyGenR vs AmpliSAT.</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>Lastly, the genotypes can be compared against their respective variants to have an idea of variants being dropped when genotyping:</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>dropped_var <span class="ot">&lt;-</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>      <span class="fu">list</span>(<span class="at">variants =</span> variants_x<span class="sc">$</span>ind_bs0,</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">genotypes =</span> genotypes_x<span class="sc">$</span>ind_bs0)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>comp_var_gen <span class="ot">&lt;-</span> <span class="fu">compare_calls</span>(dropped_var)</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(comp_var_gen, <span class="st">"output/comp_var_gen.rds"</span>)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>comp_var_gen<span class="sc">$</span>plot1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-dropped_var" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="benchmarking_files/figure-html/fig-dropped_var-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;6: Identified variants dropped during genotyping.</figcaption><p></p>
</figure>
</div>
</div>
</div>
</section>
<section id="session-info" class="level1">
<h1>Session Info</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.0 (2024-04-24)
Platform: x86_64-apple-darwin20
Running under: macOS 15.3.2

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Europe/Madrid
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_3.5.1   patchwork_1.3.0 dplyr_1.1.4     tidyGenR_0.8.3 

loaded via a namespace (and not attached):
 [1] writexl_1.5.1               tidyselect_1.2.1           
 [3] farver_2.1.2                Biostrings_2.72.1          
 [5] bitops_1.0-9                fastmap_1.2.0              
 [7] dada2_1.32.0                GenomicAlignments_1.40.0   
 [9] digest_0.6.37               lifecycle_1.0.4            
[11] pwalign_1.0.0               magrittr_2.0.3             
[13] compiler_4.4.0              rlang_1.1.5                
[15] tools_4.4.0                 yaml_2.3.10                
[17] knitr_1.49                  labeling_0.4.3             
[19] S4Arrays_1.4.1              htmlwidgets_1.6.4          
[21] interp_1.1-6                DelayedArray_0.30.1        
[23] plyr_1.8.9                  RColorBrewer_1.1-3         
[25] abind_1.4-8                 ShortRead_1.62.0           
[27] BiocParallel_1.38.0         withr_3.0.2                
[29] purrr_1.0.4                 hwriter_1.3.2.1            
[31] BiocGenerics_0.50.0         grid_4.4.0                 
[33] stats4_4.4.0                latticeExtra_0.6-30        
[35] colorspace_2.1-1            scales_1.3.0               
[37] SummarizedExperiment_1.34.0 cli_3.6.4                  
[39] rmarkdown_2.29              crayon_1.5.3               
[41] ragg_1.3.3                  generics_0.1.3             
[43] RcppParallel_5.1.10         rstudioapi_0.17.1          
[45] httr_1.4.7                  reshape2_1.4.4             
[47] tzdb_0.4.0                  DBI_1.2.3                  
[49] stringr_1.5.1               zlibbioc_1.50.0            
[51] parallel_4.4.0              XVector_0.44.0             
[53] matrixStats_1.5.0           vctrs_0.6.5                
[55] Matrix_1.7-2                jsonlite_1.8.9             
[57] IRanges_2.38.1              hms_1.1.3                  
[59] S4Vectors_0.42.1            systemfonts_1.2.1          
[61] jpeg_0.1-10                 tidyr_1.3.1                
[63] glue_1.8.0                  codetools_0.2-20           
[65] stringi_1.8.4               gtable_0.3.6               
[67] GenomeInfoDb_1.40.1         deldir_2.0-4               
[69] GenomicRanges_1.56.2        UCSC.utils_1.0.0           
[71] munsell_0.5.1               tibble_3.2.1               
[73] pillar_1.10.1               htmltools_0.5.8.1          
[75] GenomeInfoDbData_1.2.12     R6_2.6.1                   
[77] textshaping_1.0.0           evaluate_1.0.3             
[79] lattice_0.22-6              Biobase_2.64.0             
[81] readr_2.1.5                 png_0.1-8                  
[83] Rsamtools_2.20.0            DECIPHER_3.0.0             
[85] Rcpp_1.0.14                 SparseArray_1.4.8          
[87] xfun_0.50                   MatrixGenerics_1.16.0      
[89] pkgconfig_2.0.3            </code></pre>
</div>
</div>
<!-- -->

</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb44" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> Miguel Camacho Sánchez</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "tidyGenR benchmarking"</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span><span class="co"> visual</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="an">theme:</span><span class="co"> minty</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>Attach libraries:</span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-17"><a href="#cb44-17" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb44-18"><a href="#cb44-18" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb44-19"><a href="#cb44-19" aria-hidden="true" tabindex="-1"></a><span class="co">#knitr::opts_knit$set(root.dir = "../")</span></span>
<span id="cb44-20"><a href="#cb44-20" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-21"><a href="#cb44-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-22"><a href="#cb44-22" aria-hidden="true" tabindex="-1"></a><span class="in">```{r attach packages}</span></span>
<span id="cb44-23"><a href="#cb44-23" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb44-24"><a href="#cb44-24" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb44-25"><a href="#cb44-25" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyGenR)</span>
<span id="cb44-26"><a href="#cb44-26" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb44-27"><a href="#cb44-27" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb44-28"><a href="#cb44-28" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb44-29"><a href="#cb44-29" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-30"><a href="#cb44-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-31"><a href="#cb44-31" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eval_conditionals}</span></span>
<span id="cb44-32"><a href="#cb44-32" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb44-33"><a href="#cb44-33" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb44-34"><a href="#cb44-34" aria-hidden="true" tabindex="-1"></a><span class="co"># declare logicals to trigger EVAL in intensive code chunks</span></span>
<span id="cb44-35"><a href="#cb44-35" aria-hidden="true" tabindex="-1"></a><span class="co"># they are not run if output already exists.</span></span>
<span id="cb44-36"><a href="#cb44-36" aria-hidden="true" tabindex="-1"></a><span class="co"># to run fresh analysis outputs need to be removed.</span></span>
<span id="cb44-37"><a href="#cb44-37" aria-hidden="true" tabindex="-1"></a><span class="co"># demultiplex</span></span>
<span id="cb44-38"><a href="#cb44-38" aria-hidden="true" tabindex="-1"></a>c_log <span class="ot">&lt;-</span> <span class="st">"output/cutadapt.log"</span></span>
<span id="cb44-39"><a href="#cb44-39" aria-hidden="true" tabindex="-1"></a>eval_demultiplex <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(c_log)</span>
<span id="cb44-40"><a href="#cb44-40" aria-hidden="true" tabindex="-1"></a><span class="co"># fastqc</span></span>
<span id="cb44-41"><a href="#cb44-41" aria-hidden="true" tabindex="-1"></a>multiqc_report <span class="ot">&lt;-</span> <span class="st">"output/multiqc_report.html"</span></span>
<span id="cb44-42"><a href="#cb44-42" aria-hidden="true" tabindex="-1"></a>eval_fastqc <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(multiqc_report)</span>
<span id="cb44-43"><a href="#cb44-43" aria-hidden="true" tabindex="-1"></a><span class="co"># truncate</span></span>
<span id="cb44-44"><a href="#cb44-44" aria-hidden="true" tabindex="-1"></a>tr_path <span class="ot">&lt;-</span> <span class="st">"output/trunc_in-out.rds"</span></span>
<span id="cb44-45"><a href="#cb44-45" aria-hidden="true" tabindex="-1"></a>eval_truncate <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(tr_path)</span>
<span id="cb44-46"><a href="#cb44-46" aria-hidden="true" tabindex="-1"></a><span class="co"># explore_dada</span></span>
<span id="cb44-47"><a href="#cb44-47" aria-hidden="true" tabindex="-1"></a>x_out_path <span class="ot">&lt;-</span> <span class="st">"output/explore_dada.rds"</span></span>
<span id="cb44-48"><a href="#cb44-48" aria-hidden="true" tabindex="-1"></a>eval_x <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(x_out_path)</span>
<span id="cb44-49"><a href="#cb44-49" aria-hidden="true" tabindex="-1"></a><span class="co"># variant_call</span></span>
<span id="cb44-50"><a href="#cb44-50" aria-hidden="true" tabindex="-1"></a>x_variants_path <span class="ot">&lt;-</span>  <span class="st">"output/variants_x.rds"</span></span>
<span id="cb44-51"><a href="#cb44-51" aria-hidden="true" tabindex="-1"></a>eval_variant_call <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(x_variants_path)</span>
<span id="cb44-52"><a href="#cb44-52" aria-hidden="true" tabindex="-1"></a><span class="co"># genotype</span></span>
<span id="cb44-53"><a href="#cb44-53" aria-hidden="true" tabindex="-1"></a>x_genotypes_path <span class="ot">&lt;-</span>  <span class="st">"output/genotypes_x.rds"</span></span>
<span id="cb44-54"><a href="#cb44-54" aria-hidden="true" tabindex="-1"></a>eval_genotype <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(x_genotypes_path)</span>
<span id="cb44-55"><a href="#cb44-55" aria-hidden="true" tabindex="-1"></a><span class="co"># amplisas</span></span>
<span id="cb44-56"><a href="#cb44-56" aria-hidden="true" tabindex="-1"></a>amplisas_results <span class="ot">&lt;-</span> <span class="st">""</span></span>
<span id="cb44-57"><a href="#cb44-57" aria-hidden="true" tabindex="-1"></a>eval_amplisas <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(amplisas_results)</span>
<span id="cb44-58"><a href="#cb44-58" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-59"><a href="#cb44-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-60"><a href="#cb44-60" aria-hidden="true" tabindex="-1"></a><span class="fu">## Benchmarking *tidyGenR*</span></span>
<span id="cb44-61"><a href="#cb44-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-62"><a href="#cb44-62" aria-hidden="true" tabindex="-1"></a>This repository accompanies the R package **tidyGenR** available at <span class="ot">&lt;https://github.com/csmiguel/tidyGenR&gt;</span>. It covers (1) all steps for genotype calling with *tidyGenR*, (2) exploration of the parameters for variant calling, and (3) comparison of *tidyGenR* against [AmpliSAS](https://doi.org/10.1111/1755-0998.12453). It uses real data from a population genetics study of the wild rodent *Rattus baluensis* (Camacho-Sanchez et al. *in preparation*).</span>
<span id="cb44-63"><a href="#cb44-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-64"><a href="#cb44-64" aria-hidden="true" tabindex="-1"></a><span class="fu">## Preparation of input data</span></span>
<span id="cb44-65"><a href="#cb44-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-66"><a href="#cb44-66" aria-hidden="true" tabindex="-1"></a>Raw sequences are deposited in NCBI under the BioStudy <span class="co">[</span><span class="ot">SRP293699</span><span class="co">](https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP293699)</span>, in the BioProject <span class="co">[</span><span class="ot">PRJNA680166</span><span class="co">](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA680166)</span>. From the BioStudy, we can download the <span class="co">[</span><span class="ot">SRA metadata</span><span class="co">](data/raw/SraRunTable.csv)</span> to mapping later sample IDs to SRRs.</span>
<span id="cb44-67"><a href="#cb44-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-68"><a href="#cb44-68" aria-hidden="true" tabindex="-1"></a><span class="in">```{r SRR}</span></span>
<span id="cb44-69"><a href="#cb44-69" aria-hidden="true" tabindex="-1"></a><span class="co"># table with mapped SRR-sampleIDs</span></span>
<span id="cb44-70"><a href="#cb44-70" aria-hidden="true" tabindex="-1"></a>srr <span class="ot">&lt;-</span></span>
<span id="cb44-71"><a href="#cb44-71" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read.csv</span>(<span class="st">"data/raw/SraRunTable.csv"</span>) <span class="sc">|&gt;</span></span>
<span id="cb44-72"><a href="#cb44-72" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Run, <span class="st">"Sample.Name"</span>) <span class="sc">|&gt;</span></span>
<span id="cb44-73"><a href="#cb44-73" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">run =</span> Run,</span>
<span id="cb44-74"><a href="#cb44-74" aria-hidden="true" tabindex="-1"></a>         <span class="at">sample =</span> <span class="st">"Sample.Name"</span>)</span>
<span id="cb44-75"><a href="#cb44-75" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(srr, <span class="dv">3</span>))</span>
<span id="cb44-76"><a href="#cb44-76" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-77"><a href="#cb44-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-78"><a href="#cb44-78" aria-hidden="true" tabindex="-1"></a>Raw FASTQ reads from the SRRs can be downloaded with <span class="co">[</span><span class="ot">SRA-toolkit</span><span class="co">](https://github.com/ncbi/sra-tools)</span>. Then, they are renamed based on the <span class="in">`data/raw/SraRunTable.csv`</span> mapping file.</span>
<span id="cb44-79"><a href="#cb44-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-80"><a href="#cb44-80" aria-hidden="true" tabindex="-1"></a><span class="in">```{r download_raw_data}</span></span>
<span id="cb44-81"><a href="#cb44-81" aria-hidden="true" tabindex="-1"></a><span class="co">#|  message: false</span></span>
<span id="cb44-82"><a href="#cb44-82" aria-hidden="true" tabindex="-1"></a><span class="co">#|  warning: false</span></span>
<span id="cb44-83"><a href="#cb44-83" aria-hidden="true" tabindex="-1"></a><span class="co">#|  error: false</span></span>
<span id="cb44-84"><a href="#cb44-84" aria-hidden="true" tabindex="-1"></a><span class="co">#|  code-fold: true</span></span>
<span id="cb44-85"><a href="#cb44-85" aria-hidden="true" tabindex="-1"></a><span class="co">#|  results: hide</span></span>
<span id="cb44-86"><a href="#cb44-86" aria-hidden="true" tabindex="-1"></a><span class="co"># download raw reads from NCBI</span></span>
<span id="cb44-87"><a href="#cb44-87" aria-hidden="true" tabindex="-1"></a><span class="co"># they are not downloaded if already present in 'data/raw'.</span></span>
<span id="cb44-88"><a href="#cb44-88" aria-hidden="true" tabindex="-1"></a><span class="fu">invisible</span>(</span>
<span id="cb44-89"><a href="#cb44-89" aria-hidden="true" tabindex="-1"></a>  <span class="fu">apply</span>(srr, <span class="dv">1</span>, <span class="cf">function</span>(x) {</span>
<span id="cb44-90"><a href="#cb44-90" aria-hidden="true" tabindex="-1"></a>  <span class="co"># name paths</span></span>
<span id="cb44-91"><a href="#cb44-91" aria-hidden="true" tabindex="-1"></a>  srr_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, x[<span class="dv">1</span>])</span>
<span id="cb44-92"><a href="#cb44-92" aria-hidden="true" tabindex="-1"></a>  srr1 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">1</span>], <span class="st">"_1.fastq"</span>))</span>
<span id="cb44-93"><a href="#cb44-93" aria-hidden="true" tabindex="-1"></a>  srr2 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">1</span>], <span class="st">"_2.fastq"</span>))</span>
<span id="cb44-94"><a href="#cb44-94" aria-hidden="true" tabindex="-1"></a>  sam1 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">2</span>], <span class="st">"_1.fastq"</span>))</span>
<span id="cb44-95"><a href="#cb44-95" aria-hidden="true" tabindex="-1"></a>  sam2 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">2</span>], <span class="st">"_2.fastq"</span>))</span>
<span id="cb44-96"><a href="#cb44-96" aria-hidden="true" tabindex="-1"></a>  <span class="co"># fetch SRR</span></span>
<span id="cb44-97"><a href="#cb44-97" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">any</span>(<span class="fu">dir.exists</span>(srr_path) <span class="sc">||</span> <span class="fu">file.exists</span>(sam1))) {</span>
<span id="cb44-98"><a href="#cb44-98" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"prefetch"</span>, <span class="fu">paste</span>(<span class="st">"-O"</span>, <span class="st">"data/raw/"</span>, x[<span class="dv">1</span>]))</span>
<span id="cb44-99"><a href="#cb44-99" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb44-100"><a href="#cb44-100" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(sam1)) {</span>
<span id="cb44-101"><a href="#cb44-101" aria-hidden="true" tabindex="-1"></a>  <span class="co"># reformat to FASTQ</span></span>
<span id="cb44-102"><a href="#cb44-102" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"fasterq-dump"</span>, <span class="fu">paste</span>(<span class="st">"-O"</span>, <span class="st">"data/raw/"</span>, srr_path))</span>
<span id="cb44-103"><a href="#cb44-103" aria-hidden="true" tabindex="-1"></a>  <span class="co"># rename SRR with sample codes</span></span>
<span id="cb44-104"><a href="#cb44-104" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"mv"</span>, <span class="fu">paste</span>(srr1, sam1))</span>
<span id="cb44-105"><a href="#cb44-105" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"mv"</span>, <span class="fu">paste</span>(srr2, sam2))</span>
<span id="cb44-106"><a href="#cb44-106" aria-hidden="true" tabindex="-1"></a>    <span class="co"># rm SRR</span></span>
<span id="cb44-107"><a href="#cb44-107" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"rm"</span>, <span class="fu">paste0</span>(<span class="st">"-r "</span>, srr_path, <span class="st">"*"</span>))</span>
<span id="cb44-108"><a href="#cb44-108" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb44-109"><a href="#cb44-109" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb44-110"><a href="#cb44-110" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb44-111"><a href="#cb44-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-112"><a href="#cb44-112" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-113"><a href="#cb44-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-114"><a href="#cb44-114" aria-hidden="true" tabindex="-1"></a>List some of the downloaded files:</span>
<span id="cb44-115"><a href="#cb44-115" aria-hidden="true" tabindex="-1"></a><span class="in">```{r glimpse fastq}</span></span>
<span id="cb44-116"><a href="#cb44-116" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(<span class="st">"data/raw"</span>, <span class="at">pattern =</span> <span class="st">"fastq"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb44-117"><a href="#cb44-117" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-118"><a href="#cb44-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-119"><a href="#cb44-119" aria-hidden="true" tabindex="-1"></a>Check input raw fastq:</span>
<span id="cb44-120"><a href="#cb44-120" aria-hidden="true" tabindex="-1"></a><span class="in">```{r check_raw_reads}</span></span>
<span id="cb44-121"><a href="#cb44-121" aria-hidden="true" tabindex="-1"></a>freads <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/raw"</span>, <span class="at">pattern =</span> <span class="st">"1.fastq"</span>,</span>
<span id="cb44-122"><a href="#cb44-122" aria-hidden="true" tabindex="-1"></a>                    <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb44-123"><a href="#cb44-123" aria-hidden="true" tabindex="-1"></a>rreads <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/raw"</span>, <span class="at">pattern =</span> <span class="st">"2.fastq"</span>,</span>
<span id="cb44-124"><a href="#cb44-124" aria-hidden="true" tabindex="-1"></a>                    <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb44-125"><a href="#cb44-125" aria-hidden="true" tabindex="-1"></a>chr <span class="ot">&lt;-</span> <span class="fu">check_raw_reads</span>(freads, rreads, <span class="at">low_readcount =</span> <span class="dv">10</span>)</span>
<span id="cb44-126"><a href="#cb44-126" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-127"><a href="#cb44-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-128"><a href="#cb44-128" aria-hidden="true" tabindex="-1"></a>Input FASTQ complies with expected input for *tidyGenR*. A total of <span class="in">`r length(chr$samples)`</span> samples are detected:</span>
<span id="cb44-129"><a href="#cb44-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-132"><a href="#cb44-132" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-133"><a href="#cb44-133" aria-hidden="true" tabindex="-1"></a>chr<span class="sc">$</span>samples</span>
<span id="cb44-134"><a href="#cb44-134" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-135"><a href="#cb44-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-136"><a href="#cb44-136" aria-hidden="true" tabindex="-1"></a><span class="fu">## Demultiplex by locus</span></span>
<span id="cb44-137"><a href="#cb44-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-138"><a href="#cb44-138" aria-hidden="true" tabindex="-1"></a>Reads are demultiplexed by locus using primer sequences in paired-end mode.</span>
<span id="cb44-139"><a href="#cb44-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-140"><a href="#cb44-140" aria-hidden="true" tabindex="-1"></a><span class="in">```{r demultiplex_objects}</span></span>
<span id="cb44-141"><a href="#cb44-141" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb44-142"><a href="#cb44-142" aria-hidden="true" tabindex="-1"></a><span class="co"># load primer data</span></span>
<span id="cb44-143"><a href="#cb44-143" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"primers"</span>)</span>
<span id="cb44-144"><a href="#cb44-144" aria-hidden="true" tabindex="-1"></a><span class="co"># path to cutadapt</span></span>
<span id="cb44-145"><a href="#cb44-145" aria-hidden="true" tabindex="-1"></a>cutadapt <span class="ot">&lt;-</span> <span class="fu">system</span>(<span class="st">"which cutadapt"</span>, <span class="at">intern =</span> <span class="cn">TRUE</span>)</span>
<span id="cb44-146"><a href="#cb44-146" aria-hidden="true" tabindex="-1"></a><span class="co"># path to folder save locus-demultiplexed FASTQ</span></span>
<span id="cb44-147"><a href="#cb44-147" aria-hidden="true" tabindex="-1"></a>demult <span class="ot">&lt;-</span> <span class="st">"data/intermediate/demultiplexed"</span></span>
<span id="cb44-148"><a href="#cb44-148" aria-hidden="true" tabindex="-1"></a><span class="co"># print primers</span></span>
<span id="cb44-149"><a href="#cb44-149" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(primers, <span class="dv">3</span>))</span>
<span id="cb44-150"><a href="#cb44-150" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-151"><a href="#cb44-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-152"><a href="#cb44-152" aria-hidden="true" tabindex="-1"></a><span class="in">```{r demultiplex}</span></span>
<span id="cb44-153"><a href="#cb44-153" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr eval_demultiplex</span></span>
<span id="cb44-154"><a href="#cb44-154" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb44-155"><a href="#cb44-155" aria-hidden="true" tabindex="-1"></a><span class="co"># demultiplex</span></span>
<span id="cb44-156"><a href="#cb44-156" aria-hidden="true" tabindex="-1"></a><span class="fu">demultiplex</span>(</span>
<span id="cb44-157"><a href="#cb44-157" aria-hidden="true" tabindex="-1"></a>  <span class="at">interpreter =</span> <span class="st">"/bin/bash"</span>,</span>
<span id="cb44-158"><a href="#cb44-158" aria-hidden="true" tabindex="-1"></a>  <span class="at">cutadapt =</span> cutadapt,</span>
<span id="cb44-159"><a href="#cb44-159" aria-hidden="true" tabindex="-1"></a>  <span class="at">freads =</span> freads,</span>
<span id="cb44-160"><a href="#cb44-160" aria-hidden="true" tabindex="-1"></a>  <span class="at">rreads =</span> rreads,</span>
<span id="cb44-161"><a href="#cb44-161" aria-hidden="true" tabindex="-1"></a>  <span class="at">primers =</span> primers,</span>
<span id="cb44-162"><a href="#cb44-162" aria-hidden="true" tabindex="-1"></a>  <span class="at">sh_out =</span> <span class="st">"code/demultiplex.sh"</span>,</span>
<span id="cb44-163"><a href="#cb44-163" aria-hidden="true" tabindex="-1"></a>  <span class="at">write_demultiplexed =</span> demult,</span>
<span id="cb44-164"><a href="#cb44-164" aria-hidden="true" tabindex="-1"></a>  <span class="at">log_out =</span> c_log,</span>
<span id="cb44-165"><a href="#cb44-165" aria-hidden="true" tabindex="-1"></a>  <span class="at">mode =</span> <span class="st">"pe"</span>,</span>
<span id="cb44-166"><a href="#cb44-166" aria-hidden="true" tabindex="-1"></a>  <span class="at">run =</span> <span class="cn">TRUE</span>)</span>
<span id="cb44-167"><a href="#cb44-167" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-168"><a href="#cb44-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-169"><a href="#cb44-169" aria-hidden="true" tabindex="-1"></a>Glimpse demultiplexed FASTQ:</span>
<span id="cb44-172"><a href="#cb44-172" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-173"><a href="#cb44-173" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(demult, <span class="at">pattern =</span> <span class="st">"fastq"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb44-174"><a href="#cb44-174" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-175"><a href="#cb44-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-176"><a href="#cb44-176" aria-hidden="true" tabindex="-1"></a>Remove files with few reads:</span>
<span id="cb44-177"><a href="#cb44-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-178"><a href="#cb44-178" aria-hidden="true" tabindex="-1"></a><span class="in">```{r rm_after_dem}</span></span>
<span id="cb44-179"><a href="#cb44-179" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb44-180"><a href="#cb44-180" aria-hidden="true" tabindex="-1"></a><span class="fu">remove_poor_fastq</span>(demult,</span>
<span id="cb44-181"><a href="#cb44-181" aria-hidden="true" tabindex="-1"></a>                   <span class="at">min_reads =</span> <span class="dv">10</span>)</span>
<span id="cb44-182"><a href="#cb44-182" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-183"><a href="#cb44-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-184"><a href="#cb44-184" aria-hidden="true" tabindex="-1"></a>Make sequence-quality reports using FastQC and MultiQC. Instead of running it over the <span class="sc">\&gt;</span>2000 demultiplexed files, it will be run over 100 random files.</span>
<span id="cb44-185"><a href="#cb44-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-188"><a href="#cb44-188" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-189"><a href="#cb44-189" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr eval_fastqc</span></span>
<span id="cb44-190"><a href="#cb44-190" aria-hidden="true" tabindex="-1"></a><span class="co"># instead of running fastqc to all files &gt; 2000, I run</span></span>
<span id="cb44-191"><a href="#cb44-191" aria-hidden="true" tabindex="-1"></a><span class="co">#    it on 100 random files.</span></span>
<span id="cb44-192"><a href="#cb44-192" aria-hidden="true" tabindex="-1"></a>dem_files <span class="ot">&lt;-</span></span>
<span id="cb44-193"><a href="#cb44-193" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>(demult, <span class="st">"fastq.gz"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb44-194"><a href="#cb44-194" aria-hidden="true" tabindex="-1"></a><span class="co"># select 100 random files</span></span>
<span id="cb44-195"><a href="#cb44-195" aria-hidden="true" tabindex="-1"></a>random_dem <span class="ot">&lt;-</span></span>
<span id="cb44-196"><a href="#cb44-196" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample</span>(dem_files, <span class="dv">100</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb44-197"><a href="#cb44-197" aria-hidden="true" tabindex="-1"></a><span class="co"># copy to temp dir</span></span>
<span id="cb44-198"><a href="#cb44-198" aria-hidden="true" tabindex="-1"></a>temp_dem <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">tempdir</span>(), <span class="st">"dem"</span>)</span>
<span id="cb44-199"><a href="#cb44-199" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(temp_dem)</span>
<span id="cb44-200"><a href="#cb44-200" aria-hidden="true" tabindex="-1"></a><span class="fu">file.copy</span>(<span class="at">from =</span> random_dem, <span class="at">to =</span> temp_dem)</span>
<span id="cb44-201"><a href="#cb44-201" aria-hidden="true" tabindex="-1"></a><span class="co"># run fastqc</span></span>
<span id="cb44-202"><a href="#cb44-202" aria-hidden="true" tabindex="-1"></a><span class="fu">system2</span>(<span class="st">"fastqc"</span>,</span>
<span id="cb44-203"><a href="#cb44-203" aria-hidden="true" tabindex="-1"></a>        <span class="fu">paste</span>(<span class="st">"--noextract -o"</span>,</span>
<span id="cb44-204"><a href="#cb44-204" aria-hidden="true" tabindex="-1"></a>              temp_dem,</span>
<span id="cb44-205"><a href="#cb44-205" aria-hidden="true" tabindex="-1"></a>              <span class="fu">paste0</span>(temp_dem, <span class="st">"/*fastq.gz"</span>)))</span>
<span id="cb44-206"><a href="#cb44-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-207"><a href="#cb44-207" aria-hidden="true" tabindex="-1"></a><span class="co"># run multiqc</span></span>
<span id="cb44-208"><a href="#cb44-208" aria-hidden="true" tabindex="-1"></a><span class="fu">system2</span>(<span class="st">"multiqc"</span>,</span>
<span id="cb44-209"><a href="#cb44-209" aria-hidden="true" tabindex="-1"></a>        <span class="fu">paste</span>(<span class="st">"-o output"</span>, temp_dem))</span>
<span id="cb44-210"><a href="#cb44-210" aria-hidden="true" tabindex="-1"></a>           </span>
<span id="cb44-211"><a href="#cb44-211" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-212"><a href="#cb44-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-213"><a href="#cb44-213" aria-hidden="true" tabindex="-1"></a><span class="fu">## Truncate reads</span></span>
<span id="cb44-214"><a href="#cb44-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-215"><a href="#cb44-215" aria-hidden="true" tabindex="-1"></a>Reads are truncated to a given length for each locus. The truncation lengths depend on the sequence qualities for forward and reverse reads. After per-base sequence-qualities in the <span class="co">[</span><span class="ot">output/multiqc_report.html</span><span class="co">](output/multiqc_report.html)</span>, **270** nt for forward reads and **180** for reverse reads seem reasonable truncation lengths. For some loci, specific truncation lengths were set to maximize the number of reads returned. That is, a <span class="in">`data.frame`</span> with truncation lengths for forward and reverse reads was built to maximize the amount of information yielded by each locus. For instance, the amplicons for some loci, as *nfkbia*, are long and it is a good trade-off to keep the low quality ends but to be sure both F and R reads overlap.</span>
<span id="cb44-216"><a href="#cb44-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-217"><a href="#cb44-217" aria-hidden="true" tabindex="-1"></a>A <span class="in">`data.frame`</span> with locus-specific truncation lengths was built.</span>
<span id="cb44-218"><a href="#cb44-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-219"><a href="#cb44-219" aria-hidden="true" tabindex="-1"></a><span class="in">```{r build_trunc_fr}</span></span>
<span id="cb44-220"><a href="#cb44-220" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb44-221"><a href="#cb44-221" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: true</span></span>
<span id="cb44-222"><a href="#cb44-222" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb44-223"><a href="#cb44-223" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb44-224"><a href="#cb44-224" aria-hidden="true" tabindex="-1"></a>loci <span class="ot">&lt;-</span> </span>
<span id="cb44-225"><a href="#cb44-225" aria-hidden="true" tabindex="-1"></a>  tidyGenR<span class="sc">:::</span><span class="fu">check_names_demultiplexed</span>(demult,</span>
<span id="cb44-226"><a href="#cb44-226" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">fw_pattern =</span> <span class="st">"1.fastq.gz"</span>,</span>
<span id="cb44-227"><a href="#cb44-227" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">rv_pattern =</span> <span class="st">"2.fastq.gz"</span>)<span class="sc">$</span>loci</span>
<span id="cb44-228"><a href="#cb44-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-229"><a href="#cb44-229" aria-hidden="true" tabindex="-1"></a>trunc_fr <span class="ot">&lt;-</span></span>
<span id="cb44-230"><a href="#cb44-230" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(<span class="at">locus =</span> loci,</span>
<span id="cb44-231"><a href="#cb44-231" aria-hidden="true" tabindex="-1"></a>             <span class="at">trunc_f =</span> <span class="dv">270</span>,</span>
<span id="cb44-232"><a href="#cb44-232" aria-hidden="true" tabindex="-1"></a>             <span class="at">trunc_r =</span> <span class="dv">180</span>)</span>
<span id="cb44-233"><a href="#cb44-233" aria-hidden="true" tabindex="-1"></a><span class="co"># introduce manual values</span></span>
<span id="cb44-234"><a href="#cb44-234" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"rgd735029"</span>), <span class="st">"trunc_f"</span>] <span class="ot">&lt;-</span>  <span class="dv">245</span></span>
<span id="cb44-235"><a href="#cb44-235" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"rgd735029"</span>), <span class="st">"trunc_r"</span>] <span class="ot">&lt;-</span>  <span class="dv">155</span></span>
<span id="cb44-236"><a href="#cb44-236" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"fancg"</span>), <span class="st">"trunc_r"</span>] <span class="ot">&lt;-</span> <span class="dv">190</span></span>
<span id="cb44-237"><a href="#cb44-237" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"nfkbia"</span>), <span class="st">"trunc_r"</span>] <span class="ot">&lt;-</span> <span class="dv">240</span></span>
<span id="cb44-238"><a href="#cb44-238" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"tmem87a"</span>), <span class="st">"trunc_r"</span>] <span class="ot">&lt;-</span> <span class="dv">160</span></span>
<span id="cb44-239"><a href="#cb44-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-240"><a href="#cb44-240" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse data.frame with locus-specific truncation lengths</span></span>
<span id="cb44-241"><a href="#cb44-241" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(trunc_fr, <span class="dv">3</span>))</span>
<span id="cb44-242"><a href="#cb44-242" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-243"><a href="#cb44-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-244"><a href="#cb44-244" aria-hidden="true" tabindex="-1"></a>Truncate reads according to locus-specific truncation lengths:</span>
<span id="cb44-245"><a href="#cb44-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-246"><a href="#cb44-246" aria-hidden="true" tabindex="-1"></a><span class="in">```{r declare_tr_dir}</span></span>
<span id="cb44-247"><a href="#cb44-247" aria-hidden="true" tabindex="-1"></a>tr_dir <span class="ot">&lt;-</span> <span class="st">"data/intermediate/truncated"</span></span>
<span id="cb44-248"><a href="#cb44-248" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-249"><a href="#cb44-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-250"><a href="#cb44-250" aria-hidden="true" tabindex="-1"></a><span class="in">```{r truncate}</span></span>
<span id="cb44-251"><a href="#cb44-251" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb44-252"><a href="#cb44-252" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb44-253"><a href="#cb44-253" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr eval_truncate</span></span>
<span id="cb44-254"><a href="#cb44-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-255"><a href="#cb44-255" aria-hidden="true" tabindex="-1"></a><span class="co"># truncate</span></span>
<span id="cb44-256"><a href="#cb44-256" aria-hidden="true" tabindex="-1"></a>trunc_out <span class="ot">&lt;-</span></span>
<span id="cb44-257"><a href="#cb44-257" aria-hidden="true" tabindex="-1"></a>  <span class="fu">trunc_amp</span>(<span class="at">in_dir =</span> demult,</span>
<span id="cb44-258"><a href="#cb44-258" aria-hidden="true" tabindex="-1"></a>          <span class="at">fw_pattern =</span> <span class="st">"1.fastq.gz"</span>,</span>
<span id="cb44-259"><a href="#cb44-259" aria-hidden="true" tabindex="-1"></a>          <span class="at">rv_pattern =</span> <span class="st">"2.fastq.gz"</span>,</span>
<span id="cb44-260"><a href="#cb44-260" aria-hidden="true" tabindex="-1"></a>          <span class="at">trunc_fr =</span> trunc_fr,</span>
<span id="cb44-261"><a href="#cb44-261" aria-hidden="true" tabindex="-1"></a>          <span class="at">write_trun =</span> tr_dir,</span>
<span id="cb44-262"><a href="#cb44-262" aria-hidden="true" tabindex="-1"></a>          <span class="at">max_ee =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">5</span>),</span>
<span id="cb44-263"><a href="#cb44-263" aria-hidden="true" tabindex="-1"></a>          <span class="at">trunc_q =</span> <span class="dv">2</span>)</span>
<span id="cb44-264"><a href="#cb44-264" aria-hidden="true" tabindex="-1"></a><span class="co"># save reads in and out</span></span>
<span id="cb44-265"><a href="#cb44-265" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(trunc_out, tr_path)</span>
<span id="cb44-266"><a href="#cb44-266" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-267"><a href="#cb44-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-268"><a href="#cb44-268" aria-hidden="true" tabindex="-1"></a>The output from <span class="in">`trunc_amp()`</span> is a list of matrices with IN and OUT reads after truncation.</span>
<span id="cb44-269"><a href="#cb44-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-272"><a href="#cb44-272" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-273"><a href="#cb44-273" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb44-274"><a href="#cb44-274" aria-hidden="true" tabindex="-1"></a>trunc_out <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(tr_path)</span>
<span id="cb44-275"><a href="#cb44-275" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-276"><a href="#cb44-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-279"><a href="#cb44-279" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-280"><a href="#cb44-280" aria-hidden="true" tabindex="-1"></a><span class="co"># see trunc_out</span></span>
<span id="cb44-281"><a href="#cb44-281" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(trunc_out[<span class="fu">seq_len</span>(<span class="dv">3</span>)], head, <span class="dv">3</span>)</span>
<span id="cb44-282"><a href="#cb44-282" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-283"><a href="#cb44-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-284"><a href="#cb44-284" aria-hidden="true" tabindex="-1"></a>Truncated FASTQ with low number of reads can be removed:</span>
<span id="cb44-285"><a href="#cb44-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-288"><a href="#cb44-288" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-289"><a href="#cb44-289" aria-hidden="true" tabindex="-1"></a><span class="fu">remove_poor_fastq</span>(tr_dir,</span>
<span id="cb44-290"><a href="#cb44-290" aria-hidden="true" tabindex="-1"></a>                   <span class="at">min_reads =</span> <span class="dv">10</span>)</span>
<span id="cb44-291"><a href="#cb44-291" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-292"><a href="#cb44-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-293"><a href="#cb44-293" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exploration of the parameter space</span></span>
<span id="cb44-294"><a href="#cb44-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-295"><a href="#cb44-295" aria-hidden="true" tabindex="-1"></a>The function <span class="in">`explore_dada()`</span> can be used to explore the effect of some DADA2 parameters ("OMEGA_A", "BAND_SIZE", "pool") on the sensitivity on the variant calling.</span>
<span id="cb44-296"><a href="#cb44-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-297"><a href="#cb44-297" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>*omega_a*: threshold for variants to be significant overabundant *log(-log(birth_pval))* (see <span class="co">[</span><span class="ot">Rosen et al. 2012</span><span class="co">](https://doi.org/10.1186/1471-2105-13-283)</span>).</span>
<span id="cb44-298"><a href="#cb44-298" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>*band_size*: positive numbers set a band size in Needleman-Wunsch alignments and ends free alignment is performed. A value of zero turns off banding, triggering full Needleman-Wunsch alignments, in which gapless alignment is performed (see <span class="co">[</span><span class="ot">issue</span><span class="co">](https://github.com/benjjneb/dada2/issues/1982)</span>).</span>
<span id="cb44-299"><a href="#cb44-299" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>*pool*: calling variants pooling samples can increase sensitivity (see <span class="co">[</span><span class="ot">dicussion</span><span class="co">](https://benjjneb.github.io/dada2/pseudo.html)</span>).</span>
<span id="cb44-300"><a href="#cb44-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-301"><a href="#cb44-301" aria-hidden="true" tabindex="-1"></a>The returned plots can be used to guide the selection of the best *OMEGA_A* in `variant_call()`, or frequency (*maf*) and abundance thresholds (*ad*) for filtering variants.</span>
<span id="cb44-302"><a href="#cb44-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-303"><a href="#cb44-303" aria-hidden="true" tabindex="-1"></a>Explore variants:</span>
<span id="cb44-304"><a href="#cb44-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-307"><a href="#cb44-307" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-308"><a href="#cb44-308" aria-hidden="true" tabindex="-1"></a><span class="co"># declare candidate OMEGA_A to use in variant_call()</span></span>
<span id="cb44-309"><a href="#cb44-309" aria-hidden="true" tabindex="-1"></a>candidate_omega_a <span class="ot">&lt;-</span> <span class="dv">10</span><span class="sc">^-</span><span class="dv">2</span></span>
<span id="cb44-310"><a href="#cb44-310" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-311"><a href="#cb44-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-312"><a href="#cb44-312" aria-hidden="true" tabindex="-1"></a><span class="in">```{r explore_dada}</span></span>
<span id="cb44-313"><a href="#cb44-313" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr eval_x</span></span>
<span id="cb44-314"><a href="#cb44-314" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb44-315"><a href="#cb44-315" aria-hidden="true" tabindex="-1"></a><span class="co"># paths to forward and reverse truncated reads</span></span>
<span id="cb44-316"><a href="#cb44-316" aria-hidden="true" tabindex="-1"></a>ftrun <span class="ot">&lt;-</span></span>
<span id="cb44-317"><a href="#cb44-317" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>(tr_dir, <span class="at">pattern =</span> <span class="st">"_F_"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb44-318"><a href="#cb44-318" aria-hidden="true" tabindex="-1"></a>rtrun <span class="ot">&lt;-</span></span>
<span id="cb44-319"><a href="#cb44-319" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>(tr_dir, <span class="at">pattern =</span> <span class="st">"_R_"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb44-320"><a href="#cb44-320" aria-hidden="true" tabindex="-1"></a><span class="co"># candidate omega value to annotate vline in plots.</span></span>
<span id="cb44-321"><a href="#cb44-321" aria-hidden="true" tabindex="-1"></a>v_line <span class="ot">&lt;-</span> <span class="fu">log</span>(<span class="sc">-</span><span class="fu">log</span>(candidate_omega_a))</span>
<span id="cb44-322"><a href="#cb44-322" aria-hidden="true" tabindex="-1"></a><span class="co"># run explore_dada() with band_size = 0, non pooling and omega_a = 0.9</span></span>
<span id="cb44-323"><a href="#cb44-323" aria-hidden="true" tabindex="-1"></a><span class="co"># forward</span></span>
<span id="cb44-324"><a href="#cb44-324" aria-hidden="true" tabindex="-1"></a>F_ind_0 <span class="ot">&lt;-</span></span>
<span id="cb44-325"><a href="#cb44-325" aria-hidden="true" tabindex="-1"></a>  <span class="fu">explore_dada</span>(ftrun, <span class="at">band_size =</span> <span class="dv">0</span>, <span class="at">vline =</span> v_line, <span class="at">hline_fr =</span> <span class="fl">0.1</span>)</span>
<span id="cb44-326"><a href="#cb44-326" aria-hidden="true" tabindex="-1"></a><span class="co"># reverse</span></span>
<span id="cb44-327"><a href="#cb44-327" aria-hidden="true" tabindex="-1"></a>R_ind_0 <span class="ot">&lt;-</span></span>
<span id="cb44-328"><a href="#cb44-328" aria-hidden="true" tabindex="-1"></a>  <span class="fu">explore_dada</span>(rtrun, <span class="at">band_size =</span> <span class="dv">0</span>, <span class="at">vline =</span> v_line, <span class="at">hline_fr =</span> <span class="fl">0.1</span>)</span>
<span id="cb44-329"><a href="#cb44-329" aria-hidden="true" tabindex="-1"></a><span class="co"># save results</span></span>
<span id="cb44-330"><a href="#cb44-330" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(<span class="fu">list</span>(<span class="at">F_ind_0 =</span> F_ind_0, <span class="at">R_ind_0 =</span> R_ind_0), x_out_path)</span>
<span id="cb44-331"><a href="#cb44-331" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-332"><a href="#cb44-332" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-333"><a href="#cb44-333" aria-hidden="true" tabindex="-1"></a>Exploration of DADA2 clustering for forward (A, C) and reverse (B, D) reads in @fig-explore_dada. The Y-axis represents the frequency of the variant in each locus and sample. The *log(-log(birth_pval))* transformation in the X-axis is related to the *p*-value of a variant being significantly overabundant. Larger x-values represent likely true variants. For representation purposes *birth_pval* of 0 (thus negative infinite), are converted to 10. Points are color-coded according to the variant rank in read abundance for its given locus and sample. For diploid individuals, <span class="co">[</span><span class="ot">green</span><span class="co">]</span>{style="color: green;"} are likely true variants and <span class="co">[</span><span class="ot">red</span><span class="co">]</span>{style="color: red;"} are likely false variants. <span class="co">[</span><span class="ot">Grey</span><span class="co">]</span>{style="color: grey;"} dashed lines are thresholds used for <span class="in">`variant_call()`</span>:</span>
<span id="cb44-334"><a href="#cb44-334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-335"><a href="#cb44-335" aria-hidden="true" tabindex="-1"></a><span class="in">```{r x_attach_if_not_run}</span></span>
<span id="cb44-336"><a href="#cb44-336" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb44-337"><a href="#cb44-337" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb44-338"><a href="#cb44-338" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span>(<span class="fu">exists</span>(<span class="st">"F_ind_0"</span>) <span class="sc">||</span> <span class="fu">exists</span>(<span class="st">"R_ind_0"</span>))) {</span>
<span id="cb44-339"><a href="#cb44-339" aria-hidden="true" tabindex="-1"></a>    x_dada <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(x_out_path)</span>
<span id="cb44-340"><a href="#cb44-340" aria-hidden="true" tabindex="-1"></a>    <span class="fu">attach</span>(x_dada)</span>
<span id="cb44-341"><a href="#cb44-341" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb44-342"><a href="#cb44-342" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-343"><a href="#cb44-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-344"><a href="#cb44-344" aria-hidden="true" tabindex="-1"></a><span class="in">```{r plot_explore_dada}</span></span>
<span id="cb44-345"><a href="#cb44-345" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-explore_dada</span></span>
<span id="cb44-346"><a href="#cb44-346" aria-hidden="true" tabindex="-1"></a><span class="co">#| warnings: false</span></span>
<span id="cb44-347"><a href="#cb44-347" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Exploration of DADA2 variants."</span></span>
<span id="cb44-348"><a href="#cb44-348" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb44-349"><a href="#cb44-349" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb44-350"><a href="#cb44-350" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb44-351"><a href="#cb44-351" aria-hidden="true" tabindex="-1"></a>ppool <span class="ot">&lt;-</span></span>
<span id="cb44-352"><a href="#cb44-352" aria-hidden="true" tabindex="-1"></a>  (F_ind_0<span class="sc">$</span>p1 <span class="sc">|</span> R_ind_0<span class="sc">$</span>p1) <span class="sc">/</span> (F_ind_0<span class="sc">$</span>p2 <span class="sc">|</span> R_ind_0<span class="sc">$</span>p2) <span class="sc">+</span></span>
<span id="cb44-353"><a href="#cb44-353" aria-hidden="true" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="st">"Dada F/R pool = F, omega_a 0.9, band_size = 0"</span>,</span>
<span id="cb44-354"><a href="#cb44-354" aria-hidden="true" tabindex="-1"></a>                             <span class="at">tag_levels =</span> <span class="st">"A"</span>)</span>
<span id="cb44-355"><a href="#cb44-355" aria-hidden="true" tabindex="-1"></a><span class="co"># save plot with combined loci</span></span>
<span id="cb44-356"><a href="#cb44-356" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"output/explore_dada.pdf"</span>, ppool, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">5</span>)</span>
<span id="cb44-357"><a href="#cb44-357" aria-hidden="true" tabindex="-1"></a><span class="co"># print plot</span></span>
<span id="cb44-358"><a href="#cb44-358" aria-hidden="true" tabindex="-1"></a>ppool</span>
<span id="cb44-359"><a href="#cb44-359" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-360"><a href="#cb44-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-361"><a href="#cb44-361" aria-hidden="true" tabindex="-1"></a>After the exploration variants in @fig-explore_dada, it seems an *OMEGA_A* = <span class="in">`r candidate_omega_a`</span>, implying a cut-off of <span class="in">`r sprintf("%.2f", log(-log(candidate_omega_a)))`</span> in the X-axis and a frequency threshold (Y-axis) of 0.1, excludes most <span class="co">[</span><span class="ot">artifacts</span><span class="co">]</span>{style="color: red;"} while maximizing <span class="co">[</span><span class="ot">true positives</span><span class="co">]</span>{style="color: green;"}.</span>
<span id="cb44-362"><a href="#cb44-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-363"><a href="#cb44-363" aria-hidden="true" tabindex="-1"></a>The results can also be explored **per-locus**. For instance, @fig-explore_dada B can be expanded per locus in @fig-explore_dada_loci.</span>
<span id="cb44-364"><a href="#cb44-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-367"><a href="#cb44-367" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-368"><a href="#cb44-368" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: true</span></span>
<span id="cb44-369"><a href="#cb44-369" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb44-370"><a href="#cb44-370" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb44-371"><a href="#cb44-371" aria-hidden="true" tabindex="-1"></a><span class="co"># list of plots per locus</span></span>
<span id="cb44-372"><a href="#cb44-372" aria-hidden="true" tabindex="-1"></a>lplots <span class="ot">&lt;-</span></span>
<span id="cb44-373"><a href="#cb44-373" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(<span class="at">loci_f_ind_0_logp =</span> F_ind_0<span class="sc">$</span>p3,</span>
<span id="cb44-374"><a href="#cb44-374" aria-hidden="true" tabindex="-1"></a>       <span class="at">loci_r_ind_0_logp =</span> R_ind_0<span class="sc">$</span>p3,</span>
<span id="cb44-375"><a href="#cb44-375" aria-hidden="true" tabindex="-1"></a>       <span class="at">loci_f_ind_0_abun =</span> F_ind_0<span class="sc">$</span>p4,</span>
<span id="cb44-376"><a href="#cb44-376" aria-hidden="true" tabindex="-1"></a>       <span class="at">loci_r_ind_0_abun =</span> R_ind_0<span class="sc">$</span>p4)</span>
<span id="cb44-377"><a href="#cb44-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-378"><a href="#cb44-378" aria-hidden="true" tabindex="-1"></a><span class="co"># save plots per locus</span></span>
<span id="cb44-379"><a href="#cb44-379" aria-hidden="true" tabindex="-1"></a><span class="fu">invisible</span>(</span>
<span id="cb44-380"><a href="#cb44-380" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(<span class="fu">seq_along</span>(lplots), <span class="cf">function</span>(x) {</span>
<span id="cb44-381"><a href="#cb44-381" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">paste0</span>(<span class="st">"output/"</span>, <span class="fu">names</span>(lplots)[x], <span class="st">".pdf"</span>),</span>
<span id="cb44-382"><a href="#cb44-382" aria-hidden="true" tabindex="-1"></a>         lplots[[x]], <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">6</span>)</span>
<span id="cb44-383"><a href="#cb44-383" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb44-384"><a href="#cb44-384" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb44-385"><a href="#cb44-385" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-386"><a href="#cb44-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-389"><a href="#cb44-389" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-390"><a href="#cb44-390" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-explore_dada_loci</span></span>
<span id="cb44-391"><a href="#cb44-391" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb44-392"><a href="#cb44-392" aria-hidden="true" tabindex="-1"></a><span class="co">#| warnings: false</span></span>
<span id="cb44-393"><a href="#cb44-393" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Exploration of DADA2 variants per locus for R reads."</span></span>
<span id="cb44-394"><a href="#cb44-394" aria-hidden="true" tabindex="-1"></a><span class="co">#| out-width: '90%'</span></span>
<span id="cb44-395"><a href="#cb44-395" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb44-396"><a href="#cb44-396" aria-hidden="true" tabindex="-1"></a>lplots<span class="sc">$</span>loci_r_ind_0_logp</span>
<span id="cb44-397"><a href="#cb44-397" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-398"><a href="#cb44-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-399"><a href="#cb44-399" aria-hidden="true" tabindex="-1"></a><span class="fu">## Variant and genotype calling</span></span>
<span id="cb44-400"><a href="#cb44-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-401"><a href="#cb44-401" aria-hidden="true" tabindex="-1"></a>Variant calling is run using *OMEGA_A* = <span class="in">`r candidate_omega_a`</span>, and under different parameters:</span>
<span id="cb44-402"><a href="#cb44-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-403"><a href="#cb44-403" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>*band_size*: 0, 16</span>
<span id="cb44-404"><a href="#cb44-404" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>*pool*: TRUE, FALSE</span>
<span id="cb44-405"><a href="#cb44-405" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-406"><a href="#cb44-406" aria-hidden="true" tabindex="-1"></a><span class="in">```{r variant_call}</span></span>
<span id="cb44-407"><a href="#cb44-407" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb44-408"><a href="#cb44-408" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr eval_variant_call</span></span>
<span id="cb44-409"><a href="#cb44-409" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb44-410"><a href="#cb44-410" aria-hidden="true" tabindex="-1"></a><span class="co"># list of parameters to run variant_call() with 4 different set of values:</span></span>
<span id="cb44-411"><a href="#cb44-411" aria-hidden="true" tabindex="-1"></a>variant_call_params <span class="ot">&lt;-</span></span>
<span id="cb44-412"><a href="#cb44-412" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(</span>
<span id="cb44-413"><a href="#cb44-413" aria-hidden="true" tabindex="-1"></a>    <span class="at">ind_bs16 =</span> <span class="fu">c</span>(<span class="cn">FALSE</span>, <span class="dv">16</span>), <span class="co"># pool = F, default band size == 16</span></span>
<span id="cb44-414"><a href="#cb44-414" aria-hidden="true" tabindex="-1"></a>    <span class="at">ind_bs0 =</span> <span class="fu">c</span>(<span class="cn">FALSE</span>, <span class="dv">0</span>), <span class="co"># pool = F, default band size == 0</span></span>
<span id="cb44-415"><a href="#cb44-415" aria-hidden="true" tabindex="-1"></a>    <span class="at">pool_bs16 =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="dv">16</span>), <span class="co"># pool = T, default band size == 16</span></span>
<span id="cb44-416"><a href="#cb44-416" aria-hidden="true" tabindex="-1"></a>    <span class="at">pool_bs0 =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="dv">0</span>) <span class="co"># pool = T, default band size == 0</span></span>
<span id="cb44-417"><a href="#cb44-417" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb44-418"><a href="#cb44-418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-419"><a href="#cb44-419" aria-hidden="true" tabindex="-1"></a>variants_x <span class="ot">&lt;-</span></span>
<span id="cb44-420"><a href="#cb44-420" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(variant_call_params, <span class="cf">function</span>(x) {</span>
<span id="cb44-421"><a href="#cb44-421" aria-hidden="true" tabindex="-1"></a>    <span class="fu">variant_call</span>(<span class="at">in_folder =</span> tr_dir,</span>
<span id="cb44-422"><a href="#cb44-422" aria-hidden="true" tabindex="-1"></a>                 <span class="at">rv_pattern =</span> <span class="st">"R_filt.fastq.gz"</span>,</span>
<span id="cb44-423"><a href="#cb44-423" aria-hidden="true" tabindex="-1"></a>                 <span class="at">c_unmerged =</span> <span class="cn">TRUE</span>,</span>
<span id="cb44-424"><a href="#cb44-424" aria-hidden="true" tabindex="-1"></a>                 <span class="at">multithread =</span> <span class="cn">TRUE</span>,</span>
<span id="cb44-425"><a href="#cb44-425" aria-hidden="true" tabindex="-1"></a>                 <span class="at">pool =</span> <span class="fu">as.logical</span>(x[<span class="dv">1</span>]),</span>
<span id="cb44-426"><a href="#cb44-426" aria-hidden="true" tabindex="-1"></a>                 <span class="at">omega_a_f =</span> candidate_omega_a,</span>
<span id="cb44-427"><a href="#cb44-427" aria-hidden="true" tabindex="-1"></a>                 <span class="at">omega_a_r =</span> candidate_omega_a,</span>
<span id="cb44-428"><a href="#cb44-428" aria-hidden="true" tabindex="-1"></a>                 <span class="at">band_size =</span> x[<span class="dv">2</span>],</span>
<span id="cb44-429"><a href="#cb44-429" aria-hidden="true" tabindex="-1"></a>                 <span class="at">ad =</span> <span class="dv">10</span>,</span>
<span id="cb44-430"><a href="#cb44-430" aria-hidden="true" tabindex="-1"></a>                 <span class="at">maf =</span> <span class="fl">0.1</span>)</span>
<span id="cb44-431"><a href="#cb44-431" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb44-432"><a href="#cb44-432" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(variants_x, <span class="st">"output/variants_x.rds"</span>)</span>
<span id="cb44-433"><a href="#cb44-433" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-434"><a href="#cb44-434" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-437"><a href="#cb44-437" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-438"><a href="#cb44-438" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb44-439"><a href="#cb44-439" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"variants_x"</span>))</span>
<span id="cb44-440"><a href="#cb44-440" aria-hidden="true" tabindex="-1"></a>  variants_x <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/variants_x.rds"</span>)</span>
<span id="cb44-441"><a href="#cb44-441" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-442"><a href="#cb44-442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-443"><a href="#cb44-443" aria-hidden="true" tabindex="-1"></a>Samples are genotyped from variant data with defaults <span class="in">`ploidy = 2`</span> and <span class="in">`ADt = 10`</span>:</span>
<span id="cb44-444"><a href="#cb44-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-447"><a href="#cb44-447" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-448"><a href="#cb44-448" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb44-449"><a href="#cb44-449" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr eval_genotype</span></span>
<span id="cb44-450"><a href="#cb44-450" aria-hidden="true" tabindex="-1"></a>genotypes_x <span class="ot">&lt;-</span></span>
<span id="cb44-451"><a href="#cb44-451" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(variants_x, genotype)</span>
<span id="cb44-452"><a href="#cb44-452" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-453"><a href="#cb44-453" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-456"><a href="#cb44-456" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-457"><a href="#cb44-457" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb44-458"><a href="#cb44-458" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr eval_genotype</span></span>
<span id="cb44-459"><a href="#cb44-459" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(genotypes_x, <span class="at">file =</span> <span class="st">"output/genotypes_x.rds"</span>)</span>
<span id="cb44-460"><a href="#cb44-460" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-461"><a href="#cb44-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-464"><a href="#cb44-464" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-465"><a href="#cb44-465" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb44-466"><a href="#cb44-466" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"genotypes_x"</span>))</span>
<span id="cb44-467"><a href="#cb44-467" aria-hidden="true" tabindex="-1"></a>  genotypes_x <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/genotypes_x.rds"</span>)</span>
<span id="cb44-468"><a href="#cb44-468" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-469"><a href="#cb44-469" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-470"><a href="#cb44-470" aria-hidden="true" tabindex="-1"></a><span class="fu">## Tidy data</span></span>
<span id="cb44-471"><a href="#cb44-471" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-472"><a href="#cb44-472" aria-hidden="true" tabindex="-1"></a>A strength from *tidyGenR* is that variants and genotypes from <span class="in">`variant_call()`</span> and <span class="in">`genotype()`</span> are returned as <span class="in">`tidy`</span> tables: one row per observation and variables in columns. Lets have a look at the data structure.</span>
<span id="cb44-473"><a href="#cb44-473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-474"><a href="#cb44-474" aria-hidden="true" tabindex="-1"></a><span class="in">```{r glimpse_tidy_variants}</span></span>
<span id="cb44-475"><a href="#cb44-475" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse tidy variants</span></span>
<span id="cb44-476"><a href="#cb44-476" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(variants_x<span class="sc">$</span>ind_bs0))</span>
<span id="cb44-477"><a href="#cb44-477" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-478"><a href="#cb44-478" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-479"><a href="#cb44-479" aria-hidden="true" tabindex="-1"></a><span class="in">```{r glimpse_tidy_genotypes}</span></span>
<span id="cb44-480"><a href="#cb44-480" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse tidy genotypes</span></span>
<span id="cb44-481"><a href="#cb44-481" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(genotypes_x<span class="sc">$</span>ind_bs0))</span>
<span id="cb44-482"><a href="#cb44-482" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-483"><a href="#cb44-483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-484"><a href="#cb44-484" aria-hidden="true" tabindex="-1"></a><span class="fu">## Genotype with AmpliSAT</span></span>
<span id="cb44-485"><a href="#cb44-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-486"><a href="#cb44-486" aria-hidden="true" tabindex="-1"></a>Genotyping was also carried with <span class="co">[</span><span class="ot">AmpliSAT</span><span class="co">](https://doi.org/10.1111/1755-0998.12453)</span>, a software written in PERL with similar characteristics to *tidyGenR*. To compare their performance we run AmpliSAT in a DOCKER container to genotype the same raw data. The steps are detailed <span class="co">[</span><span class="ot">here</span><span class="co">](code/amplisas)</span>. AmpliSAS returns results in a multisheet EXCEL and in plain text files, one per locus. The function <span class="in">`amplisas2tidy()`</span> permits to read plain text results from AmpliSAS into tidy variants.</span>
<span id="cb44-487"><a href="#cb44-487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-488"><a href="#cb44-488" aria-hidden="true" tabindex="-1"></a>Create input for AmpliSAS:</span>
<span id="cb44-489"><a href="#cb44-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-492"><a href="#cb44-492" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-493"><a href="#cb44-493" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb44-494"><a href="#cb44-494" aria-hidden="true" tabindex="-1"></a><span class="co"># amplicon metadata</span></span>
<span id="cb44-495"><a href="#cb44-495" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"code/amplisas/01_create_amplicon_data.R"</span>)</span>
<span id="cb44-496"><a href="#cb44-496" aria-hidden="true" tabindex="-1"></a><span class="co"># append barcodes to reads</span></span>
<span id="cb44-497"><a href="#cb44-497" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"code/amplisas/02_add_barcodes.R"</span>)</span>
<span id="cb44-498"><a href="#cb44-498" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-499"><a href="#cb44-499" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-500"><a href="#cb44-500" aria-hidden="true" tabindex="-1"></a>AmpliSAS is run in a DOCKER container. The steps are detailed <span class="co">[</span><span class="ot">here</span><span class="co">](code/amplisat_analysis.md)</span>.</span>
<span id="cb44-501"><a href="#cb44-501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-502"><a href="#cb44-502" aria-hidden="true" tabindex="-1"></a>AmpliSAS results can be read to tidy variants:</span>
<span id="cb44-503"><a href="#cb44-503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-504"><a href="#cb44-504" aria-hidden="true" tabindex="-1"></a><span class="in">```{r read_amplisas}</span></span>
<span id="cb44-505"><a href="#cb44-505" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb44-506"><a href="#cb44-506" aria-hidden="true" tabindex="-1"></a>fp <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/intermediate/amplisas/results_amplisas/filtered"</span>,</span>
<span id="cb44-507"><a href="#cb44-507" aria-hidden="true" tabindex="-1"></a>                 <span class="at">pattern =</span> <span class="st">"txt$"</span>,</span>
<span id="cb44-508"><a href="#cb44-508" aria-hidden="true" tabindex="-1"></a>                 <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb44-509"><a href="#cb44-509" aria-hidden="true" tabindex="-1"></a><span class="co"># amplisas results to tidy variants</span></span>
<span id="cb44-510"><a href="#cb44-510" aria-hidden="true" tabindex="-1"></a>var_amplisas <span class="ot">&lt;-</span></span>
<span id="cb44-511"><a href="#cb44-511" aria-hidden="true" tabindex="-1"></a>  <span class="fu">amplisas2tidy</span>(fp)</span>
<span id="cb44-512"><a href="#cb44-512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-513"><a href="#cb44-513" aria-hidden="true" tabindex="-1"></a><span class="co"># to genotypes</span></span>
<span id="cb44-514"><a href="#cb44-514" aria-hidden="true" tabindex="-1"></a>gen_amplisas <span class="ot">&lt;-</span></span>
<span id="cb44-515"><a href="#cb44-515" aria-hidden="true" tabindex="-1"></a>  <span class="fu">genotype</span>(var_amplisas)</span>
<span id="cb44-516"><a href="#cb44-516" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(var_amplisas, <span class="dv">3</span>))</span>
<span id="cb44-517"><a href="#cb44-517" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-518"><a href="#cb44-518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-519"><a href="#cb44-519" aria-hidden="true" tabindex="-1"></a><span class="fu">## Compare results</span></span>
<span id="cb44-520"><a href="#cb44-520" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-521"><a href="#cb44-521" aria-hidden="true" tabindex="-1"></a><span class="in">`compare_calls()`</span> was used for comparing results of *tidyGenR* run with different sets of parameters between them and agains AmpliSAT results. The comparison indicates that the set of parameters that returns the most similar genotype calls to AmpliSAT is using a band size = 0.</span>
<span id="cb44-522"><a href="#cb44-522" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-525"><a href="#cb44-525" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-526"><a href="#cb44-526" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb44-527"><a href="#cb44-527" aria-hidden="true" tabindex="-1"></a><span class="co"># compare genotypes of the four runs plus amplisat</span></span>
<span id="cb44-528"><a href="#cb44-528" aria-hidden="true" tabindex="-1"></a><span class="co">#   genotypes</span></span>
<span id="cb44-529"><a href="#cb44-529" aria-hidden="true" tabindex="-1"></a>gen_comb <span class="ot">&lt;-</span></span>
<span id="cb44-530"><a href="#cb44-530" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(genotypes_x,</span>
<span id="cb44-531"><a href="#cb44-531" aria-hidden="true" tabindex="-1"></a>    <span class="fu">list</span>(gen_amplisas))</span>
<span id="cb44-532"><a href="#cb44-532" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(gen_comb) <span class="ot">&lt;-</span> <span class="fu">c</span>(stringr<span class="sc">::</span><span class="fu">str_c</span>(<span class="st">"tidyGenR_"</span>, <span class="fu">names</span>(genotypes_x)), <span class="st">"AmpliSAT"</span>)</span>
<span id="cb44-533"><a href="#cb44-533" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-534"><a href="#cb44-534" aria-hidden="true" tabindex="-1"></a>comp_gen <span class="ot">&lt;-</span></span>
<span id="cb44-535"><a href="#cb44-535" aria-hidden="true" tabindex="-1"></a>  <span class="fu">suppressMessages</span>(<span class="fu">compare_calls</span>(gen_comb,</span>
<span id="cb44-536"><a href="#cb44-536" aria-hidden="true" tabindex="-1"></a>                <span class="st">"output/comp_gen_noReads.xlsx"</span>,</span>
<span id="cb44-537"><a href="#cb44-537" aria-hidden="true" tabindex="-1"></a>                <span class="at">creads =</span> <span class="cn">FALSE</span>))</span>
<span id="cb44-538"><a href="#cb44-538" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(comp_gen<span class="sc">$</span>dist)</span>
<span id="cb44-539"><a href="#cb44-539" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-540"><a href="#cb44-540" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-543"><a href="#cb44-543" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-544"><a href="#cb44-544" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb44-545"><a href="#cb44-545" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(comp_gen, <span class="st">"output/comp_gen.rds"</span>)</span>
<span id="cb44-546"><a href="#cb44-546" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-547"><a href="#cb44-547" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-548"><a href="#cb44-548" aria-hidden="true" tabindex="-1"></a>A plot produced by <span class="in">`compare_calls()`</span> depicts the differences between strategies. It is useful for spotting potential biases and problematic loci (@fig-comp_gen_all).</span>
<span id="cb44-549"><a href="#cb44-549" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-552"><a href="#cb44-552" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-553"><a href="#cb44-553" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Comparison between genotype calls."</span></span>
<span id="cb44-554"><a href="#cb44-554" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-comp_gen_all</span></span>
<span id="cb44-555"><a href="#cb44-555" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 12</span></span>
<span id="cb44-556"><a href="#cb44-556" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 10</span></span>
<span id="cb44-557"><a href="#cb44-557" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb44-558"><a href="#cb44-558" aria-hidden="true" tabindex="-1"></a>comp_gen<span class="sc">$</span>plot2</span>
<span id="cb44-559"><a href="#cb44-559" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-560"><a href="#cb44-560" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-561"><a href="#cb44-561" aria-hidden="true" tabindex="-1"></a>Problematic loci, with low coverage and with conflicting genotypes between different strategies were checked manually by aligning them and compared the with dereplicated reads. Setting band_size = 0 yielded the best results as it recovered alleles which had indels at their ends. Pooling samples did not affect genotype calls, although it has some influence on the number of reads supporting each allele. Thus samples not sharing alleles (i.e. different species) can be genotyped equally efficiently as samples sharing alleles (i.e. same species). Then, the run with *band size = 0* and *non-pooled* samples was chosen as the best strategy and compared pairwise with AmpliSAT.</span>
<span id="cb44-562"><a href="#cb44-562" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-565"><a href="#cb44-565" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-566"><a href="#cb44-566" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb44-567"><a href="#cb44-567" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb44-568"><a href="#cb44-568" aria-hidden="true" tabindex="-1"></a><span class="co"># compare best call with amplisas</span></span>
<span id="cb44-569"><a href="#cb44-569" aria-hidden="true" tabindex="-1"></a><span class="co">#   variants</span></span>
<span id="cb44-570"><a href="#cb44-570" aria-hidden="true" tabindex="-1"></a>comp_gen_sel <span class="ot">&lt;-</span></span>
<span id="cb44-571"><a href="#cb44-571" aria-hidden="true" tabindex="-1"></a>  <span class="fu">compare_calls</span>(<span class="fu">list</span>(<span class="at">tidyGenR_ind_bs0 =</span> genotypes_x<span class="sc">$</span>ind_bs0,</span>
<span id="cb44-572"><a href="#cb44-572" aria-hidden="true" tabindex="-1"></a>                      <span class="at">AmpliSAT =</span> gen_amplisas),</span>
<span id="cb44-573"><a href="#cb44-573" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"output/comp_gen_indbs0_amplisat.xlsx"</span>,</span>
<span id="cb44-574"><a href="#cb44-574" aria-hidden="true" tabindex="-1"></a>                  <span class="at">creads =</span> <span class="cn">TRUE</span>)</span>
<span id="cb44-575"><a href="#cb44-575" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(comp_gen_sel, <span class="st">"output/comp_gen_sel.rds"</span>)</span>
<span id="cb44-576"><a href="#cb44-576" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-577"><a href="#cb44-577" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-578"><a href="#cb44-578" aria-hidden="true" tabindex="-1"></a><span class="in">```{r plot_amplisat_vs_tidygenr}</span></span>
<span id="cb44-579"><a href="#cb44-579" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-compgen-amplisas</span></span>
<span id="cb44-580"><a href="#cb44-580" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb44-581"><a href="#cb44-581" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Comparison of genotype call tidyGenR vs AmpliSAT"</span></span>
<span id="cb44-582"><a href="#cb44-582" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb44-583"><a href="#cb44-583" aria-hidden="true" tabindex="-1"></a>pcomp1 <span class="ot">&lt;-</span> comp_gen_sel<span class="sc">$</span>plot2</span>
<span id="cb44-584"><a href="#cb44-584" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(pcomp1,</span>
<span id="cb44-585"><a href="#cb44-585" aria-hidden="true" tabindex="-1"></a>       <span class="at">file =</span> <span class="st">"output/comp_amplisat_indo.pdf"</span>,</span>
<span id="cb44-586"><a href="#cb44-586" aria-hidden="true" tabindex="-1"></a>       <span class="at">width =</span> <span class="dv">10</span>,</span>
<span id="cb44-587"><a href="#cb44-587" aria-hidden="true" tabindex="-1"></a>       <span class="at">height =</span> <span class="fl">5.4</span>)</span>
<span id="cb44-588"><a href="#cb44-588" aria-hidden="true" tabindex="-1"></a>pcomp1</span>
<span id="cb44-589"><a href="#cb44-589" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-590"><a href="#cb44-590" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-591"><a href="#cb44-591" aria-hidden="true" tabindex="-1"></a><span class="in">```{r plot_n_reads_comp}</span></span>
<span id="cb44-592"><a href="#cb44-592" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-compgen-amplisas_nreads</span></span>
<span id="cb44-593"><a href="#cb44-593" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb44-594"><a href="#cb44-594" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Comparison of number of reads supporting alleles in tidyGenR vs AmpliSAT."</span></span>
<span id="cb44-595"><a href="#cb44-595" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb44-596"><a href="#cb44-596" aria-hidden="true" tabindex="-1"></a>pcomp2 <span class="ot">&lt;-</span> comp_gen_sel<span class="sc">$</span>plot3</span>
<span id="cb44-597"><a href="#cb44-597" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(pcomp2,</span>
<span id="cb44-598"><a href="#cb44-598" aria-hidden="true" tabindex="-1"></a>       <span class="at">file =</span> <span class="st">"output/comp_amplisat_ind0_boxplot.pdf"</span>,</span>
<span id="cb44-599"><a href="#cb44-599" aria-hidden="true" tabindex="-1"></a>       <span class="at">width =</span> <span class="dv">5</span>,</span>
<span id="cb44-600"><a href="#cb44-600" aria-hidden="true" tabindex="-1"></a>       <span class="at">height =</span> <span class="dv">3</span>)</span>
<span id="cb44-601"><a href="#cb44-601" aria-hidden="true" tabindex="-1"></a>pcomp2</span>
<span id="cb44-602"><a href="#cb44-602" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-603"><a href="#cb44-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-604"><a href="#cb44-604" aria-hidden="true" tabindex="-1"></a>Lastly, the genotypes can be compared against their respective variants to have an idea of variants being dropped when genotyping:</span>
<span id="cb44-605"><a href="#cb44-605" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-606"><a href="#cb44-606" aria-hidden="true" tabindex="-1"></a><span class="in">```{r drop_variants}</span></span>
<span id="cb44-607"><a href="#cb44-607" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-dropped_var</span></span>
<span id="cb44-608"><a href="#cb44-608" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb44-609"><a href="#cb44-609" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb44-610"><a href="#cb44-610" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Identified variants dropped during genotyping."</span></span>
<span id="cb44-611"><a href="#cb44-611" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb44-612"><a href="#cb44-612" aria-hidden="true" tabindex="-1"></a>dropped_var <span class="ot">&lt;-</span></span>
<span id="cb44-613"><a href="#cb44-613" aria-hidden="true" tabindex="-1"></a>      <span class="fu">list</span>(<span class="at">variants =</span> variants_x<span class="sc">$</span>ind_bs0,</span>
<span id="cb44-614"><a href="#cb44-614" aria-hidden="true" tabindex="-1"></a>            <span class="at">genotypes =</span> genotypes_x<span class="sc">$</span>ind_bs0)</span>
<span id="cb44-615"><a href="#cb44-615" aria-hidden="true" tabindex="-1"></a>comp_var_gen <span class="ot">&lt;-</span> <span class="fu">compare_calls</span>(dropped_var)</span>
<span id="cb44-616"><a href="#cb44-616" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(comp_var_gen, <span class="st">"output/comp_var_gen.rds"</span>)</span>
<span id="cb44-617"><a href="#cb44-617" aria-hidden="true" tabindex="-1"></a>comp_var_gen<span class="sc">$</span>plot1</span>
<span id="cb44-618"><a href="#cb44-618" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-619"><a href="#cb44-619" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb44-620"><a href="#cb44-620" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-621"><a href="#cb44-621" aria-hidden="true" tabindex="-1"></a><span class="fu"># Session Info</span></span>
<span id="cb44-622"><a href="#cb44-622" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-625"><a href="#cb44-625" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb44-626"><a href="#cb44-626" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb44-627"><a href="#cb44-627" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>