<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Miguel Camacho Sánchez">

<title>tidyGenR benchmarking</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="README_files/libs/clipboard/clipboard.min.js"></script>
<script src="README_files/libs/quarto-html/quarto.js"></script>
<script src="README_files/libs/quarto-html/popper.min.js"></script>
<script src="README_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="README_files/libs/quarto-html/anchor.min.js"></script>
<link href="README_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="README_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="README_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="README_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="README_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#benchmarking-tidygenr" id="toc-benchmarking-tidygenr" class="nav-link active" data-scroll-target="#benchmarking-tidygenr">Benchmarking <em>tidyGenR</em></a></li>
  <li><a href="#preparation-of-input-data" id="toc-preparation-of-input-data" class="nav-link" data-scroll-target="#preparation-of-input-data">Preparation of input data</a></li>
  <li><a href="#demultiplex-by-locus" id="toc-demultiplex-by-locus" class="nav-link" data-scroll-target="#demultiplex-by-locus">Demultiplex by locus</a></li>
  <li><a href="#truncate-reads" id="toc-truncate-reads" class="nav-link" data-scroll-target="#truncate-reads">Truncate reads</a></li>
  <li><a href="#exploration-of-the-parameter-space" id="toc-exploration-of-the-parameter-space" class="nav-link" data-scroll-target="#exploration-of-the-parameter-space">Exploration of the parameter space</a></li>
  <li><a href="#variant-and-genotype-calling" id="toc-variant-and-genotype-calling" class="nav-link" data-scroll-target="#variant-and-genotype-calling">Variant and genotype calling</a></li>
  <li><a href="#tidy-data" id="toc-tidy-data" class="nav-link" data-scroll-target="#tidy-data">Tidy data</a></li>
  <li><a href="#tidygenr-vs-amplisas" id="toc-tidygenr-vs-amplisas" class="nav-link" data-scroll-target="#tidygenr-vs-amplisas"><em>tidyGenR</em> vs AmpliSAS</a></li>
  <li><a href="#integration-of-alleles-with-publised-data" id="toc-integration-of-alleles-with-publised-data" class="nav-link" data-scroll-target="#integration-of-alleles-with-publised-data">Integration of alleles with publised data</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">session info</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">tidyGenR benchmarking</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Miguel Camacho Sánchez </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<p>Attach libraries:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyGenR)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">

</div>
<section id="benchmarking-tidygenr" class="level2">
<h2 class="anchored" data-anchor-id="benchmarking-tidygenr">Benchmarking <em>tidyGenR</em></h2>
<p>This repository accompanies the R package <strong>tidyGenR</strong> available at <a href="https://github.com/csmiguel/tidyGenR" class="uri">https://github.com/csmiguel/tidyGenR</a>. It covers (1) all steps for genotype calling with <em>tidyGenR</em>, (2) exploration of the parameters for variant calling, and (3) comparison of <em>tidyGenR</em> against <a href="https://doi.org/10.1111/1755-0998.12453">AmpliSAS</a>. It uses real data from a population genetics study of the wild rodent <em>Rattus baluensis</em> (Camacho-Sanchez et al.&nbsp;<em>in preparation</em>).</p>
</section>
<section id="preparation-of-input-data" class="level2">
<h2 class="anchored" data-anchor-id="preparation-of-input-data">Preparation of input data</h2>
<p>Raw sequences are deposited in NCBI under the BioStudy <a href="https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP293699">SRP293699</a>, in the BioProject <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA680166">PRJNA680166</a>. From the BioStudy, we can download the <a href="data/raw/SraRunTable.csv">SRA metadata</a> to map sample IDs to SRRs:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># table with mapped SRR-sampleIDs</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>srr <span class="ot">&lt;-</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read.csv</span>(<span class="st">"data/raw/SraRunTable.csv"</span>) <span class="sc">|&gt;</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Run, <span class="st">"Sample.Name"</span>) <span class="sc">|&gt;</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">run =</span> Run,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">sample =</span> <span class="st">"Sample.Name"</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(srr, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">run</th>
<th style="text-align: left;">sample</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">SRR13106799</td>
<td style="text-align: left;">BOR1063</td>
</tr>
<tr class="even">
<td style="text-align: left;">SRR13106800</td>
<td style="text-align: left;">BOR1061</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SRR13106802</td>
<td style="text-align: left;">BOR626</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The FASTQ reads from the SRRs can be downloaded with <a href="https://github.com/ncbi/sra-tools">SRA-toolkit</a>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download raw reads from NCBI</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># they are not downloaded if already present in 'data/raw'.</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(srr, <span class="dv">1</span>, <span class="cf">function</span>(x) {</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># name paths</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  srr_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, x[<span class="dv">1</span>])</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  srr1 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">1</span>], <span class="st">"_1.fastq"</span>))</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  srr2 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">1</span>], <span class="st">"_2.fastq"</span>))</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  sam1 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">2</span>], <span class="st">"_1.fastq"</span>))</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>  sam2 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">2</span>], <span class="st">"_2.fastq"</span>))</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># fetch SRR</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">any</span>(<span class="fu">dir.exists</span>(srr_path) <span class="sc">||</span> <span class="fu">file.exists</span>(sam1))) {</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"prefetch"</span>, <span class="fu">paste</span>(<span class="st">"-O"</span>, <span class="st">"data/raw/"</span>, x[<span class="dv">1</span>]))</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># reformat to FASTQ</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(sam1)) {</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"fasterq-dump"</span>, <span class="fu">paste</span>(<span class="st">"-O"</span>, <span class="st">"data/raw/"</span>, srr_path))</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"mv"</span>, <span class="fu">paste</span>(srr1, sam1))</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"mv"</span>, <span class="fu">paste</span>(srr2, sam2))</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># rm SRR</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"rm"</span>, <span class="fu">paste0</span>(<span class="st">"-r "</span>, srr_path, <span class="st">"*"</span>))</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>  })</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>NULL</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse raw FASTQ</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(<span class="st">"data/raw"</span>, <span class="at">pattern =</span> <span class="st">"fastq"</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "BOR1061_1.fastq" "BOR1061_2.fastq" "BOR1063_1.fastq"</code></pre>
</div>
</div>
<p>Check input data files:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>freads <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/raw"</span>, <span class="at">pattern =</span> <span class="st">"1.fastq"</span>,</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>rreads <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/raw"</span>, <span class="at">pattern =</span> <span class="st">"2.fastq"</span>,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>chr <span class="ot">&lt;-</span> <span class="fu">check_raw_reads</span>(freads, rreads, <span class="at">low_readcount =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>All F and (R files) passed check on number of reads above 10.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Sample names are unique.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>All F files have their corresponding R file.</code></pre>
</div>
</div>
<p>Input FASTQ is conditions for <em>tidyGenR</em>. A total of 44 are detected:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>chr<span class="sc">$</span>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "BOR1061" "BOR1063" "BOR1069" "BOR1070" "BOR1071" "BOR1074" "BOR1075"
 [8] "BOR1076" "BOR1077" "BOR1079" "BOR1080" "BOR1082" "BOR1086" "BOR1088"
[15] "BOR1090" "BOR1091" "BOR1097" "BOR1098" "BOR1101" "BOR1107" "BOR1111"
[22] "BOR1113" "BOR1115" "BOR1125" "BOR1127" "BOR1129" "BOR1131" "BOR1133"
[29] "BOR1135" "BOR1137" "BOR1139" "BOR1141" "BOR1143" "BOR1147" "BOR1149"
[36] "BOR1151" "BOR1155" "BOR614"  "BOR615"  "BOR616"  "BOR618"  "BOR619" 
[43] "BOR621"  "BOR626" </code></pre>
</div>
</div>
</section>
<section id="demultiplex-by-locus" class="level2">
<h2 class="anchored" data-anchor-id="demultiplex-by-locus">Demultiplex by locus</h2>
<p>Reads are demultiplexed by locus using primer sequences in paired-end mode.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load primer data</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"primers"</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># path to cutadapt</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>cutadapt <span class="ot">&lt;-</span> <span class="fu">system</span>(<span class="st">"which cutadapt"</span>, <span class="at">intern =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># path to folder save locus-demultiplexed FASTQ</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>demult <span class="ot">&lt;-</span> <span class="st">"data/intermediate/demultiplexed"</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co"># print primers</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(primers, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">locus</th>
<th style="text-align: left;">fw</th>
<th style="text-align: left;">rv</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">nfkbia</td>
<td style="text-align: left;">GCCTCCAAACACACAGTCAT</td>
<td style="text-align: left;">TGAGGAGAGCTATGACACGG</td>
</tr>
<tr class="even">
<td style="text-align: left;">chrna9</td>
<td style="text-align: left;">TTATCTGGGAGAGCGTGACC</td>
<td style="text-align: left;">TTGGGAAARGATGAACCGGC</td>
</tr>
<tr class="odd">
<td style="text-align: left;">rogdi</td>
<td style="text-align: left;">AGAARCCGGCTCACTACCC</td>
<td style="text-align: left;">GAGGCACAGCTTGTTGAGG</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># demultiplex</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">demultiplex</span>(</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">interpreter =</span> <span class="st">"/bin/bash"</span>,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">cutadapt =</span> cutadapt,</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">freads =</span> freads,</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">rreads =</span> rreads,</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">primers =</span> primers,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">sh_out =</span> <span class="st">"code/demultiplex.sh"</span>,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">write_demultiplexed =</span> demult,</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">log_out =</span> c_log,</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">mode =</span> <span class="st">"pe"</span>,</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">run =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse demultiplexed FASTQ</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(demult, <span class="at">pattern =</span> <span class="st">"fastq"</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "BOR1061.abcg8.1.fastq.gz"  "BOR1061.abcg8.2.fastq.gz" 
[3] "BOR1061.alkbh7.1.fastq.gz"</code></pre>
</div>
</div>
<p>Remove files with few reads:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">remove_poor_fastq</span>(demult,</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">min_reads =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="truncate-reads" class="level2">
<h2 class="anchored" data-anchor-id="truncate-reads">Truncate reads</h2>
<p>Reads are truncated to a given length for each locus. A <code>data.frame</code> with truncation lengths for forward and reverse reads was built to maximize the amount of information yielded by each locus. For instance, the amplicons for some loci, as <em>nfkbia</em>, are long and it is a good trade-off to leave the low quality ends in but to be sure both F and R reads overlap.</p>
<p>Building of <code>data.frame</code> with locus-specific truncation lengths:</p>
<p>Truncate reads according to locus-specific truncation lengths:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>tr_dir <span class="ot">&lt;-</span> <span class="st">"data/intermediate/truncated"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># truncate</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>trunc_out <span class="ot">&lt;-</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">trunc_amp</span>(<span class="at">in_dir =</span> demult,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">fw_pattern =</span> <span class="st">"1.fastq.gz"</span>,</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">rv_pattern =</span> <span class="st">"2.fastq.gz"</span>,</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">trunc_fr =</span> trunc_fr,</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">write_trun =</span> tr_dir,</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">max_ee =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">5</span>),</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">trunc_q =</span> <span class="dv">2</span>)</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co"># save reads in and out</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(trunc_out, tr_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The output from <code>trunc_amp()</code> is a list of matrices with IN and OUT reads after truncation:</p>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># see trunc_out</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(trunc_out[<span class="fu">seq_len</span>(<span class="dv">3</span>)], head, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$abcg8
        reads.in reads.out
BOR1061     1716      1039
BOR1063     1616       992
BOR1069     1634       961

$alkbh7
        reads.in reads.out
BOR1061      546       398
BOR1063      499       361
BOR1069      546       374

$apeh17
        reads.in reads.out
BOR1061     3719      2773
BOR1063     3761      2793
BOR1069     3756      2838</code></pre>
</div>
</div>
</section>
<section id="exploration-of-the-parameter-space" class="level2">
<h2 class="anchored" data-anchor-id="exploration-of-the-parameter-space">Exploration of the parameter space</h2>
<p>The function <code>explore_dada()</code> can be used to explore the effect of DADA2 parameters (“OMEGA_A”, “BAND_SIZE”, “pool”) on the sensitivity on the variant calling.</p>
<ul>
<li><em>omega_a</em>: threshold for variants to be significant overabundant <em>log(-log(birth_pval))</em> (see <a href="https://doi.org/10.1186/1471-2105-13-283.">Rosen et al.&nbsp;2012</a>).</li>
<li><em>band_size</em>: positive numbers set a band size in Needleman-Wunsch alignments. In this context, ends free alignment is performed. A value of zero turns off banding, triggering full Needleman-Wunsch alignments, in which gapless alignment is performed (see <a href="https://github.com/benjjneb/dada2/issues/1982">issue</a>).</li>
<li><em>pool</em>: calling variants pooling samples can increase sensitivity (see <a href="https://benjjneb.github.io/dada2/pseudo.html">dicussion</a>).</li>
</ul>
<p>The returned plots can be used to guide the election of the best <em>OMEGA_A</em> in <code>variant_call()</code>, or frequency (<em>maf</em>) and abundance thresholds (<em>ad</em>) for filtering variants.</p>
<p>Explore variants:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># declare candidate OMEGA_A to use in variant_call()</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>candidate_omega_a <span class="ot">&lt;-</span> <span class="dv">10</span><span class="sc">^-</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># paths to forward and reverse truncated reads</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>ftrun <span class="ot">&lt;-</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>(tr_dir, <span class="at">pattern =</span> <span class="st">"_F_"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>rtrun <span class="ot">&lt;-</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>(tr_dir, <span class="at">pattern =</span> <span class="st">"_R_"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="co"># candidate omegavalue to annotate vline in plots.</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>v_line <span class="ot">&lt;-</span> <span class="fu">log</span>(<span class="sc">-</span><span class="fu">log</span>(candidate_omega_a))</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="co"># run explore_dada() with band_size = 0, non pooling and omega_a = 0.9</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="co"># forward</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>F_ind_0 <span class="ot">&lt;-</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">explore_dada</span>(ftrun, <span class="at">band_size =</span> <span class="dv">0</span>, <span class="at">vline =</span> v_line, <span class="at">hline_fr =</span> <span class="fl">0.1</span>)</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="co"># reverse</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>R_ind_0 <span class="ot">&lt;-</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">explore_dada</span>(rtrun, <span class="at">band_size =</span> <span class="dv">0</span>, <span class="at">vline =</span> v_line, <span class="at">hline_fr =</span> <span class="fl">0.1</span>)</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a><span class="co"># save results</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(<span class="fu">list</span>(<span class="at">F_ind_0 =</span> F_ind_0, <span class="at">R_ind_0 =</span> R_ind_0), x_out_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Exploration of DADA2 clustering for forward (A, C) and reverse (B, D) reads. The Y-axis represents the frequency of the variant in each locus and sample. The <em>log(-log(birth_pval))</em> transformation in the X-axis is related to the p-value of a variant being significantly overabundant. Larger x-values represent likely true variants. For representation purposes <em>birth_pval</em> of 0 (thus negative infinite), are converted to 10. Points are color-coded according to the variant rank in read abundance for its given locus and sample. For diploid individuals, <span style="color: green;">green</span> are likely true variants and <span style="color: red;">red</span> are likely false variants. <span style="color: grey;">Grey</span> dashed lines are thresholds used for <code>variant_call()</code>:</p>
<div class="cell">

</div>
<div class="cell" data-warnings="false">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>ppool <span class="ot">&lt;-</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  (F_ind_0<span class="sc">$</span>p1 <span class="sc">|</span> R_ind_0<span class="sc">$</span>p1) <span class="sc">/</span> (F_ind_0<span class="sc">$</span>p2 <span class="sc">|</span> R_ind_0<span class="sc">$</span>p2) <span class="sc">+</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="st">"Dada F/R pool = F, omega_a 0.9, band_size = 0"</span>,</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">tag_levels =</span> <span class="st">"A"</span>)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co"># save plot with combined loci</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"output/explore_dada.pdf"</span>, ppool, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">5</span>)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="co"># print plot</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>ppool</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-explore_dada" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="README_files/figure-html/fig-explore_dada-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;1: Exploration of DADA2 variants</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>After the exploration variants in <a href="#fig-explore_dada">Figure&nbsp;1</a>, it seems an <em>OMEGA_A</em> = 0.01, implying a cutoff of 1.5271796 in the X-axis and a frequency threshold (Y-axis) of 0.1, excludes most <span style="color: red;">artifacts</span> while maximizing <span style="color: red;">true positives</span>.</p>
<p>The results can also be explored <strong>per-locus</strong>. For instance, <a href="#fig-explore_dada">Figure&nbsp;1</a> B can be expanded per locus as follows:</p>
<div class="cell" data-warnings="false">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># list of plots per locus</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>lplots <span class="ot">&lt;-</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(<span class="at">loci_f_ind_0_logp =</span> F_ind_0<span class="sc">$</span>p3,</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">loci_r_ind_0_logp =</span> R_ind_0<span class="sc">$</span>p3,</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">loci_f_ind_0_abun =</span> F_ind_0<span class="sc">$</span>p4,</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">loci_r_ind_0_abun =</span> R_ind_0<span class="sc">$</span>p4)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save plots per locus</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(<span class="fu">seq_along</span>(lplots), <span class="cf">function</span>(x) {</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">paste0</span>(<span class="st">"output/"</span>, <span class="fu">names</span>(lplots)[x], <span class="st">".pdf"</span>),</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>         lplots[[x]], <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">6</span>)</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a><span class="do">## [[1]]</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "output/loci_f_ind_0_logp.pdf"</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a><span class="do">## [[2]]</span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "output/loci_r_ind_0_logp.pdf"</span></span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a><span class="do">## [[3]]</span></span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "output/loci_f_ind_0_abun.pdf"</span></span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a><span class="do">## [[4]]</span></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "output/loci_r_ind_0_abun.pdf"</span></span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a>lplots<span class="sc">$</span>loci_r_ind_0_logp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-explore_dada_loci" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="README_files/figure-html/fig-explore_dada_loci-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2: Exploration of DADA2 variants per locus for R reads</figcaption><p></p>
</figure>
</div>
</div>
</div>
</section>
<section id="variant-and-genotype-calling" class="level2">
<h2 class="anchored" data-anchor-id="variant-and-genotype-calling">Variant and genotype calling</h2>
<p>Variant calling is run using <em>OMEGA_A</em> = 0.01, and under different parameters:</p>
<ul>
<li><em>band_size</em>: 0, 16</li>
<li><em>pool</em>: TRUE, FALSE</li>
</ul>
<div class="cell">

</div>
<p>Variants are used to genotype individuals with defaults <code>ploidy = 2</code> and <code>ADt = 10</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>genotypes_x <span class="ot">&lt;-</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(variants_x, genotype)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(genotypes_x, <span class="at">file =</span> <span class="st">"output/genotypes_x.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">

</div>
</section>
<section id="tidy-data" class="level2">
<h2 class="anchored" data-anchor-id="tidy-data">Tidy data</h2>
<p>A strong point from <em>tidyGenR</em> is that variants and genotypes from <code>variant_call()</code> and <code>genotype()</code> are returned in <code>tidy</code> format: one row per observation and variables in columns. Lets have a look at the data structure.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse tidy variants</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(variants_x<span class="sc">$</span>ind_bs0))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 6%">
<col style="width: 86%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">sample</th>
<th style="text-align: left;">locus</th>
<th style="text-align: left;">variant</th>
<th style="text-align: right;">reads</th>
<th style="text-align: right;">nt</th>
<th style="text-align: left;">md5</th>
<th style="text-align: left;">sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">abcg8</td>
<td style="text-align: left;">01</td>
<td style="text-align: right;">1028</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">c3eab42bea56937040aa79a6bbea2724</td>
<td style="text-align: left;">TTGCCCACCCTGTTCATCCATGGAGCAGAAGCCTGCCTGATGTCTCTCATCATTGGCTTCCTTTACTACGGCCACGCAGATAAGCCGCTCTCCTTCGTGGACATGGCAGCCCTCCTGTTCATGATAGGAGCGCTCATTCCTTTTAATGTCATTCTGGATGTCGTCTCCAAATGTGAGTGTCACCCGCCCTCCTCACCAGACATCGGGACAGTGGGACAGCCTCCCTGGGCACTGCACTGAGGCCAAGCTCTGTGCTTCCGCTGGTACCCACGGCATTACAAGAGATGCGACCTCAGTAACACTCTTCGCTCATTCACCTCCCTCTCCCCCATCTCCAGGTCACTCGGAGCGCTCGCTGCTGTACTATGAACTGGAGGACGGACTGTACAC</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">alkbh7</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">393</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">620057f2e547e688d1220db91a3a1d01</td>
<td style="text-align: left;">GCCTTCCTGGCCCCATCCCCTCTGGGAGGGAGCGGCAAATCACTGAGATGCGTCGGCCCCGGGGGACCCGGTGAGCCCCAAAAAATGATTCTTCATCTCTAAGGATCTCATGGGAGAAGTCATATCGGGCTGAACCCCTGCAAGAAAATAAAGGCCAAGTAGGTGAAGGGAGGGAAGTCTGTCCCTTCATCATTTCTGTACTTTATCTGGGTAGTGGTGATGGTGACTTCCTCTGCTATGAGAGCAGAAACAAGAGCTGTTGGAAACACTTGGCCTCATCTGGGGGTTCTAGGCTAGGTCATACCTTAGGATATAGAGAGAACCAGGTTCCAGCAACAGTTCCAGCCACTGCTCAGGTTCCTGTGTATGAACCAGCTTCATAACACTTGG</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">apeh17</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">2747</td>
<td style="text-align: right;">415</td>
<td style="text-align: left;">394a483d3ae2ce5605f1674b2400986c</td>
<td style="text-align: left;">AAAGCCAGTGGAGCCACGATAGTTCACTGCAAGTGAGACAAGATAGTCGGGCAATGTTCTCAGTCCCCAACCCAATTCCATGTGTCTGTGACATAGGCCCTAGGACCAGCTGCATCACGTGCTGTCAAGGGGAGGTAGCCAAGAGAGATGAAGCTGCTACCCTGAGACCACAGCATCCTTTCCACAGAAGCTTCAAGTTAATCCACGTGATCACCAGGTTATGAAGCCAGAGGCTAAGCCAGGGCAATTTCTAGTAGTAGTTTCTTTTTGTTTTCAAAACAGGGTTTTTCTGTGCAGTCCTGGATGATCTTGGACTCAAACCTTCACCTGCCTCTGCCTCCTAAGTGCTGGATTAAAGGAGTGTGCCCCCATTGCCCAGCCCGTTTCTAAAGAGTAAGTACACCATAGAAAAGCA</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">cd27</td>
<td style="text-align: left;">01</td>
<td style="text-align: right;">1143</td>
<td style="text-align: right;">379</td>
<td style="text-align: left;">a897a500c934797d4b3662415fc92456</td>
<td style="text-align: left;">AGTCTTCCTGGATAGGGATGACGCTGCCCTCCTCCTCCCTGGGGCAGCTGTAAGGACAAAGCTCTTCAGGTACTGCCTGGCTATCTTCATCTGTGCAAAGACAATTAGCCAAGTGTTGGTCAGCAGTGGAGAGAAGAGAGGGGAAGGTGAGGAGAGAGGAGAAGGCCGAGTGGAGGCTGGGGCATGGGGGAGCCAGGGGAGCCTGTGGGAAGGACACTTGAAGAACCAGAGAAGGTGGGTGAAGGTGGGATGGGGGCTTTAGGTGTGGGTGGCAGAGCTGAGAGGGCAGGAGGGAAGGCCTGTGCCTTACTTGGCCCGTGATTTCTTCTTTGACGGAAGAACAAGATCCCACCCAGGACGAAGACAAGAAGCATGCTGG</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">chrna9</td>
<td style="text-align: left;">01</td>
<td style="text-align: right;">21</td>
<td style="text-align: right;">408</td>
<td style="text-align: left;">3d77c726462281336567895361ceebc1</td>
<td style="text-align: left;">TGCAGTGTGACATTCAGCACCGCGTCCGTATCCTCGACTGGACGCAGAGCACTGGAGTAGTCTTCAAAAAGATCGCTGAACAATTTCTGAGCATATTTCCCATTTGCTGTCTCTACGGCTGTTCAAAGAGAAGCACCCGGATGGGCATTTCAAAACAAGACCAACTCTGGGGTCAAAATGGGGGATGGCAACTTTGGATGAGTTCTTTTTTGGGGGCAGGGGACTGGTCTTACCTACCTTGATGATCGAGAGTCAGAAAGAGGTGCGAGCACGCCCATGTACTCACGTCAGTCCTCCTGTCTCTCTCACCAAGGCTCTAAGGAGCCTCTATGGAGCACTGCGCTAGCCCCTCACCTCTGATTCCAGAAGCAGCAAAATACATCCAGCAAAAGGAGATGCAGGACTGGG</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">chrna9</td>
<td style="text-align: left;">02</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">408</td>
<td style="text-align: left;">9b805048a29d6233f1ac41f2a6aa1421</td>
<td style="text-align: left;">TGCAGTGTGACATTCAGCACCGCGTCCGTATCCTCGACTGGACGCAGAGCACTGGAGTAGTCTTCAAAAAGATCGCTGAATAATTTCTGAGCATATTTCCCATTTGCTGTCTCTACGGCTGTTCAAAGAGAAGCACCCGGATGGGCATTTCAAAACAAGACCAACTCTGGGATCAAAATGGGGGATGGCAACTTTGGATGAGTTCTTTTTTGGGGGCAGGGGACTGGTCTTACCTACCTTGATGATCAAGAGTCAGAAAGAGGTGCGAGCACGCCCATGTACTCACGTCAGTCCTCCTGTCTCTCTCACCAAGGCTCTAAGGAGCCTCTATGGAGCACTGCGCTAGCCCCTCACCTCTGATTCCAGAAGCAGCAAAATACATCCAGCAAAAGGAGATGCAGGACTGGG</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse tidy genotypes</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(genotypes_x<span class="sc">$</span>ind_bs0))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 6%">
<col style="width: 84%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">sample</th>
<th style="text-align: left;">locus</th>
<th style="text-align: left;">allele</th>
<th style="text-align: right;">allele_no</th>
<th style="text-align: right;">reads</th>
<th style="text-align: right;">nt</th>
<th style="text-align: left;">md5</th>
<th style="text-align: left;">sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">abcg8</td>
<td style="text-align: left;">01</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">514.0</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">c3eab42bea56937040aa79a6bbea2724</td>
<td style="text-align: left;">TTGCCCACCCTGTTCATCCATGGAGCAGAAGCCTGCCTGATGTCTCTCATCATTGGCTTCCTTTACTACGGCCACGCAGATAAGCCGCTCTCCTTCGTGGACATGGCAGCCCTCCTGTTCATGATAGGAGCGCTCATTCCTTTTAATGTCATTCTGGATGTCGTCTCCAAATGTGAGTGTCACCCGCCCTCCTCACCAGACATCGGGACAGTGGGACAGCCTCCCTGGGCACTGCACTGAGGCCAAGCTCTGTGCTTCCGCTGGTACCCACGGCATTACAAGAGATGCGACCTCAGTAACACTCTTCGCTCATTCACCTCCCTCTCCCCCATCTCCAGGTCACTCGGAGCGCTCGCTGCTGTACTATGAACTGGAGGACGGACTGTACAC</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">abcg8</td>
<td style="text-align: left;">01</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">514.0</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">c3eab42bea56937040aa79a6bbea2724</td>
<td style="text-align: left;">TTGCCCACCCTGTTCATCCATGGAGCAGAAGCCTGCCTGATGTCTCTCATCATTGGCTTCCTTTACTACGGCCACGCAGATAAGCCGCTCTCCTTCGTGGACATGGCAGCCCTCCTGTTCATGATAGGAGCGCTCATTCCTTTTAATGTCATTCTGGATGTCGTCTCCAAATGTGAGTGTCACCCGCCCTCCTCACCAGACATCGGGACAGTGGGACAGCCTCCCTGGGCACTGCACTGAGGCCAAGCTCTGTGCTTCCGCTGGTACCCACGGCATTACAAGAGATGCGACCTCAGTAACACTCTTCGCTCATTCACCTCCCTCTCCCCCATCTCCAGGTCACTCGGAGCGCTCGCTGCTGTACTATGAACTGGAGGACGGACTGTACAC</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">alkbh7</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">196.5</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">620057f2e547e688d1220db91a3a1d01</td>
<td style="text-align: left;">GCCTTCCTGGCCCCATCCCCTCTGGGAGGGAGCGGCAAATCACTGAGATGCGTCGGCCCCGGGGGACCCGGTGAGCCCCAAAAAATGATTCTTCATCTCTAAGGATCTCATGGGAGAAGTCATATCGGGCTGAACCCCTGCAAGAAAATAAAGGCCAAGTAGGTGAAGGGAGGGAAGTCTGTCCCTTCATCATTTCTGTACTTTATCTGGGTAGTGGTGATGGTGACTTCCTCTGCTATGAGAGCAGAAACAAGAGCTGTTGGAAACACTTGGCCTCATCTGGGGGTTCTAGGCTAGGTCATACCTTAGGATATAGAGAGAACCAGGTTCCAGCAACAGTTCCAGCCACTGCTCAGGTTCCTGTGTATGAACCAGCTTCATAACACTTGG</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">alkbh7</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">196.5</td>
<td style="text-align: right;">390</td>
<td style="text-align: left;">620057f2e547e688d1220db91a3a1d01</td>
<td style="text-align: left;">GCCTTCCTGGCCCCATCCCCTCTGGGAGGGAGCGGCAAATCACTGAGATGCGTCGGCCCCGGGGGACCCGGTGAGCCCCAAAAAATGATTCTTCATCTCTAAGGATCTCATGGGAGAAGTCATATCGGGCTGAACCCCTGCAAGAAAATAAAGGCCAAGTAGGTGAAGGGAGGGAAGTCTGTCCCTTCATCATTTCTGTACTTTATCTGGGTAGTGGTGATGGTGACTTCCTCTGCTATGAGAGCAGAAACAAGAGCTGTTGGAAACACTTGGCCTCATCTGGGGGTTCTAGGCTAGGTCATACCTTAGGATATAGAGAGAACCAGGTTCCAGCAACAGTTCCAGCCACTGCTCAGGTTCCTGTGTATGAACCAGCTTCATAACACTTGG</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">apeh17</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1373.5</td>
<td style="text-align: right;">415</td>
<td style="text-align: left;">394a483d3ae2ce5605f1674b2400986c</td>
<td style="text-align: left;">AAAGCCAGTGGAGCCACGATAGTTCACTGCAAGTGAGACAAGATAGTCGGGCAATGTTCTCAGTCCCCAACCCAATTCCATGTGTCTGTGACATAGGCCCTAGGACCAGCTGCATCACGTGCTGTCAAGGGGAGGTAGCCAAGAGAGATGAAGCTGCTACCCTGAGACCACAGCATCCTTTCCACAGAAGCTTCAAGTTAATCCACGTGATCACCAGGTTATGAAGCCAGAGGCTAAGCCAGGGCAATTTCTAGTAGTAGTTTCTTTTTGTTTTCAAAACAGGGTTTTTCTGTGCAGTCCTGGATGATCTTGGACTCAAACCTTCACCTGCCTCTGCCTCCTAAGTGCTGGATTAAAGGAGTGTGCCCCCATTGCCCAGCCCGTTTCTAAAGAGTAAGTACACCATAGAAAAGCA</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: left;">apeh17</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1373.5</td>
<td style="text-align: right;">415</td>
<td style="text-align: left;">394a483d3ae2ce5605f1674b2400986c</td>
<td style="text-align: left;">AAAGCCAGTGGAGCCACGATAGTTCACTGCAAGTGAGACAAGATAGTCGGGCAATGTTCTCAGTCCCCAACCCAATTCCATGTGTCTGTGACATAGGCCCTAGGACCAGCTGCATCACGTGCTGTCAAGGGGAGGTAGCCAAGAGAGATGAAGCTGCTACCCTGAGACCACAGCATCCTTTCCACAGAAGCTTCAAGTTAATCCACGTGATCACCAGGTTATGAAGCCAGAGGCTAAGCCAGGGCAATTTCTAGTAGTAGTTTCTTTTTGTTTTCAAAACAGGGTTTTTCTGTGCAGTCCTGGATGATCTTGGACTCAAACCTTCACCTGCCTCTGCCTCCTAAGTGCTGGATTAAAGGAGTGTGCCCCCATTGCCCAGCCCGTTTCTAAAGAGTAAGTACACCATAGAAAAGCA</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="tidygenr-vs-amplisas" class="level2">
<h2 class="anchored" data-anchor-id="tidygenr-vs-amplisas"><em>tidyGenR</em> vs AmpliSAS</h2>
<p><a href="https://doi.org/10.1111/1755-0998.12453">AmpliSAS</a> is a software written in PERL with similar characteristics to <em>tidyGenR</em>. To compare their performance we run AmpliSAS in a docker container to genotype our raw sequences. The steps are detailed <a href="code/amplisas">here</a>. AmpliSAS returns results in a multisheet EXCEL and in plain text files, one per locus. The function <code>amplisas2tidy()</code> permits to read plain text results from AmpliSAS into tidy variants.</p>
<div class="cell">

</div>
</section>
<section id="integration-of-alleles-with-publised-data" class="level1">
<h1>Integration of alleles with publised data</h1>
<div class="cell">

</div>
</section>
<section id="session-info" class="level1">
<h1>session info</h1>
<p>Details on the sessionInfo() are <a href="session_info.txt">here</a>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">writeLines</span>(<span class="fu">capture.output</span>(<span class="fu">sessionInfo</span>()), <span class="at">con =</span> <span class="st">"session_info.txt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- -->

</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb30" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> Miguel Camacho Sánchez</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "tidyGenR benchmarking"</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span><span class="co"> visual</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="an">theme:</span><span class="co"> minty</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>Attach libraries:</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a><span class="in">```{r attach packages}</span></span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyGenR)</span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eval_conditionals}</span></span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb30-24"><a href="#cb30-24" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb30-25"><a href="#cb30-25" aria-hidden="true" tabindex="-1"></a><span class="co"># declare logicals to trigger EVAL in intensive code chunks</span></span>
<span id="cb30-26"><a href="#cb30-26" aria-hidden="true" tabindex="-1"></a><span class="co"># they are not run if output already exists.</span></span>
<span id="cb30-27"><a href="#cb30-27" aria-hidden="true" tabindex="-1"></a><span class="co"># to run a fresh analysis outputs need to be removed.</span></span>
<span id="cb30-28"><a href="#cb30-28" aria-hidden="true" tabindex="-1"></a><span class="co"># demultiplex</span></span>
<span id="cb30-29"><a href="#cb30-29" aria-hidden="true" tabindex="-1"></a>c_log <span class="ot">&lt;-</span> <span class="st">"output/cutadapt.log"</span></span>
<span id="cb30-30"><a href="#cb30-30" aria-hidden="true" tabindex="-1"></a>eval_demultiplex <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(c_log)</span>
<span id="cb30-31"><a href="#cb30-31" aria-hidden="true" tabindex="-1"></a><span class="co"># truncate</span></span>
<span id="cb30-32"><a href="#cb30-32" aria-hidden="true" tabindex="-1"></a>tr_path <span class="ot">&lt;-</span> <span class="st">"output/trunc_in-out.rds"</span></span>
<span id="cb30-33"><a href="#cb30-33" aria-hidden="true" tabindex="-1"></a>eval_truncate <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(tr_path)</span>
<span id="cb30-34"><a href="#cb30-34" aria-hidden="true" tabindex="-1"></a><span class="co"># explore_dada</span></span>
<span id="cb30-35"><a href="#cb30-35" aria-hidden="true" tabindex="-1"></a>x_out_path <span class="ot">&lt;-</span> <span class="st">"output/explore_dada.rds"</span></span>
<span id="cb30-36"><a href="#cb30-36" aria-hidden="true" tabindex="-1"></a>eval_x <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(x_out_path)</span>
<span id="cb30-37"><a href="#cb30-37" aria-hidden="true" tabindex="-1"></a><span class="co"># variant_call</span></span>
<span id="cb30-38"><a href="#cb30-38" aria-hidden="true" tabindex="-1"></a>x_variants_path <span class="ot">&lt;-</span>  <span class="st">"output/variants_x.rds"</span></span>
<span id="cb30-39"><a href="#cb30-39" aria-hidden="true" tabindex="-1"></a>eval_variant_call <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(x_variants_path)</span>
<span id="cb30-40"><a href="#cb30-40" aria-hidden="true" tabindex="-1"></a><span class="co"># genotype</span></span>
<span id="cb30-41"><a href="#cb30-41" aria-hidden="true" tabindex="-1"></a>x_genotypes_path <span class="ot">&lt;-</span>  <span class="st">"output/genotypes_x.rds"</span></span>
<span id="cb30-42"><a href="#cb30-42" aria-hidden="true" tabindex="-1"></a>eval_genotype <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">file.exists</span>(x_genotypes_path)</span>
<span id="cb30-43"><a href="#cb30-43" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-44"><a href="#cb30-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-45"><a href="#cb30-45" aria-hidden="true" tabindex="-1"></a><span class="fu">## Benchmarking *tidyGenR*</span></span>
<span id="cb30-46"><a href="#cb30-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-47"><a href="#cb30-47" aria-hidden="true" tabindex="-1"></a>This repository accompanies the R package **tidyGenR** available at <span class="ot">&lt;https://github.com/csmiguel/tidyGenR&gt;</span>. It covers (1) all steps for genotype calling with *tidyGenR*, (2) exploration of the parameters for variant calling, and (3) comparison of _tidyGenR_ against [AmpliSAS](https://doi.org/10.1111/1755-0998.12453). It uses real data from a population genetics study of the wild rodent *Rattus baluensis* (Camacho-Sanchez et al. _in preparation_).</span>
<span id="cb30-48"><a href="#cb30-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-49"><a href="#cb30-49" aria-hidden="true" tabindex="-1"></a><span class="fu">## Preparation of input data</span></span>
<span id="cb30-50"><a href="#cb30-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-51"><a href="#cb30-51" aria-hidden="true" tabindex="-1"></a>Raw sequences are deposited in NCBI under the BioStudy <span class="co">[</span><span class="ot">SRP293699</span><span class="co">](https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP293699)</span>, in the BioProject <span class="co">[</span><span class="ot">PRJNA680166</span><span class="co">](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA680166)</span>. From the BioStudy, we can download the <span class="co">[</span><span class="ot">SRA metadata</span><span class="co">](data/raw/SraRunTable.csv)</span> to map sample IDs to SRRs:</span>
<span id="cb30-52"><a href="#cb30-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-53"><a href="#cb30-53" aria-hidden="true" tabindex="-1"></a><span class="in">```{r SRR}</span></span>
<span id="cb30-54"><a href="#cb30-54" aria-hidden="true" tabindex="-1"></a><span class="co"># table with mapped SRR-sampleIDs</span></span>
<span id="cb30-55"><a href="#cb30-55" aria-hidden="true" tabindex="-1"></a>srr <span class="ot">&lt;-</span></span>
<span id="cb30-56"><a href="#cb30-56" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read.csv</span>(<span class="st">"data/raw/SraRunTable.csv"</span>) <span class="sc">|&gt;</span></span>
<span id="cb30-57"><a href="#cb30-57" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Run, <span class="st">"Sample.Name"</span>) <span class="sc">|&gt;</span></span>
<span id="cb30-58"><a href="#cb30-58" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">run =</span> Run,</span>
<span id="cb30-59"><a href="#cb30-59" aria-hidden="true" tabindex="-1"></a>         <span class="at">sample =</span> <span class="st">"Sample.Name"</span>)</span>
<span id="cb30-60"><a href="#cb30-60" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(srr, <span class="dv">3</span>))</span>
<span id="cb30-61"><a href="#cb30-61" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-62"><a href="#cb30-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-63"><a href="#cb30-63" aria-hidden="true" tabindex="-1"></a>The FASTQ reads from the SRRs can be downloaded with <span class="co">[</span><span class="ot">SRA-toolkit</span><span class="co">](https://github.com/ncbi/sra-tools)</span>:</span>
<span id="cb30-64"><a href="#cb30-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-65"><a href="#cb30-65" aria-hidden="true" tabindex="-1"></a><span class="in">```{r download_raw_data}</span></span>
<span id="cb30-66"><a href="#cb30-66" aria-hidden="true" tabindex="-1"></a><span class="co">#|  message: false</span></span>
<span id="cb30-67"><a href="#cb30-67" aria-hidden="true" tabindex="-1"></a><span class="co">#|  warning: false</span></span>
<span id="cb30-68"><a href="#cb30-68" aria-hidden="true" tabindex="-1"></a><span class="co">#|  error: false</span></span>
<span id="cb30-69"><a href="#cb30-69" aria-hidden="true" tabindex="-1"></a><span class="co">#|  collapse: true</span></span>
<span id="cb30-70"><a href="#cb30-70" aria-hidden="true" tabindex="-1"></a><span class="co"># download raw reads from NCBI</span></span>
<span id="cb30-71"><a href="#cb30-71" aria-hidden="true" tabindex="-1"></a><span class="co"># they are not downloaded if already present in 'data/raw'.</span></span>
<span id="cb30-72"><a href="#cb30-72" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(srr, <span class="dv">1</span>, <span class="cf">function</span>(x) {</span>
<span id="cb30-73"><a href="#cb30-73" aria-hidden="true" tabindex="-1"></a>  <span class="co"># name paths</span></span>
<span id="cb30-74"><a href="#cb30-74" aria-hidden="true" tabindex="-1"></a>  srr_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, x[<span class="dv">1</span>])</span>
<span id="cb30-75"><a href="#cb30-75" aria-hidden="true" tabindex="-1"></a>  srr1 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">1</span>], <span class="st">"_1.fastq"</span>))</span>
<span id="cb30-76"><a href="#cb30-76" aria-hidden="true" tabindex="-1"></a>  srr2 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">1</span>], <span class="st">"_2.fastq"</span>))</span>
<span id="cb30-77"><a href="#cb30-77" aria-hidden="true" tabindex="-1"></a>  sam1 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">2</span>], <span class="st">"_1.fastq"</span>))</span>
<span id="cb30-78"><a href="#cb30-78" aria-hidden="true" tabindex="-1"></a>  sam2 <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/raw"</span>, <span class="fu">paste0</span>(x[<span class="dv">2</span>], <span class="st">"_2.fastq"</span>))</span>
<span id="cb30-79"><a href="#cb30-79" aria-hidden="true" tabindex="-1"></a>  <span class="co"># fetch SRR</span></span>
<span id="cb30-80"><a href="#cb30-80" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">any</span>(<span class="fu">dir.exists</span>(srr_path) <span class="sc">||</span> <span class="fu">file.exists</span>(sam1))) {</span>
<span id="cb30-81"><a href="#cb30-81" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"prefetch"</span>, <span class="fu">paste</span>(<span class="st">"-O"</span>, <span class="st">"data/raw/"</span>, x[<span class="dv">1</span>]))</span>
<span id="cb30-82"><a href="#cb30-82" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb30-83"><a href="#cb30-83" aria-hidden="true" tabindex="-1"></a>  <span class="co"># reformat to FASTQ</span></span>
<span id="cb30-84"><a href="#cb30-84" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(sam1)) {</span>
<span id="cb30-85"><a href="#cb30-85" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"fasterq-dump"</span>, <span class="fu">paste</span>(<span class="st">"-O"</span>, <span class="st">"data/raw/"</span>, srr_path))</span>
<span id="cb30-86"><a href="#cb30-86" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"mv"</span>, <span class="fu">paste</span>(srr1, sam1))</span>
<span id="cb30-87"><a href="#cb30-87" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"mv"</span>, <span class="fu">paste</span>(srr2, sam2))</span>
<span id="cb30-88"><a href="#cb30-88" aria-hidden="true" tabindex="-1"></a>    <span class="co"># rm SRR</span></span>
<span id="cb30-89"><a href="#cb30-89" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system2</span>(<span class="st">"rm"</span>, <span class="fu">paste0</span>(<span class="st">"-r "</span>, srr_path, <span class="st">"*"</span>))</span>
<span id="cb30-90"><a href="#cb30-90" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb30-91"><a href="#cb30-91" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb30-92"><a href="#cb30-92" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse raw FASTQ</span></span>
<span id="cb30-93"><a href="#cb30-93" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(<span class="st">"data/raw"</span>, <span class="at">pattern =</span> <span class="st">"fastq"</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb30-94"><a href="#cb30-94" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-95"><a href="#cb30-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-96"><a href="#cb30-96" aria-hidden="true" tabindex="-1"></a>Check input data files:</span>
<span id="cb30-97"><a href="#cb30-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-98"><a href="#cb30-98" aria-hidden="true" tabindex="-1"></a><span class="in">```{r check_raw_reads}</span></span>
<span id="cb30-99"><a href="#cb30-99" aria-hidden="true" tabindex="-1"></a>freads <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/raw"</span>, <span class="at">pattern =</span> <span class="st">"1.fastq"</span>,</span>
<span id="cb30-100"><a href="#cb30-100" aria-hidden="true" tabindex="-1"></a>                    <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-101"><a href="#cb30-101" aria-hidden="true" tabindex="-1"></a>rreads <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/raw"</span>, <span class="at">pattern =</span> <span class="st">"2.fastq"</span>,</span>
<span id="cb30-102"><a href="#cb30-102" aria-hidden="true" tabindex="-1"></a>                    <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-103"><a href="#cb30-103" aria-hidden="true" tabindex="-1"></a>chr <span class="ot">&lt;-</span> <span class="fu">check_raw_reads</span>(freads, rreads, <span class="at">low_readcount =</span> <span class="dv">10</span>)</span>
<span id="cb30-104"><a href="#cb30-104" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-105"><a href="#cb30-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-106"><a href="#cb30-106" aria-hidden="true" tabindex="-1"></a>Input FASTQ is conditions for *tidyGenR*. A total of <span class="in">`r length(chr$samples)`</span> are detected:</span>
<span id="cb30-107"><a href="#cb30-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-110"><a href="#cb30-110" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-111"><a href="#cb30-111" aria-hidden="true" tabindex="-1"></a>chr<span class="sc">$</span>samples</span>
<span id="cb30-112"><a href="#cb30-112" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-113"><a href="#cb30-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-114"><a href="#cb30-114" aria-hidden="true" tabindex="-1"></a><span class="fu">## Demultiplex by locus</span></span>
<span id="cb30-115"><a href="#cb30-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-116"><a href="#cb30-116" aria-hidden="true" tabindex="-1"></a>Reads are demultiplexed by locus using primer sequences in paired-end mode.</span>
<span id="cb30-117"><a href="#cb30-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-118"><a href="#cb30-118" aria-hidden="true" tabindex="-1"></a><span class="in">```{r demultiplex_objects}</span></span>
<span id="cb30-119"><a href="#cb30-119" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb30-120"><a href="#cb30-120" aria-hidden="true" tabindex="-1"></a><span class="co"># load primer data</span></span>
<span id="cb30-121"><a href="#cb30-121" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"primers"</span>)</span>
<span id="cb30-122"><a href="#cb30-122" aria-hidden="true" tabindex="-1"></a><span class="co"># path to cutadapt</span></span>
<span id="cb30-123"><a href="#cb30-123" aria-hidden="true" tabindex="-1"></a>cutadapt <span class="ot">&lt;-</span> <span class="fu">system</span>(<span class="st">"which cutadapt"</span>, <span class="at">intern =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-124"><a href="#cb30-124" aria-hidden="true" tabindex="-1"></a><span class="co"># path to folder save locus-demultiplexed FASTQ</span></span>
<span id="cb30-125"><a href="#cb30-125" aria-hidden="true" tabindex="-1"></a>demult <span class="ot">&lt;-</span> <span class="st">"data/intermediate/demultiplexed"</span></span>
<span id="cb30-126"><a href="#cb30-126" aria-hidden="true" tabindex="-1"></a><span class="co"># print primers</span></span>
<span id="cb30-127"><a href="#cb30-127" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(primers, <span class="dv">3</span>))</span>
<span id="cb30-128"><a href="#cb30-128" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-129"><a href="#cb30-129" aria-hidden="true" tabindex="-1"></a><span class="in">```{r demultiplex}</span></span>
<span id="cb30-130"><a href="#cb30-130" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr eval_demultiplex</span></span>
<span id="cb30-131"><a href="#cb30-131" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb30-132"><a href="#cb30-132" aria-hidden="true" tabindex="-1"></a><span class="co"># demultiplex</span></span>
<span id="cb30-133"><a href="#cb30-133" aria-hidden="true" tabindex="-1"></a><span class="fu">demultiplex</span>(</span>
<span id="cb30-134"><a href="#cb30-134" aria-hidden="true" tabindex="-1"></a>  <span class="at">interpreter =</span> <span class="st">"/bin/bash"</span>,</span>
<span id="cb30-135"><a href="#cb30-135" aria-hidden="true" tabindex="-1"></a>  <span class="at">cutadapt =</span> cutadapt,</span>
<span id="cb30-136"><a href="#cb30-136" aria-hidden="true" tabindex="-1"></a>  <span class="at">freads =</span> freads,</span>
<span id="cb30-137"><a href="#cb30-137" aria-hidden="true" tabindex="-1"></a>  <span class="at">rreads =</span> rreads,</span>
<span id="cb30-138"><a href="#cb30-138" aria-hidden="true" tabindex="-1"></a>  <span class="at">primers =</span> primers,</span>
<span id="cb30-139"><a href="#cb30-139" aria-hidden="true" tabindex="-1"></a>  <span class="at">sh_out =</span> <span class="st">"code/demultiplex.sh"</span>,</span>
<span id="cb30-140"><a href="#cb30-140" aria-hidden="true" tabindex="-1"></a>  <span class="at">write_demultiplexed =</span> demult,</span>
<span id="cb30-141"><a href="#cb30-141" aria-hidden="true" tabindex="-1"></a>  <span class="at">log_out =</span> c_log,</span>
<span id="cb30-142"><a href="#cb30-142" aria-hidden="true" tabindex="-1"></a>  <span class="at">mode =</span> <span class="st">"pe"</span>,</span>
<span id="cb30-143"><a href="#cb30-143" aria-hidden="true" tabindex="-1"></a>  <span class="at">run =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-144"><a href="#cb30-144" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-147"><a href="#cb30-147" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-148"><a href="#cb30-148" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse demultiplexed FASTQ</span></span>
<span id="cb30-149"><a href="#cb30-149" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(demult, <span class="at">pattern =</span> <span class="st">"fastq"</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb30-150"><a href="#cb30-150" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-151"><a href="#cb30-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-152"><a href="#cb30-152" aria-hidden="true" tabindex="-1"></a>Remove files with few reads:</span>
<span id="cb30-153"><a href="#cb30-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-154"><a href="#cb30-154" aria-hidden="true" tabindex="-1"></a><span class="in">```{r rm_after_dem}</span></span>
<span id="cb30-155"><a href="#cb30-155" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb30-156"><a href="#cb30-156" aria-hidden="true" tabindex="-1"></a><span class="fu">remove_poor_fastq</span>(demult,</span>
<span id="cb30-157"><a href="#cb30-157" aria-hidden="true" tabindex="-1"></a>                   <span class="at">min_reads =</span> <span class="dv">10</span>)</span>
<span id="cb30-158"><a href="#cb30-158" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-159"><a href="#cb30-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-160"><a href="#cb30-160" aria-hidden="true" tabindex="-1"></a><span class="fu">## Truncate reads</span></span>
<span id="cb30-161"><a href="#cb30-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-162"><a href="#cb30-162" aria-hidden="true" tabindex="-1"></a>Reads are truncated to a given length for each locus. A <span class="in">`data.frame`</span> with truncation lengths for forward and reverse reads was built to maximize the amount of information yielded by each locus. For instance, the amplicons for some loci, as *nfkbia*, are long and it is a good trade-off to leave the low quality ends in but to be sure both F and R reads overlap.</span>
<span id="cb30-163"><a href="#cb30-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-164"><a href="#cb30-164" aria-hidden="true" tabindex="-1"></a>Building of <span class="in">`data.frame`</span> with locus-specific truncation lengths:</span>
<span id="cb30-165"><a href="#cb30-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-166"><a href="#cb30-166" aria-hidden="true" tabindex="-1"></a><span class="in">```{r build_trunc_fr}</span></span>
<span id="cb30-167"><a href="#cb30-167" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb30-168"><a href="#cb30-168" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb30-169"><a href="#cb30-169" aria-hidden="true" tabindex="-1"></a>loci <span class="ot">&lt;-</span> </span>
<span id="cb30-170"><a href="#cb30-170" aria-hidden="true" tabindex="-1"></a>  tidyGenR<span class="sc">:::</span><span class="fu">check_names_demultiplexed</span>(demult,</span>
<span id="cb30-171"><a href="#cb30-171" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">fw_pattern =</span> <span class="st">"1.fastq.gz"</span>,</span>
<span id="cb30-172"><a href="#cb30-172" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">rv_pattern =</span> <span class="st">"2.fastq.gz"</span>)<span class="sc">$</span>loci</span>
<span id="cb30-173"><a href="#cb30-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-174"><a href="#cb30-174" aria-hidden="true" tabindex="-1"></a>trunc_fr <span class="ot">&lt;-</span></span>
<span id="cb30-175"><a href="#cb30-175" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(<span class="at">locus =</span> loci,</span>
<span id="cb30-176"><a href="#cb30-176" aria-hidden="true" tabindex="-1"></a>             <span class="at">trunc_f =</span> <span class="dv">270</span>,</span>
<span id="cb30-177"><a href="#cb30-177" aria-hidden="true" tabindex="-1"></a>             <span class="at">trunc_r =</span> <span class="dv">180</span>)</span>
<span id="cb30-178"><a href="#cb30-178" aria-hidden="true" tabindex="-1"></a><span class="co"># introduce manual values</span></span>
<span id="cb30-179"><a href="#cb30-179" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"rgd735029"</span>), <span class="st">"trunc_f"</span>] <span class="ot">&lt;-</span>  <span class="dv">245</span></span>
<span id="cb30-180"><a href="#cb30-180" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"rgd735029"</span>), <span class="st">"trunc_r"</span>] <span class="ot">&lt;-</span>  <span class="dv">155</span></span>
<span id="cb30-181"><a href="#cb30-181" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"fancg"</span>), <span class="st">"trunc_r"</span>] <span class="ot">&lt;-</span> <span class="dv">190</span></span>
<span id="cb30-182"><a href="#cb30-182" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"nfkbia"</span>), <span class="st">"trunc_r"</span>] <span class="ot">&lt;-</span> <span class="dv">240</span></span>
<span id="cb30-183"><a href="#cb30-183" aria-hidden="true" tabindex="-1"></a>trunc_fr[<span class="fu">which</span>(loci <span class="sc">==</span> <span class="st">"tmem87a"</span>), <span class="st">"trunc_r"</span>] <span class="ot">&lt;-</span> <span class="dv">160</span></span>
<span id="cb30-184"><a href="#cb30-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-185"><a href="#cb30-185" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse data.frame with locus-specific truncation lengths</span></span>
<span id="cb30-186"><a href="#cb30-186" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(trunc_fr, <span class="dv">3</span>))</span>
<span id="cb30-187"><a href="#cb30-187" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-188"><a href="#cb30-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-189"><a href="#cb30-189" aria-hidden="true" tabindex="-1"></a>Truncate reads according to locus-specific truncation lengths:</span>
<span id="cb30-190"><a href="#cb30-190" aria-hidden="true" tabindex="-1"></a><span class="in">```{r declare_tr_dir}</span></span>
<span id="cb30-191"><a href="#cb30-191" aria-hidden="true" tabindex="-1"></a>tr_dir <span class="ot">&lt;-</span> <span class="st">"data/intermediate/truncated"</span></span>
<span id="cb30-192"><a href="#cb30-192" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-193"><a href="#cb30-193" aria-hidden="true" tabindex="-1"></a><span class="in">```{r truncate}</span></span>
<span id="cb30-194"><a href="#cb30-194" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb30-195"><a href="#cb30-195" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb30-196"><a href="#cb30-196" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr eval_truncate</span></span>
<span id="cb30-197"><a href="#cb30-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-198"><a href="#cb30-198" aria-hidden="true" tabindex="-1"></a><span class="co"># truncate</span></span>
<span id="cb30-199"><a href="#cb30-199" aria-hidden="true" tabindex="-1"></a>trunc_out <span class="ot">&lt;-</span></span>
<span id="cb30-200"><a href="#cb30-200" aria-hidden="true" tabindex="-1"></a>  <span class="fu">trunc_amp</span>(<span class="at">in_dir =</span> demult,</span>
<span id="cb30-201"><a href="#cb30-201" aria-hidden="true" tabindex="-1"></a>          <span class="at">fw_pattern =</span> <span class="st">"1.fastq.gz"</span>,</span>
<span id="cb30-202"><a href="#cb30-202" aria-hidden="true" tabindex="-1"></a>          <span class="at">rv_pattern =</span> <span class="st">"2.fastq.gz"</span>,</span>
<span id="cb30-203"><a href="#cb30-203" aria-hidden="true" tabindex="-1"></a>          <span class="at">trunc_fr =</span> trunc_fr,</span>
<span id="cb30-204"><a href="#cb30-204" aria-hidden="true" tabindex="-1"></a>          <span class="at">write_trun =</span> tr_dir,</span>
<span id="cb30-205"><a href="#cb30-205" aria-hidden="true" tabindex="-1"></a>          <span class="at">max_ee =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">5</span>),</span>
<span id="cb30-206"><a href="#cb30-206" aria-hidden="true" tabindex="-1"></a>          <span class="at">trunc_q =</span> <span class="dv">2</span>)</span>
<span id="cb30-207"><a href="#cb30-207" aria-hidden="true" tabindex="-1"></a><span class="co"># save reads in and out</span></span>
<span id="cb30-208"><a href="#cb30-208" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(trunc_out, tr_path)</span>
<span id="cb30-209"><a href="#cb30-209" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-210"><a href="#cb30-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-211"><a href="#cb30-211" aria-hidden="true" tabindex="-1"></a>The output from <span class="in">`trunc_amp()`</span> is a list of matrices with IN and OUT reads after truncation:</span>
<span id="cb30-212"><a href="#cb30-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-215"><a href="#cb30-215" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-216"><a href="#cb30-216" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb30-217"><a href="#cb30-217" aria-hidden="true" tabindex="-1"></a>trunc_out <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(tr_path)</span>
<span id="cb30-218"><a href="#cb30-218" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-221"><a href="#cb30-221" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-222"><a href="#cb30-222" aria-hidden="true" tabindex="-1"></a><span class="co"># see trunc_out</span></span>
<span id="cb30-223"><a href="#cb30-223" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(trunc_out[<span class="fu">seq_len</span>(<span class="dv">3</span>)], head, <span class="dv">3</span>)</span>
<span id="cb30-224"><a href="#cb30-224" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-225"><a href="#cb30-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-226"><a href="#cb30-226" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exploration of the parameter space</span></span>
<span id="cb30-227"><a href="#cb30-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-228"><a href="#cb30-228" aria-hidden="true" tabindex="-1"></a>The function <span class="in">`explore_dada()`</span> can be used to explore the effect of DADA2 parameters ("OMEGA_A", "BAND_SIZE", "pool") on the sensitivity on the variant calling.</span>
<span id="cb30-229"><a href="#cb30-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-230"><a href="#cb30-230" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>*omega_a*: threshold for variants to be significant overabundant _log(-log(birth_pval))_ (see <span class="co">[</span><span class="ot">Rosen et al. 2012</span><span class="co">](https://doi.org/10.1186/1471-2105-13-283.)</span>).</span>
<span id="cb30-231"><a href="#cb30-231" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>*band_size*: positive numbers set a band size in Needleman-Wunsch alignments. In this context, ends free alignment is performed. A value of zero turns off banding, triggering full Needleman-Wunsch alignments, in which gapless alignment is performed (see <span class="co">[</span><span class="ot">issue</span><span class="co">](https://github.com/benjjneb/dada2/issues/1982)</span>).</span>
<span id="cb30-232"><a href="#cb30-232" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>*pool*: calling variants pooling samples can increase sensitivity (see <span class="co">[</span><span class="ot">dicussion</span><span class="co">](https://benjjneb.github.io/dada2/pseudo.html)</span>).</span>
<span id="cb30-233"><a href="#cb30-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-234"><a href="#cb30-234" aria-hidden="true" tabindex="-1"></a>The returned plots can be used to guide the election of the best *OMEGA_A* in `variant_call()`, or frequency (*maf*) and abundance thresholds (*ad*) for filtering variants.</span>
<span id="cb30-235"><a href="#cb30-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-236"><a href="#cb30-236" aria-hidden="true" tabindex="-1"></a>Explore variants:</span>
<span id="cb30-239"><a href="#cb30-239" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-240"><a href="#cb30-240" aria-hidden="true" tabindex="-1"></a><span class="co"># declare candidate OMEGA_A to use in variant_call()</span></span>
<span id="cb30-241"><a href="#cb30-241" aria-hidden="true" tabindex="-1"></a>candidate_omega_a <span class="ot">&lt;-</span> <span class="dv">10</span><span class="sc">^-</span><span class="dv">2</span></span>
<span id="cb30-242"><a href="#cb30-242" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-243"><a href="#cb30-243" aria-hidden="true" tabindex="-1"></a><span class="in">```{r explore_dada}</span></span>
<span id="cb30-244"><a href="#cb30-244" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr eval_x</span></span>
<span id="cb30-245"><a href="#cb30-245" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb30-246"><a href="#cb30-246" aria-hidden="true" tabindex="-1"></a><span class="co"># paths to forward and reverse truncated reads</span></span>
<span id="cb30-247"><a href="#cb30-247" aria-hidden="true" tabindex="-1"></a>ftrun <span class="ot">&lt;-</span></span>
<span id="cb30-248"><a href="#cb30-248" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>(tr_dir, <span class="at">pattern =</span> <span class="st">"_F_"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb30-249"><a href="#cb30-249" aria-hidden="true" tabindex="-1"></a>rtrun <span class="ot">&lt;-</span></span>
<span id="cb30-250"><a href="#cb30-250" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>(tr_dir, <span class="at">pattern =</span> <span class="st">"_R_"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb30-251"><a href="#cb30-251" aria-hidden="true" tabindex="-1"></a><span class="co"># candidate omegavalue to annotate vline in plots.</span></span>
<span id="cb30-252"><a href="#cb30-252" aria-hidden="true" tabindex="-1"></a>v_line <span class="ot">&lt;-</span> <span class="fu">log</span>(<span class="sc">-</span><span class="fu">log</span>(candidate_omega_a))</span>
<span id="cb30-253"><a href="#cb30-253" aria-hidden="true" tabindex="-1"></a><span class="co"># run explore_dada() with band_size = 0, non pooling and omega_a = 0.9</span></span>
<span id="cb30-254"><a href="#cb30-254" aria-hidden="true" tabindex="-1"></a><span class="co"># forward</span></span>
<span id="cb30-255"><a href="#cb30-255" aria-hidden="true" tabindex="-1"></a>F_ind_0 <span class="ot">&lt;-</span></span>
<span id="cb30-256"><a href="#cb30-256" aria-hidden="true" tabindex="-1"></a>  <span class="fu">explore_dada</span>(ftrun, <span class="at">band_size =</span> <span class="dv">0</span>, <span class="at">vline =</span> v_line, <span class="at">hline_fr =</span> <span class="fl">0.1</span>)</span>
<span id="cb30-257"><a href="#cb30-257" aria-hidden="true" tabindex="-1"></a><span class="co"># reverse</span></span>
<span id="cb30-258"><a href="#cb30-258" aria-hidden="true" tabindex="-1"></a>R_ind_0 <span class="ot">&lt;-</span></span>
<span id="cb30-259"><a href="#cb30-259" aria-hidden="true" tabindex="-1"></a>  <span class="fu">explore_dada</span>(rtrun, <span class="at">band_size =</span> <span class="dv">0</span>, <span class="at">vline =</span> v_line, <span class="at">hline_fr =</span> <span class="fl">0.1</span>)</span>
<span id="cb30-260"><a href="#cb30-260" aria-hidden="true" tabindex="-1"></a><span class="co"># save results</span></span>
<span id="cb30-261"><a href="#cb30-261" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(<span class="fu">list</span>(<span class="at">F_ind_0 =</span> F_ind_0, <span class="at">R_ind_0 =</span> R_ind_0), x_out_path)</span>
<span id="cb30-262"><a href="#cb30-262" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-263"><a href="#cb30-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-264"><a href="#cb30-264" aria-hidden="true" tabindex="-1"></a>Exploration of DADA2 clustering for forward (A, C) and reverse (B, D) reads. The Y-axis represents the frequency of the variant in each locus and sample. The _log(-log(birth_pval))_ transformation in the X-axis is related to the p-value of a variant being significantly overabundant. Larger x-values represent likely true variants. For representation purposes _birth_pval_ of 0 (thus negative infinite), are converted to 10. Points are color-coded according to the variant rank in read abundance for its given locus and sample. For diploid individuals, <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color: green;"</span><span class="kw">&gt;</span>green<span class="kw">&lt;/span&gt;</span> are likely true variants and <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color: red;"</span><span class="kw">&gt;</span>red<span class="kw">&lt;/span&gt;</span> are likely false variants. <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color: grey;"</span><span class="kw">&gt;</span>Grey<span class="kw">&lt;/span&gt;</span> dashed lines are thresholds used for <span class="in">`variant_call()`</span>:</span>
<span id="cb30-265"><a href="#cb30-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-266"><a href="#cb30-266" aria-hidden="true" tabindex="-1"></a><span class="in">```{r x_attach_if_not_run}</span></span>
<span id="cb30-267"><a href="#cb30-267" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb30-268"><a href="#cb30-268" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb30-269"><a href="#cb30-269" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span>(<span class="fu">exists</span>(<span class="st">"F_ind_0"</span>) <span class="sc">||</span> <span class="fu">exists</span>(<span class="st">"R_ind_0"</span>))) {</span>
<span id="cb30-270"><a href="#cb30-270" aria-hidden="true" tabindex="-1"></a>    x_dada <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(x_out_path)</span>
<span id="cb30-271"><a href="#cb30-271" aria-hidden="true" tabindex="-1"></a>    <span class="fu">attach</span>(x_dada)</span>
<span id="cb30-272"><a href="#cb30-272" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb30-273"><a href="#cb30-273" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-274"><a href="#cb30-274" aria-hidden="true" tabindex="-1"></a><span class="in">```{r plot_explore_dada}</span></span>
<span id="cb30-275"><a href="#cb30-275" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-explore_dada</span></span>
<span id="cb30-276"><a href="#cb30-276" aria-hidden="true" tabindex="-1"></a><span class="co">#| warnings: false</span></span>
<span id="cb30-277"><a href="#cb30-277" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Exploration of DADA2 variants"</span></span>
<span id="cb30-278"><a href="#cb30-278" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb30-279"><a href="#cb30-279" aria-hidden="true" tabindex="-1"></a>ppool <span class="ot">&lt;-</span></span>
<span id="cb30-280"><a href="#cb30-280" aria-hidden="true" tabindex="-1"></a>  (F_ind_0<span class="sc">$</span>p1 <span class="sc">|</span> R_ind_0<span class="sc">$</span>p1) <span class="sc">/</span> (F_ind_0<span class="sc">$</span>p2 <span class="sc">|</span> R_ind_0<span class="sc">$</span>p2) <span class="sc">+</span></span>
<span id="cb30-281"><a href="#cb30-281" aria-hidden="true" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="st">"Dada F/R pool = F, omega_a 0.9, band_size = 0"</span>,</span>
<span id="cb30-282"><a href="#cb30-282" aria-hidden="true" tabindex="-1"></a>                             <span class="at">tag_levels =</span> <span class="st">"A"</span>)</span>
<span id="cb30-283"><a href="#cb30-283" aria-hidden="true" tabindex="-1"></a><span class="co"># save plot with combined loci</span></span>
<span id="cb30-284"><a href="#cb30-284" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"output/explore_dada.pdf"</span>, ppool, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">5</span>)</span>
<span id="cb30-285"><a href="#cb30-285" aria-hidden="true" tabindex="-1"></a><span class="co"># print plot</span></span>
<span id="cb30-286"><a href="#cb30-286" aria-hidden="true" tabindex="-1"></a>ppool</span>
<span id="cb30-287"><a href="#cb30-287" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-288"><a href="#cb30-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-289"><a href="#cb30-289" aria-hidden="true" tabindex="-1"></a>After the exploration variants in @fig-explore_dada, it seems an *OMEGA_A* = <span class="in">`r candidate_omega_a`</span>, implying a cutoff of <span class="in">`r log(-log(candidate_omega_a))`</span> in the X-axis and a frequency threshold (Y-axis) of 0.1, excludes most <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color: red;"</span><span class="kw">&gt;</span>artifacts<span class="kw">&lt;/span&gt;</span> while maximizing <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color: red;"</span><span class="kw">&gt;</span>true positives<span class="kw">&lt;/span&gt;</span>.</span>
<span id="cb30-290"><a href="#cb30-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-291"><a href="#cb30-291" aria-hidden="true" tabindex="-1"></a>The results can also be explored **per-locus**. For instance, @fig-explore_dada B can be expanded per locus as follows:</span>
<span id="cb30-294"><a href="#cb30-294" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-295"><a href="#cb30-295" aria-hidden="true" tabindex="-1"></a><span class="co">#| collapse: true</span></span>
<span id="cb30-296"><a href="#cb30-296" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-explore_dada_loci</span></span>
<span id="cb30-297"><a href="#cb30-297" aria-hidden="true" tabindex="-1"></a><span class="co">#| warnings: false</span></span>
<span id="cb30-298"><a href="#cb30-298" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Exploration of DADA2 variants per locus for R reads"</span></span>
<span id="cb30-299"><a href="#cb30-299" aria-hidden="true" tabindex="-1"></a><span class="co"># list of plots per locus</span></span>
<span id="cb30-300"><a href="#cb30-300" aria-hidden="true" tabindex="-1"></a>lplots <span class="ot">&lt;-</span></span>
<span id="cb30-301"><a href="#cb30-301" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(<span class="at">loci_f_ind_0_logp =</span> F_ind_0<span class="sc">$</span>p3,</span>
<span id="cb30-302"><a href="#cb30-302" aria-hidden="true" tabindex="-1"></a>       <span class="at">loci_r_ind_0_logp =</span> R_ind_0<span class="sc">$</span>p3,</span>
<span id="cb30-303"><a href="#cb30-303" aria-hidden="true" tabindex="-1"></a>       <span class="at">loci_f_ind_0_abun =</span> F_ind_0<span class="sc">$</span>p4,</span>
<span id="cb30-304"><a href="#cb30-304" aria-hidden="true" tabindex="-1"></a>       <span class="at">loci_r_ind_0_abun =</span> R_ind_0<span class="sc">$</span>p4)</span>
<span id="cb30-305"><a href="#cb30-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-306"><a href="#cb30-306" aria-hidden="true" tabindex="-1"></a><span class="co"># save plots per locus</span></span>
<span id="cb30-307"><a href="#cb30-307" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(<span class="fu">seq_along</span>(lplots), <span class="cf">function</span>(x) {</span>
<span id="cb30-308"><a href="#cb30-308" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">paste0</span>(<span class="st">"output/"</span>, <span class="fu">names</span>(lplots)[x], <span class="st">".pdf"</span>),</span>
<span id="cb30-309"><a href="#cb30-309" aria-hidden="true" tabindex="-1"></a>         lplots[[x]], <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">6</span>)</span>
<span id="cb30-310"><a href="#cb30-310" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb30-311"><a href="#cb30-311" aria-hidden="true" tabindex="-1"></a>lplots<span class="sc">$</span>loci_r_ind_0_logp</span>
<span id="cb30-312"><a href="#cb30-312" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-313"><a href="#cb30-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-314"><a href="#cb30-314" aria-hidden="true" tabindex="-1"></a><span class="fu">## Variant and genotype calling</span></span>
<span id="cb30-315"><a href="#cb30-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-316"><a href="#cb30-316" aria-hidden="true" tabindex="-1"></a>Variant calling is run using *OMEGA_A* = <span class="in">`r candidate_omega_a`</span>, and under different parameters:</span>
<span id="cb30-317"><a href="#cb30-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-318"><a href="#cb30-318" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>_band_size_: 0, 16</span>
<span id="cb30-319"><a href="#cb30-319" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>_pool_: TRUE, FALSE</span>
<span id="cb30-320"><a href="#cb30-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-321"><a href="#cb30-321" aria-hidden="true" tabindex="-1"></a><span class="in">```{r variant_call}</span></span>
<span id="cb30-322"><a href="#cb30-322" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb30-323"><a href="#cb30-323" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr eval_variant_call</span></span>
<span id="cb30-324"><a href="#cb30-324" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb30-325"><a href="#cb30-325" aria-hidden="true" tabindex="-1"></a><span class="co"># list of parameters to run variant_call() with 4 different set of values:</span></span>
<span id="cb30-326"><a href="#cb30-326" aria-hidden="true" tabindex="-1"></a>variant_call_params <span class="ot">&lt;-</span></span>
<span id="cb30-327"><a href="#cb30-327" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(</span>
<span id="cb30-328"><a href="#cb30-328" aria-hidden="true" tabindex="-1"></a>    <span class="at">ind_bs16 =</span> <span class="fu">c</span>(<span class="cn">FALSE</span>, <span class="dv">16</span>), <span class="co"># pool = F, default band size == 16</span></span>
<span id="cb30-329"><a href="#cb30-329" aria-hidden="true" tabindex="-1"></a>    <span class="at">ind_bs0 =</span> <span class="fu">c</span>(<span class="cn">FALSE</span>, <span class="dv">0</span>), <span class="co"># pool = F, default band size == 0</span></span>
<span id="cb30-330"><a href="#cb30-330" aria-hidden="true" tabindex="-1"></a>    <span class="at">pool_bs16 =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="dv">16</span>), <span class="co"># pool = T, default band size == 16</span></span>
<span id="cb30-331"><a href="#cb30-331" aria-hidden="true" tabindex="-1"></a>    <span class="at">pool_bs0 =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="dv">0</span>) <span class="co"># pool = T, default band size == 0</span></span>
<span id="cb30-332"><a href="#cb30-332" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb30-333"><a href="#cb30-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-334"><a href="#cb30-334" aria-hidden="true" tabindex="-1"></a>variants_x <span class="ot">&lt;-</span></span>
<span id="cb30-335"><a href="#cb30-335" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(variant_call_params, <span class="cf">function</span>(x) {</span>
<span id="cb30-336"><a href="#cb30-336" aria-hidden="true" tabindex="-1"></a>    <span class="fu">variant_call</span>(<span class="at">in_folder =</span> tr_dir,</span>
<span id="cb30-337"><a href="#cb30-337" aria-hidden="true" tabindex="-1"></a>                 <span class="at">rv_pattern =</span> <span class="st">"R_filt.fastq.gz"</span>,</span>
<span id="cb30-338"><a href="#cb30-338" aria-hidden="true" tabindex="-1"></a>                 <span class="at">c_unmerged =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-339"><a href="#cb30-339" aria-hidden="true" tabindex="-1"></a>                 <span class="at">multithread =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-340"><a href="#cb30-340" aria-hidden="true" tabindex="-1"></a>                 <span class="at">pool =</span> <span class="fu">as.logical</span>(x[<span class="dv">1</span>]),</span>
<span id="cb30-341"><a href="#cb30-341" aria-hidden="true" tabindex="-1"></a>                 <span class="at">omega_a_f =</span> candidate_omega_a,</span>
<span id="cb30-342"><a href="#cb30-342" aria-hidden="true" tabindex="-1"></a>                 <span class="at">omega_a_r =</span> candidate_omega_a,</span>
<span id="cb30-343"><a href="#cb30-343" aria-hidden="true" tabindex="-1"></a>                 <span class="at">band_size =</span> x[<span class="dv">2</span>],</span>
<span id="cb30-344"><a href="#cb30-344" aria-hidden="true" tabindex="-1"></a>                 <span class="at">ad =</span> <span class="dv">10</span>,</span>
<span id="cb30-345"><a href="#cb30-345" aria-hidden="true" tabindex="-1"></a>                 <span class="at">maf =</span> <span class="fl">0.1</span>)</span>
<span id="cb30-346"><a href="#cb30-346" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb30-347"><a href="#cb30-347" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(variants_x, <span class="st">"output/variants_x.rds"</span>)</span>
<span id="cb30-348"><a href="#cb30-348" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-349"><a href="#cb30-349" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-352"><a href="#cb30-352" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-353"><a href="#cb30-353" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb30-354"><a href="#cb30-354" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"variants_x"</span>))</span>
<span id="cb30-355"><a href="#cb30-355" aria-hidden="true" tabindex="-1"></a>  variants_x <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/variants_x.rds"</span>)</span>
<span id="cb30-356"><a href="#cb30-356" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-357"><a href="#cb30-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-358"><a href="#cb30-358" aria-hidden="true" tabindex="-1"></a>Variants are used to genotype individuals with defaults <span class="in">`ploidy = 2`</span> and <span class="in">`ADt = 10`</span>:</span>
<span id="cb30-359"><a href="#cb30-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-362"><a href="#cb30-362" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-363"><a href="#cb30-363" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb30-364"><a href="#cb30-364" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr eval_genotype</span></span>
<span id="cb30-365"><a href="#cb30-365" aria-hidden="true" tabindex="-1"></a>genotypes_x <span class="ot">&lt;-</span></span>
<span id="cb30-366"><a href="#cb30-366" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(variants_x, genotype)</span>
<span id="cb30-367"><a href="#cb30-367" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(genotypes_x, <span class="at">file =</span> <span class="st">"output/genotypes_x.rds"</span>)</span>
<span id="cb30-368"><a href="#cb30-368" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-369"><a href="#cb30-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-372"><a href="#cb30-372" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-373"><a href="#cb30-373" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb30-374"><a href="#cb30-374" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"genotypes_x"</span>))</span>
<span id="cb30-375"><a href="#cb30-375" aria-hidden="true" tabindex="-1"></a>  genotypes_x <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/genotypes_x.rds"</span>)</span>
<span id="cb30-376"><a href="#cb30-376" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-377"><a href="#cb30-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-378"><a href="#cb30-378" aria-hidden="true" tabindex="-1"></a><span class="fu">## Tidy data</span></span>
<span id="cb30-379"><a href="#cb30-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-380"><a href="#cb30-380" aria-hidden="true" tabindex="-1"></a>A strong point from _tidyGenR_ is that variants and genotypes from <span class="in">`variant_call()`</span> and <span class="in">`genotype()`</span> are returned in <span class="in">`tidy`</span> format: one row per observation and variables in columns. Lets have a look at the data structure.</span>
<span id="cb30-381"><a href="#cb30-381" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-382"><a href="#cb30-382" aria-hidden="true" tabindex="-1"></a><span class="in">```{r glimpse_tidy_variants}</span></span>
<span id="cb30-383"><a href="#cb30-383" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse tidy variants</span></span>
<span id="cb30-384"><a href="#cb30-384" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(variants_x<span class="sc">$</span>ind_bs0))</span>
<span id="cb30-385"><a href="#cb30-385" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-386"><a href="#cb30-386" aria-hidden="true" tabindex="-1"></a><span class="in">```{r glimpse_tidy_genotypes}</span></span>
<span id="cb30-387"><a href="#cb30-387" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse tidy genotypes</span></span>
<span id="cb30-388"><a href="#cb30-388" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(genotypes_x<span class="sc">$</span>ind_bs0))</span>
<span id="cb30-389"><a href="#cb30-389" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-390"><a href="#cb30-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-391"><a href="#cb30-391" aria-hidden="true" tabindex="-1"></a><span class="fu">## *tidyGenR* vs AmpliSAS</span></span>
<span id="cb30-392"><a href="#cb30-392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-393"><a href="#cb30-393" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">AmpliSAS</span><span class="co">](https://doi.org/10.1111/1755-0998.12453)</span> is a software written in PERL with similar characteristics to _tidyGenR_. To compare their performance we run AmpliSAS in a docker container to genotype our raw sequences. The steps are detailed <span class="co">[</span><span class="ot">here</span><span class="co">](code/amplisas)</span>.</span>
<span id="cb30-394"><a href="#cb30-394" aria-hidden="true" tabindex="-1"></a>AmpliSAS returns results in a multisheet EXCEL and in plain text files, one per locus. The function <span class="in">`amplisas2tidy()`</span> permits to read plain text results from AmpliSAS into tidy variants.</span>
<span id="cb30-395"><a href="#cb30-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-398"><a href="#cb30-398" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-399"><a href="#cb30-399" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-400"><a href="#cb30-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-401"><a href="#cb30-401" aria-hidden="true" tabindex="-1"></a><span class="fu"># Integration of alleles with publised data</span></span>
<span id="cb30-402"><a href="#cb30-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-405"><a href="#cb30-405" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-406"><a href="#cb30-406" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-407"><a href="#cb30-407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-408"><a href="#cb30-408" aria-hidden="true" tabindex="-1"></a><span class="fu"># session info</span></span>
<span id="cb30-409"><a href="#cb30-409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-410"><a href="#cb30-410" aria-hidden="true" tabindex="-1"></a>Details on the sessionInfo() are <span class="co">[</span><span class="ot">here</span><span class="co">](session_info.txt)</span>:</span>
<span id="cb30-411"><a href="#cb30-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-414"><a href="#cb30-414" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-415"><a href="#cb30-415" aria-hidden="true" tabindex="-1"></a><span class="fu">writeLines</span>(<span class="fu">capture.output</span>(<span class="fu">sessionInfo</span>()), <span class="at">con =</span> <span class="st">"session_info.txt"</span>)</span>
<span id="cb30-416"><a href="#cb30-416" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>