---
author: Miguel Camacho SÃ¡nchez
title: "AmpliSAS analysis"
editor: visual
theme: minty
format: html
editor_options: 
  markdown: 
    wrap: 80
---
#!/bin/bash

# navigate to directory with F and R reads.
cd data/intermediate/amplisas/raw-barcoded
#merge F and R reads
docker run -v $PWD:/workdir --rm sixthresearcher/amplisat ampliMERGE.pl \
    -i 1.fastq.gz 2.fastq.gz \
    -o merged_reads > amplisas.log

# ampliCLEAN.pl seems to collapse due to the large file size. For that reason I split the fastq in to 10 parts:
seqkit split -p 10 merged_reads.fq >> amplisas.log
rm merged_reads.fq
rm 1.fastq.gz
rm 2.fastq.gz
#clean merged reads. It can also clean demultiplexed reads.
# If the reads have been already demultiplexed into separate files (one file per sample), they can be packed into a single .zip or .tar.gz file and use it as input
# if -d amplicon-data.csv  is included the processing time increases x30 because it needs to demultiplex.
cp ../amplicon-data.csv .
for partx in merged_reads.fq.split/*fq
do
  outname=$(echo $partx | sed 's|^.*\(part.*\).fq|\1|')
  docker run -v $PWD:/workdir --rm sixthresearcher/amplisat ampliCLEAN.pl \
      -i $partx \
      -mqual 17 \
      -min 200 \
      -d amplicon-data.csv \
      -o filtered_reads_$outname >> amplisas.log
done

rm -rf merged_reads.fq.split
# concatenate split sequences
cat filtered_reads_part*fq > filtered_reads.fq
rm *part_0*fq

# edit params to mimic easyampr parameters MAF = .1 and AD = 10.
cp amplicon-data.csv amplicon-data_ampliSAS.csv

echo ">param,amplicon,value" >> amplicon-data_ampliSAS.csv
echo "min_amplicon_seq_depth,all,10" >> amplicon-data_ampliSAS.csv
echo "min_amplicon_seq_frequency,all,10" >> amplicon-data_ampliSAS.csv

# run amplisSAS
docker run -v $PWD:/workdir --rm sixthresearcher/amplisat ampliSAS.pl \
-i filtered_reads.fq \
-min 10 \
-t Illumina \
-d amplicon-data_ampliSAS.csv \
-thr 5 \
-o results_amplisas > amplisas.log

#remove intermediate files
rm filtered_reads.fq

###deprecated code
# there is no need to run amplicheck because I already checked it.
#fast analysis of amplicon sequences
# Description:
#   Fast pairwise comparison among higher depth sequences in amplicon sequencing (AS) data
#   Analyzes AS data and gives as output an Excel file where are shown the similarities among the high depth sequences
#   This results can be used to optimize the parameters for a further analysis with AmpliSAS
#
# Requires as input a FASTA or FASTQ file with sequences/reads and a CSV format file with primer/tag data information
# Example: perl a
# docker run -v $PWD:/workdir --rm sixthresearcher/amplisat ampliCHECK.pl \
#     -d amplicon-data.csv \
#     -i filtered_reads.fq \
#     -o results_amplicheck >> amplisas.log
# loci for which there where no reads:
# echo "\n\nloci for which there where no reads:" >> amplisas.log
# cat amplisas.log | grep 'de-multiplexed (0 reads' | cut -d"-" -f1 | uniq -c >> amplisas.log
