<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Miguel Camacho Sánchez">

<title>tidyGenR Benchmarking - Summary tables</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-fixed">


<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">tidyGenR Benchmarking</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html">
 <span class="menu-text">Benchmarking</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./amplisat_analysis.html">
 <span class="menu-text">AmpliSAS</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./summary-tables.html" aria-current="page">
 <span class="menu-text">Summary tables</span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#track-reads-throught-the-workflow" id="toc-track-reads-throught-the-workflow" class="nav-link active" data-scroll-target="#track-reads-throught-the-workflow">Track reads throught the workflow</a></li>
  <li><a href="#distribution-of-reads-across-samples-and-loci" id="toc-distribution-of-reads-across-samples-and-loci" class="nav-link" data-scroll-target="#distribution-of-reads-across-samples-and-loci">Distribution of reads across samples and loci</a></li>
  <li><a href="#dereplicated" id="toc-dereplicated" class="nav-link" data-scroll-target="#dereplicated">Dereplicated</a></li>
  <li><a href="#agglomerate-tables-in-an-excel" id="toc-agglomerate-tables-in-an-excel" class="nav-link" data-scroll-target="#agglomerate-tables-in-an-excel">Agglomerate tables in an EXCEL</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session Info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Summary tables</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Miguel Camacho Sánchez </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<p><em>tidyGenR</em> includes several functions to explore the sequencing data and create summary tables from the sequencing data created along the workflow: <code>reads_track()</code> and <code>reads_loci_samples()</code>, <code>dereplicate()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># attach packages</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyGenR)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(writexl)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># load tidy variants dataset</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>tidy_variants <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/variants_x.rds"</span>)<span class="sc">$</span>ind_bs0</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="track-reads-throught-the-workflow" class="level2">
<h2 class="anchored" data-anchor-id="track-reads-throught-the-workflow">Track reads throught the workflow</h2>
<p>The sequencing reads went through different steps in the workflow: <strong>raw</strong> -&gt; <strong>demultiplexed</strong> -&gt; <strong>truncated</strong> -&gt; <strong>variant_calling</strong> At each step some data can be lost, edited, or filtered. This leads to a decrease of the number of reads from the initial step to the last. It is useful to track the reads through the pipeline to detect potential steps that might have gone wrong or could be optimized. We tracked the reads from the initial 44 F/R raw FASTQ to the filtered variants:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create list with ordered elements:</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>l <span class="ot">&lt;-</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"raw"</span>           <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"data/raw"</span>,</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"1.fastq"</span>), <span class="co"># raw reads</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"demultiplexed"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"data/intermediate/demultiplexed"</span>,</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"1.fastq.gz"</span>),</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">"truncated"</span>     <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"data/intermediate/truncated"</span>,</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"F_filt.fastq.gz"</span>),</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">"supp_variants"</span> <span class="ot">=</span> tidy_variants</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># track reads</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>reads_tracked <span class="ot">&lt;-</span> <span class="fu">reads_track</span>(l)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(reads_tracked, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">sample</th>
<th style="text-align: right;">raw</th>
<th style="text-align: right;">demultiplexed</th>
<th style="text-align: right;">truncated</th>
<th style="text-align: right;">supp_variants</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">BOR1061</td>
<td style="text-align: right;">86716</td>
<td style="text-align: right;">76207</td>
<td style="text-align: right;">45939</td>
<td style="text-align: right;">45459</td>
</tr>
<tr class="even">
<td style="text-align: left;">BOR1063</td>
<td style="text-align: right;">83553</td>
<td style="text-align: right;">73917</td>
<td style="text-align: right;">44798</td>
<td style="text-align: right;">44345</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BOR1069</td>
<td style="text-align: right;">84444</td>
<td style="text-align: right;">72497</td>
<td style="text-align: right;">43782</td>
<td style="text-align: right;">43379</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="distribution-of-reads-across-samples-and-loci" class="level2">
<h2 class="anchored" data-anchor-id="distribution-of-reads-across-samples-and-loci">Distribution of reads across samples and loci</h2>
<p>The distribution of reads across samples and loci is not homogeneous. Indeed, it can be highly skewed towards some loci. Their distribution provides important information about potential biases and help to interpret the data.</p>
<p>We used <code>reads_loci_samples()</code> to get the depth of each locus and sample for demultiplexed reads, an was Demultiplexed reads can</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># retrieve depth of demultiplexed reads per locus</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>reads_demultiplx <span class="ot">&lt;-</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">reads_loci_samples</span>(<span class="st">"data/intermediate/demultiplexed"</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">pattern_fq =</span> <span class="st">"1.fastq.gz"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'all.variants' and 'var_id' are ignored when setting'path' to a folder.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># retrieve filtered variants reads per locus and sample.</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   loci are split with one variant per column when 'all.variants = T'.</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>reads_variants <span class="ot">&lt;-</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">reads_loci_samples</span>(tidy_variants,</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">all.variants =</span> <span class="cn">TRUE</span>,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>                     <span class="at">var_id =</span> <span class="st">"variant"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="dereplicated" class="level2">
<h2 class="anchored" data-anchor-id="dereplicated">Dereplicated</h2>
<p>De-replicated sequences are unique sequences in the read pool. Their count and distribution across FASTQ files can be retrieved and written to tables with <code>dereplicate()</code>. This step was performed with truncated FASTQ previous to variant calling with the aim to have tables to de-replicated frequencies to check for consistency against the variants called.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># files to dereplicate</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>fp <span class="ot">&lt;-</span> <span class="st">"data/intermediate/truncated"</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>fs <span class="ot">&lt;-</span> <span class="fu">list.files</span>(fp, <span class="at">pattern =</span> <span class="st">"fastq.gz"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>drep_xls <span class="ot">&lt;-</span> <span class="st">"output/dereplicated-truncated.xlsx"</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># run dereplicate</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>drep_tab <span class="ot">&lt;-</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dereplicate</span>(fs,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_sam_fr =</span> <span class="dv">5</span>, <span class="co"># a 5x is needed for a read count to be output in the table.</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_loc_fr =</span> <span class="fl">0.002</span>, <span class="co"># min fr of de-replicated sequence across all samples in the group.</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">"_([a-zA-Z0-9]*_[F|R])"</span>, <span class="co"># group by locus + F/R</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>            <span class="at">out_xlsx =</span> drep_xls)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>De-replicated sequences written to output/dereplicated-truncated.xlsx.
If the number of sheets in too large EXCEL might crash at opening.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(drep_tab[[<span class="dv">1</span>]], <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 16%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">sequence</th>
<th style="text-align: left;">md5</th>
<th style="text-align: right;">BOR1061_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1063_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1069_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1070_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1071_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1074_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1075_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1076_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1077_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1079_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1080_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1082_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1086_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1088_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1090_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1091_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1097_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1098_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1101_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1107_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1111_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1113_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1115_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1125_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1127_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1129_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1131_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1133_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1135_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1137_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1139_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1141_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1143_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1147_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1149_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1151_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR1155_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR614_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR615_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR616_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR618_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR619_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR621_abcg8_F_filt.fastq.gz</th>
<th style="text-align: right;">BOR626_abcg8_F_filt.fastq.gz</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">TTGCCCACCCTGTTCATCCATGGAGCAGAAGCCTGCCTGATGTCTCTCATCATTGGCTTCCTTTACTACGGCCACGCAGACAAGCCGCTCTCCTTCGTGGACATGGCAGCCCTCCTGTTCATGATAGGAGCGCTCATTCCTTTTAATGTCATTCTGGATGTCGTCTCCAAATGTGAGTGTCACCCGCCCTCCTCACCAGACATCGGGACAGTGGGACAGCCTCCCTGGGCACTGCACTGAGGCCAAGCTCTGTGCTTCCGCTGGTACCCA</td>
<td style="text-align: left;">79cdeae19796f127012f5a6d9d8fe5ba</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">332</td>
<td style="text-align: right;">261</td>
<td style="text-align: right;">281</td>
<td style="text-align: right;">261</td>
<td style="text-align: right;">461</td>
<td style="text-align: right;">329</td>
<td style="text-align: right;">197</td>
<td style="text-align: right;">441</td>
<td style="text-align: right;">222</td>
<td style="text-align: right;">555</td>
<td style="text-align: right;">196</td>
<td style="text-align: right;">374</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">227</td>
<td style="text-align: right;">490</td>
<td style="text-align: right;">633</td>
<td style="text-align: right;">347</td>
<td style="text-align: right;">556</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">388</td>
<td style="text-align: right;">341</td>
<td style="text-align: right;">618</td>
<td style="text-align: right;">318</td>
<td style="text-align: right;">665</td>
<td style="text-align: right;">323</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">330</td>
<td style="text-align: right;">238</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">577</td>
<td style="text-align: right;">264</td>
<td style="text-align: right;">454</td>
<td style="text-align: right;">494</td>
<td style="text-align: right;">569</td>
<td style="text-align: right;">1123</td>
<td style="text-align: right;">328</td>
<td style="text-align: right;">311</td>
<td style="text-align: right;">205</td>
</tr>
<tr class="even">
<td style="text-align: left;">TTGCCCACCCTGTTCATCCATGGAGCAGAAGCCTGCCTGATGTCTCTCATCATTGGCTTCCTTTACTACGGCCACGCAGATAAGCCGCTCTCCTTCGTGGACATGGCAGCCCTCCTGTTCATGATAGGAGCGCTCATTCCTTTTAATGTCATTCTGGATGTCGTCTCCAAATGTGAGTGTCACCCGCCCTCCTCACCAGACATCGGGACAGTGGGACAGCCTCCCTGGGCACTGCACTGAGGCCAAGCTCTGTGCTTCCGCTGGTACCCA</td>
<td style="text-align: left;">559fd72c1b84037c883fdd6053482736</td>
<td style="text-align: right;">680</td>
<td style="text-align: right;">657</td>
<td style="text-align: right;">306</td>
<td style="text-align: right;">264</td>
<td style="text-align: right;">229</td>
<td style="text-align: right;">273</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">198</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">251</td>
<td style="text-align: right;">251</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">333</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">605</td>
<td style="text-align: right;">264</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">405</td>
<td style="text-align: right;">388</td>
<td style="text-align: right;">390</td>
<td style="text-align: right;">365</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">298</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">336</td>
<td style="text-align: right;">332</td>
<td style="text-align: right;">132</td>
<td style="text-align: right;">236</td>
<td style="text-align: right;">204</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">207</td>
<td style="text-align: right;">484</td>
<td style="text-align: right;">493</td>
<td style="text-align: right;">489</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">325</td>
<td style="text-align: right;">259</td>
<td style="text-align: right;">197</td>
</tr>
<tr class="odd">
<td style="text-align: left;">TTGCCCACCCTGTTCATCCATGGAGCAGAAGCCTGCCTGATGTCTCTCATCATTGGCTTCCTTTACTACGGCCACGCAGACAAGCCGCTCTCCTTCGTGGACATGGCAGCCCTCCTGTTCATGATAGGAGCTCTCATTCCTTTTAATGTCATTCTGGATGTCGTCTCCAAATGTGAGTGTCACCCGCCCTCCTCACCAGACATCGGGACAGTGGGACAGCCTCCCTGGGCACTGCACTGAGGCCAAGCTCTGTGCTTCCGCTGGTACCCA</td>
<td style="text-align: left;">f9c9a80d4e1fa27cd4ce16e0cb9f6da9</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">334</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">227</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">210</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">244</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">305</td>
<td style="text-align: right;">601</td>
<td style="text-align: right;">390</td>
<td style="text-align: right;">368</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">332</td>
<td style="text-align: right;">202</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">223</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="agglomerate-tables-in-an-excel" class="level2">
<h2 class="anchored" data-anchor-id="agglomerate-tables-in-an-excel">Agglomerate tables in an EXCEL</h2>
<p>Tables above were combined in a unique EXCEL file that will be included in the manuscript as supplementary material.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ST1. tracked reads.</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co"># ST2. demultiplexed reads.</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ST3. reads supporting variants.</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>suppl_tables <span class="ot">&lt;-</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ST1. Tracked reads"</span> <span class="ot">=</span> reads_tracked,</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ST2. Demultiplexed- per locus"</span> <span class="ot">=</span> reads_demultiplx,</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ST3. Filtered variants"</span> <span class="ot">=</span> reads_variants</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="fu">write_xlsx</span>(suppl_tables,</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>           <span class="at">path =</span> <span class="st">"output/supplementary_tables.xlsx"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="session-info" class="level2">
<h2 class="anchored" data-anchor-id="session-info">Session Info</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.0 (2024-04-24)
Platform: x86_64-apple-darwin20
Running under: macOS 15.3.2

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] es_ES.UTF-8/es_ES.UTF-8/es_ES.UTF-8/C/es_ES.UTF-8/es_ES.UTF-8

time zone: Europe/Madrid
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] writexl_1.5.1  dplyr_1.1.4    tidyGenR_0.8.3

loaded via a namespace (and not attached):
 [1] SummarizedExperiment_1.34.0 gtable_0.3.6               
 [3] xfun_0.50                   hwriter_1.3.2.1            
 [5] ggplot2_3.5.1               htmlwidgets_1.6.4          
 [7] latticeExtra_0.6-30         Biobase_2.64.0             
 [9] lattice_0.22-6              tzdb_0.4.0                 
[11] vctrs_0.6.5                 tools_4.4.0                
[13] bitops_1.0-9                generics_0.1.3             
[15] stats4_4.4.0                parallel_4.4.0             
[17] tibble_3.2.1                pkgconfig_2.0.3            
[19] Matrix_1.7-2                RColorBrewer_1.1-3         
[21] S4Vectors_0.42.1            RcppParallel_5.1.10        
[23] lifecycle_1.0.4             GenomeInfoDbData_1.2.12    
[25] stringr_1.5.1               deldir_2.0-4               
[27] compiler_4.4.0              farver_2.1.2               
[29] Rsamtools_2.20.0            Biostrings_2.72.1          
[31] munsell_0.5.1               dada2_1.32.0               
[33] codetools_0.2-20            DECIPHER_3.0.0             
[35] GenomeInfoDb_1.40.1         htmltools_0.5.8.1          
[37] yaml_2.3.10                 tidyr_1.3.1                
[39] pillar_1.10.1               crayon_1.5.3               
[41] BiocParallel_1.38.0         DelayedArray_0.30.1        
[43] ShortRead_1.62.0            abind_1.4-8                
[45] tidyselect_1.2.1            digest_0.6.37              
[47] stringi_1.8.4               purrr_1.0.4                
[49] reshape2_1.4.4              fastmap_1.2.0              
[51] grid_4.4.0                  colorspace_2.1-1           
[53] cli_3.6.4                   SparseArray_1.4.8          
[55] magrittr_2.0.3              patchwork_1.3.0            
[57] S4Arrays_1.4.1              withr_3.0.2                
[59] readr_2.1.5                 scales_1.3.0               
[61] UCSC.utils_1.0.0            rmarkdown_2.29             
[63] pwalign_1.0.0               XVector_0.44.0             
[65] httr_1.4.7                  jpeg_0.1-10                
[67] matrixStats_1.5.0           interp_1.1-6               
[69] hms_1.1.3                   png_0.1-8                  
[71] evaluate_1.0.3              knitr_1.49                 
[73] GenomicRanges_1.56.2        IRanges_2.38.1             
[75] rlang_1.1.5                 Rcpp_1.0.14                
[77] glue_1.8.0                  DBI_1.2.3                  
[79] BiocGenerics_0.50.0         jsonlite_1.8.9             
[81] plyr_1.8.9                  R6_2.6.1                   
[83] MatrixGenerics_1.16.0       GenomicAlignments_1.40.0   
[85] zlibbioc_1.50.0            </code></pre>
</div>
</div>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb12" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> Miguel Camacho Sánchez</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Summary tables"</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>*tidyGenR* includes several functions to explore the sequencing data and create summary tables from the sequencing data created along the workflow: <span class="in">`reads_track()`</span> and <span class="in">`reads_loci_samples()`</span>, <span class="in">`dereplicate()`</span>.</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="in">```{r attach packages}</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="co"># attach packages</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyGenR)</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(writexl)</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a><span class="co"># load tidy variants dataset</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>tidy_variants <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/variants_x.rds"</span>)<span class="sc">$</span>ind_bs0</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a><span class="fu">## Track reads throught the workflow</span></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>The sequencing reads went through different steps in the workflow:</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>**raw** -&gt; **demultiplexed** -&gt; **truncated** -&gt; **variant_calling**</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>At each step some data can be lost, edited, or filtered. This leads to a decrease of the number of reads from the initial step to the last. It is useful to track the reads through the pipeline to detect potential steps that might have gone wrong or could be optimized.</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>We tracked the reads from the initial 44 F/R raw FASTQ to the filtered variants:</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a><span class="in">```{r track reads}</span></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a><span class="co"># create list with ordered elements:</span></span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>l <span class="ot">&lt;-</span></span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(</span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a>  <span class="st">"raw"</span>           <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"data/raw"</span>,</span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"1.fastq"</span>), <span class="co"># raw reads</span></span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>  <span class="st">"demultiplexed"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"data/intermediate/demultiplexed"</span>,</span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"1.fastq.gz"</span>),</span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a>  <span class="st">"truncated"</span>     <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"data/intermediate/truncated"</span>,</span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"F_filt.fastq.gz"</span>),</span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a>  <span class="st">"supp_variants"</span> <span class="ot">=</span> tidy_variants</span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a><span class="co"># track reads</span></span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a>reads_tracked <span class="ot">&lt;-</span> <span class="fu">reads_track</span>(l)</span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse</span></span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(reads_tracked, <span class="dv">3</span>))</span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb12-45"><a href="#cb12-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-46"><a href="#cb12-46" aria-hidden="true" tabindex="-1"></a><span class="fu">## Distribution of reads across samples and loci</span></span>
<span id="cb12-47"><a href="#cb12-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-48"><a href="#cb12-48" aria-hidden="true" tabindex="-1"></a>The distribution of reads across samples and loci is not homogeneous. Indeed, it can be highly skewed towards some loci. Their distribution provides important information about potential biases and help to interpret the data.</span>
<span id="cb12-49"><a href="#cb12-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-50"><a href="#cb12-50" aria-hidden="true" tabindex="-1"></a>We used <span class="in">`reads_loci_samples()`</span> to get the depth of each locus and sample for demultiplexed reads, an was Demultiplexed reads can</span>
<span id="cb12-51"><a href="#cb12-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-54"><a href="#cb12-54" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb12-55"><a href="#cb12-55" aria-hidden="true" tabindex="-1"></a><span class="co"># retrieve depth of demultiplexed reads per locus</span></span>
<span id="cb12-56"><a href="#cb12-56" aria-hidden="true" tabindex="-1"></a>reads_demultiplx <span class="ot">&lt;-</span></span>
<span id="cb12-57"><a href="#cb12-57" aria-hidden="true" tabindex="-1"></a>  <span class="fu">reads_loci_samples</span>(<span class="st">"data/intermediate/demultiplexed"</span>,</span>
<span id="cb12-58"><a href="#cb12-58" aria-hidden="true" tabindex="-1"></a>                     <span class="at">pattern_fq =</span> <span class="st">"1.fastq.gz"</span>)</span>
<span id="cb12-59"><a href="#cb12-59" aria-hidden="true" tabindex="-1"></a><span class="co"># retrieve filtered variants reads per locus and sample.</span></span>
<span id="cb12-60"><a href="#cb12-60" aria-hidden="true" tabindex="-1"></a><span class="co">#   loci are split with one variant per column when 'all.variants = T'.</span></span>
<span id="cb12-61"><a href="#cb12-61" aria-hidden="true" tabindex="-1"></a>reads_variants <span class="ot">&lt;-</span></span>
<span id="cb12-62"><a href="#cb12-62" aria-hidden="true" tabindex="-1"></a>  <span class="fu">reads_loci_samples</span>(tidy_variants,</span>
<span id="cb12-63"><a href="#cb12-63" aria-hidden="true" tabindex="-1"></a>                     <span class="at">all.variants =</span> <span class="cn">TRUE</span>,</span>
<span id="cb12-64"><a href="#cb12-64" aria-hidden="true" tabindex="-1"></a>                     <span class="at">var_id =</span> <span class="st">"variant"</span>)</span>
<span id="cb12-65"><a href="#cb12-65" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb12-66"><a href="#cb12-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-67"><a href="#cb12-67" aria-hidden="true" tabindex="-1"></a><span class="fu">## Dereplicated</span></span>
<span id="cb12-68"><a href="#cb12-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-69"><a href="#cb12-69" aria-hidden="true" tabindex="-1"></a>De-replicated sequences are unique sequences in the read pool. Their count and distribution across FASTQ files can be retrieved and written to tables with <span class="in">`dereplicate()`</span>. This step was performed with truncated FASTQ previous to variant calling with the aim to have tables to de-replicated frequencies to check for consistency against the variants called.</span>
<span id="cb12-70"><a href="#cb12-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-71"><a href="#cb12-71" aria-hidden="true" tabindex="-1"></a><span class="in">```{r deprelicate}</span></span>
<span id="cb12-72"><a href="#cb12-72" aria-hidden="true" tabindex="-1"></a><span class="co"># files to dereplicate</span></span>
<span id="cb12-73"><a href="#cb12-73" aria-hidden="true" tabindex="-1"></a>fp <span class="ot">&lt;-</span> <span class="st">"data/intermediate/truncated"</span></span>
<span id="cb12-74"><a href="#cb12-74" aria-hidden="true" tabindex="-1"></a>fs <span class="ot">&lt;-</span> <span class="fu">list.files</span>(fp, <span class="at">pattern =</span> <span class="st">"fastq.gz"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb12-75"><a href="#cb12-75" aria-hidden="true" tabindex="-1"></a>drep_xls <span class="ot">&lt;-</span> <span class="st">"output/dereplicated-truncated.xlsx"</span></span>
<span id="cb12-76"><a href="#cb12-76" aria-hidden="true" tabindex="-1"></a><span class="co"># run dereplicate</span></span>
<span id="cb12-77"><a href="#cb12-77" aria-hidden="true" tabindex="-1"></a>drep_tab <span class="ot">&lt;-</span></span>
<span id="cb12-78"><a href="#cb12-78" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dereplicate</span>(fs,</span>
<span id="cb12-79"><a href="#cb12-79" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_sam_fr =</span> <span class="dv">5</span>, <span class="co"># a 5x is needed for a read count to be output in the table.</span></span>
<span id="cb12-80"><a href="#cb12-80" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_loc_fr =</span> <span class="fl">0.002</span>, <span class="co"># min fr of de-replicated sequence across all samples in the group.</span></span>
<span id="cb12-81"><a href="#cb12-81" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">"_([a-zA-Z0-9]*_[F|R])"</span>, <span class="co"># group by locus + F/R</span></span>
<span id="cb12-82"><a href="#cb12-82" aria-hidden="true" tabindex="-1"></a>            <span class="at">out_xlsx =</span> drep_xls)</span>
<span id="cb12-83"><a href="#cb12-83" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse</span></span>
<span id="cb12-84"><a href="#cb12-84" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(drep_tab[[<span class="dv">1</span>]], <span class="dv">3</span>))</span>
<span id="cb12-85"><a href="#cb12-85" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb12-86"><a href="#cb12-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-87"><a href="#cb12-87" aria-hidden="true" tabindex="-1"></a><span class="fu">## Agglomerate tables in an EXCEL</span></span>
<span id="cb12-88"><a href="#cb12-88" aria-hidden="true" tabindex="-1"></a>Tables above were combined in a unique EXCEL file that will be included in the manuscript as supplementary material.</span>
<span id="cb12-89"><a href="#cb12-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-92"><a href="#cb12-92" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb12-93"><a href="#cb12-93" aria-hidden="true" tabindex="-1"></a><span class="co"># ST1. tracked reads.</span></span>
<span id="cb12-94"><a href="#cb12-94" aria-hidden="true" tabindex="-1"></a><span class="co"># ST2. demultiplexed reads.</span></span>
<span id="cb12-95"><a href="#cb12-95" aria-hidden="true" tabindex="-1"></a><span class="co"># ST3. reads supporting variants.</span></span>
<span id="cb12-96"><a href="#cb12-96" aria-hidden="true" tabindex="-1"></a>suppl_tables <span class="ot">&lt;-</span></span>
<span id="cb12-97"><a href="#cb12-97" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(</span>
<span id="cb12-98"><a href="#cb12-98" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ST1. Tracked reads"</span> <span class="ot">=</span> reads_tracked,</span>
<span id="cb12-99"><a href="#cb12-99" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ST2. Demultiplexed- per locus"</span> <span class="ot">=</span> reads_demultiplx,</span>
<span id="cb12-100"><a href="#cb12-100" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ST3. Filtered variants"</span> <span class="ot">=</span> reads_variants</span>
<span id="cb12-101"><a href="#cb12-101" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb12-102"><a href="#cb12-102" aria-hidden="true" tabindex="-1"></a><span class="fu">write_xlsx</span>(suppl_tables,</span>
<span id="cb12-103"><a href="#cb12-103" aria-hidden="true" tabindex="-1"></a>           <span class="at">path =</span> <span class="st">"output/supplementary_tables.xlsx"</span>)</span>
<span id="cb12-104"><a href="#cb12-104" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb12-105"><a href="#cb12-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-106"><a href="#cb12-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-107"><a href="#cb12-107" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session Info</span></span>
<span id="cb12-108"><a href="#cb12-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-111"><a href="#cb12-111" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb12-112"><a href="#cb12-112" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb12-113"><a href="#cb12-113" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>